{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query some results about the manually annotated test-set\n",
    "\n",
    "Also contains some additional results (somewhat redundant with Parameterestimation.ipynb, but conceptually the basis thereof)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "complement = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\", \"N\": \"N\"}\n",
    "base_order = \"TCAG\"\n",
    "codon_list = [a + b + c for a in base_order for b in base_order for c in base_order]\n",
    "\n",
    "rna_codon_list = [c.replace(\"T\", \"U\") for c in codon_list]\n",
    "\n",
    "def codon2index(codon):\n",
    "    result = base_order.index(codon[0]) * 16\n",
    "    result += base_order.index(codon[1]) * 4\n",
    "    result += base_order.index(codon[2])\n",
    "    return result\n",
    "\n",
    "\n",
    "def rev_comp(sequence: str):\n",
    "    result = \"\"\n",
    "    for i in range(len(sequence)-1, -1, -1):\n",
    "        result += complement[sequence[i]]\n",
    "    return result\n",
    "\n",
    "class Feature:\n",
    "    def __init__(self, feature: str, start: int, end: int, strand: bool):\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "        self.feature = feature\n",
    "    \n",
    "class Gene:\n",
    "    def __init__(self, strand: bool):\n",
    "        self.strand = strand\n",
    "        self.features = [] # a sorted (!) list of CDS, introns and one UTR; sorted from 5' to 3', \n",
    "        # i.e. on reverse strand, the indices of subsequent features;\n",
    "        # Sorting is trusted to the outside world; maybe check in isValid!\n",
    "        \n",
    "    def append(self, feature: Feature):\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.features)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return \" \".join([\"%s%s %s %s%s\" % (\"[\" if f.strand else \"<\", \n",
    "                                           f.start, f.feature, f.end,\n",
    "                                           \">\" if f.strand else \"]\") \n",
    "                         for f in self.features])\n",
    "\n",
    "    def get_range(self):\n",
    "        return self.features[0].start, self.features[-1].end\n",
    "    \n",
    "    # sequence is nucleotide sequence from scaffolds.fasta\n",
    "    # this produces the concatenated cds -- potentially complementarily reversed (if strand is negative)\n",
    "    def get_contiguous_cds(self, sequence: str):\n",
    "        cds = \"\".join([sequence[(f.start-1):f.end] for f in self.features if f.feature == \"CDS\"])\n",
    "        if self.strand:\n",
    "            return cds\n",
    "        else:\n",
    "            return rev_comp(cds)\n",
    "        \n",
    "    def codon_usage(self, sequence: str):\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        general = np.zeros(64)\n",
    "        stop = np.zeros(64)\n",
    "        general_n = 0\n",
    "        for i in range(0, len(cds) - 21, 3):\n",
    "            general[codon2index(cds[i:i+3])] += 1\n",
    "            general_n += 1\n",
    "            \n",
    "        for i in range(len(cds) - 21, len(cds), 3):\n",
    "            stop[codon2index(cds[i:i+3])] += 1\n",
    "            \n",
    "        print(\"term: %s : %s*UGA\" % (cds[len(cds)-21:], stop[codon2index(\"TGA\")]))\n",
    "        return general, general_n, stop\n",
    "        \n",
    "    def count_UGA(self, sequence: str):\n",
    "        inframe = 0\n",
    "        outframe_in_cds = 0\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        for i in range(len(cds)-3):\n",
    "            if sequence[i:i+3] == \"TGA\":\n",
    "                if i % 3 == 0:\n",
    "                    inframe += 1\n",
    "                else:\n",
    "                    outframe_in_cds += 1\n",
    "        # REQUIRES this gene be valid\n",
    "        if self.strand:\n",
    "            utr = sequence[(self.features[-1].start-1):self.features[-1].end].count(\"TGA\")\n",
    "        else:\n",
    "            utr = sequence[(self.features[0].start-1):self.features[0].end].count(\"TCA\")\n",
    "        \n",
    "        \n",
    "        return inframe, outframe_in_cds, utr\n",
    "        \n",
    "    def get_exon_intron_utr_stats(self):\n",
    "        exon_lengths = []\n",
    "        intron_lengths = []\n",
    "        utr_index = -1 if self.strand else 0\n",
    "        utr_length = self.features[utr_index].end - self.features[utr_index].start + 1 # +1 bc. gff is inclusive-end\n",
    "        for f in self.features:\n",
    "            if f.feature == \"CDS\":\n",
    "                exon_lengths += [f.end - f.start + 1]\n",
    "            elif f.feature == \"intron\":\n",
    "                intron_lengths += [f.end - f.start + 1]\n",
    "        \n",
    "        return exon_lengths, intron_lengths, utr_length\n",
    "        \n",
    "    def is_valid(self):\n",
    "        if len(self.features) == 0:\n",
    "            print(\"! The gene is empty\")\n",
    "            return False\n",
    "        \n",
    "        utr_index = -1 if self.strand else 0\n",
    "        first_cds_index = 0 if self.strand else -1\n",
    "        \n",
    "        # a Gene must end in a 3'UTR\n",
    "        if self.features[utr_index].feature != \"three_prime_UTR\":\n",
    "            print(\"! 3' Terminal feature is not a UTR, but a %s\" % self.features[utr_index].feature)\n",
    "            return False\n",
    "        # a Gene must start with a CDS\n",
    "        if self.features[first_cds_index].feature != \"CDS\":\n",
    "            print(\"! 5' terminal feature is not a CDS, but a %s\" % self.features[first_cds_index].feature)\n",
    "            return False\n",
    "        \n",
    "        # Gene must start with CDS, end with UTR (len >= 2 so far), and can have [intron, CDS]-pairs in between \n",
    "        # (no two cds adjacent, nor two introns)\n",
    "        if len(self.features) % 2 != 0:\n",
    "            print(\"! uneven number of features\")\n",
    "            return False\n",
    "        \n",
    "        num_utrs = 0\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].feature == \"three_prime_UTR\":\n",
    "                num_utrs += 1\n",
    "            if self.features[index+1].feature == \"three_prime_UTR\":\n",
    "                num_utrs += 1\n",
    "                \n",
    "            if self.features[index].strand != self.strand or self.features[index+1].strand != self.strand:\n",
    "                print(\"! Strand disagreement within the gene\")\n",
    "                return False\n",
    "            \n",
    "            if self.strand:\n",
    "                if not (self.features[index].feature == \"CDS\" \\\n",
    "                        and self.features[index+1].feature in [\"intron\", \"three_prime_UTR\"]):\n",
    "                    print(\"! CDS-intron-CDS-UTR-pattern violated\")\n",
    "                    return False\n",
    "            else:\n",
    "                if not (self.features[index+1].feature == \"CDS\" \\\n",
    "                        and self.features[index].feature in [\"intron\", \"three_prime_UTR\"]):\n",
    "                    print(\"! CDS-intron-CDS-UTR-pattern violated\")\n",
    "                    return False\n",
    "            \n",
    "            index +=2\n",
    "            \n",
    "        if num_utrs != 1:\n",
    "            print(\"! Too many utrs: %s\" % num_utrs)\n",
    "            return False   \n",
    "        \n",
    "        return True\n",
    "\n",
    "class Annotation:\n",
    "    def __init__(self, name: str, start: int, end: int):\n",
    "        self.name = name\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.features = [] # The list of features; could also separate this by strand\n",
    "        self.genes = []\n",
    "    \n",
    "    def add(self, feature: Feature):\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].start > feature.end:\n",
    "                self.features.insert(index, feature)\n",
    "                return\n",
    "            index += 1\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def compile_genes(self) -> bool:\n",
    "        all_valid = True\n",
    "        self.genes = []\n",
    "        last_end = self.features[0].start - 1\n",
    "        current_gene = Gene(self.features[0].strand)\n",
    "        for f in self.features:\n",
    "            if f.start != last_end + 1: # there is a break\n",
    "                self.genes.append(current_gene)\n",
    "                all_valid = all_valid and current_gene.is_valid()\n",
    "                current_gene = Gene(f.strand)\n",
    "                \n",
    "            current_gene.append(f)\n",
    "            last_end = f.end\n",
    "        \n",
    "        if len(current_gene) > 0:\n",
    "            self.genes.append(current_gene)\n",
    "            all_valid = all_valid and current_gene.is_valid()\n",
    "        \n",
    "        return all_valid\n",
    "        \n",
    "        \n",
    "    def __str__(self):\n",
    "        return self.name + \":\\n\\t\" + (\"\\n\\t\".join([\"[%s, %s]%s%s\" % (f.start, f.end,\n",
    "                                                                     f.feature, \n",
    "                                                                     \"+\" if f.strand else \"-\") \n",
    "                                               for f in self.features]))\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in an annotations-file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "annotations = []\n",
    "with open(\"september-final-benchmarks/training.gff\") as annotation_file:\n",
    "    c = 0\n",
    "    f = 0\n",
    "    line = annotation_file.readline()\n",
    "    current = None\n",
    "    while line:\n",
    "        line = annotation_file.readline()[:-1]\n",
    "        content = line.split(\"\\t\")\n",
    "        if line.startswith(\"##sequence-region\"):\n",
    "            print(current)\n",
    "            if current is not None:\n",
    "                correct = current.compile_genes()\n",
    "                if not correct:\n",
    "                    f += 1\n",
    "                print(correct)\n",
    "                print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "                annotations.append(current)\n",
    "            print(\"-  \" * 36)\n",
    "            \n",
    "            current = Annotation(content[1], int(content[2]), int(content[3]))\n",
    "            c += 1\n",
    "        elif len(line) > 0 and not line.startswith(\"#\"):\n",
    "            current.add(Feature(content[2], int(content[3]), int(content[4]), content[6] == \"+\"))\n",
    "\n",
    "print(current.compile_genes())\n",
    "print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "annotations.append(current)\n",
    "print(c, \"=\", len(annotations), \"of which\", f, \"were faulty\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in the sequences from scaffolds and query number of inframe-TGA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "annotation_dict = {a.name:a for a in annotations}\n",
    "num_of_genes = np.sum([len(a.genes) for a in annotations])\n",
    "\n",
    "counts = np.zeros((num_of_genes,3))\n",
    "codon_usage_general = np.zeros(64)\n",
    "codon_usage_stop = np.zeros(64)\n",
    "codon_number = 0\n",
    "stop_codon_number = 0\n",
    "\n",
    "start_regions = []\n",
    "introns_by_length = {}\n",
    "\n",
    "current_header = None\n",
    "current_sequence = \"\"\n",
    "gene_index = 0\n",
    "with open(\"september-final-benchmarks/annotated_nodes.basic.train+test.fasta\") as scaffolds:\n",
    "    for row, line in enumerate(scaffolds):\n",
    "        if line.startswith(\">\"):\n",
    "            # Here query the analysis\n",
    "            if current_header is not None:\n",
    "                if current_header in annotation_dict:\n",
    "                    print(current_header)\n",
    "                    ann = annotation_dict[current_header]\n",
    "                    for gene in ann.genes:\n",
    "                        counts[gene_index, :] = np.array(gene.count_UGA(current_sequence))\n",
    "                        print(\"\\t\", counts[gene_index,:])\n",
    "                        if counts[gene_index,2] == 0:\n",
    "                            print(\"\\t\\t\\t\\t!>!>!>missing STOP\")\n",
    "                        cds_codon_counts, cds_num_of_codons, stop_region = gene.codon_usage(current_sequence)\n",
    "                        codon_number += cds_num_of_codons\n",
    "                        stop_codon_number += 7\n",
    "                        codon_usage_general += cds_codon_counts\n",
    "                        codon_usage_stop += stop_region\n",
    "                        \n",
    "                        gene_index += 1\n",
    "                        \n",
    "                        # Collect intron sequences:\n",
    "                        for f in [f for f in gene.features if f.feature == \"intron\"]:\n",
    "                            curr_intron = current_sequence[(f.start-1):f.end]\n",
    "                            curr_intron_length = f.end - f.start + 1\n",
    "                            if not f.strand:\n",
    "                                curr_intron = rev_comp(curr_intron)\n",
    "                            if curr_intron_length not in introns_by_length:\n",
    "                                introns_by_length[curr_intron_length] = [curr_intron]\n",
    "                            else:\n",
    "                                introns_by_length[curr_intron_length] += [curr_intron]\n",
    "                            \n",
    "                            print(\"intron: \", curr_intron)\n",
    "                            \n",
    "                        # Collect start-region:\n",
    "                        g_start, g_end = gene.get_range()\n",
    "                        if gene.strand:\n",
    "                            start_regions += [current_sequence[max(0, \n",
    "                                                                   g_start-1-24):min(len(current_sequence),\n",
    "                                                                                        g_start+2+9)]]\n",
    "                        else:\n",
    "                            start_regions += [rev_comp(current_sequence[max(0, \n",
    "                                                                            g_end-3-9):min(len(current_sequence),\n",
    "                                                                                           g_end+24)])]\n",
    "            \n",
    "            current_header = line[1:-1]\n",
    "            current_sequence = \"\"\n",
    "        else:\n",
    "            current_sequence += line[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start region: Looking for motifs\n",
    "\n",
    "This was done _before_ the final version of the parameter-estimation was implemented, and is the data-basis for the design decision"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\".\"*24 + \"ATG\" + \"-\"*9)\n",
    "start_region_bases = np.zeros((24+3+9, 4))\n",
    "# print(\"\\n\".join(start_regions))\n",
    "for r in [r for r in start_regions if len(r) == 24+3+9]:\n",
    "    # print(r)\n",
    "    for i in range(24+3+9):\n",
    "        start_region_bases[i, base_order.index(r[i])] += 1\n",
    "        \n",
    "print(start_region_bases / 153)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intron sequence analysis:\n",
    "\n",
    "Same as for start-region: this served to guide the model-design (and is thus here limited to the training-set, which was exclusively used for this, because the testing-set did not yet exist, and should not be used, not even indirectly, for the design of the model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "ls = [k for k in introns_by_length.keys()]\n",
    "ls.sort()\n",
    "\n",
    "motif_length = 10\n",
    "intron_heads = []\n",
    "intron_tails = []\n",
    "\n",
    "for l in ls:\n",
    "    # print(l)\n",
    "    for i in introns_by_length[l]:\n",
    "        intron_heads += [i[:motif_length]]\n",
    "        intron_tails += [i[-motif_length:]]\n",
    "        print(\"\\t\" + i + \":\\t%sT %sC %sA %sG\" % (i.count(\"T\")/l, i.count(\"C\")/l, i.count(\"A\")/l, i.count(\"G\")/l))\n",
    "        \n",
    "base_freq_head = np.zeros((4, motif_length))\n",
    "base_freq_tail = np.zeros((4, motif_length))\n",
    "\n",
    "n = len(intron_heads)\n",
    "\n",
    "for i in range(n):\n",
    "    for j in range(motif_length):\n",
    "        base_freq_head[base_order.index(intron_heads[i][j]), j] += 1\n",
    "        base_freq_tail[base_order.index(intron_tails[i][j]), j] += 1\n",
    "\n",
    "print()\n",
    "print(\"  \" + \"          \".join(base_order))\n",
    "print((base_freq_head/n).T)\n",
    "print(\"   ...\")\n",
    "print((base_freq_tail/n).T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(counts, label=[\"inframe\", \"outframe_in_cds\", \"utr\"])\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def positional_base_usage(fs):\n",
    "    positional_fs = np.zeros((3, 4))\n",
    "    for c in codon_list:\n",
    "        for i in range(3): # position in codon\n",
    "            positional_fs[i, base_order.index(c[i])] += fs[codon2index(c)]\n",
    "    return positional_fs\n",
    "\n",
    "print(positional_base_usage(codon_usage_general / codon_number))\n",
    "print(positional_base_usage(codon_usage_stop / stop_codon_number))\n",
    "w = 0.5\n",
    "plt.bar(np.arange(64) - w/2, codon_usage_general / codon_number, width=w)\n",
    "plt.bar(np.arange(64) + w/2, codon_usage_stop / stop_codon_number, width=w)\n",
    "plt.xticks(range(64), rna_codon_list, rotation=90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = [\"inframe UGA\", \"outframe UGA in CDS\", \"UGA in 3'UTR\"]\n",
    "bins = [40, 40, 8]\n",
    "means = np.mean(counts, axis=0)\n",
    "medians = np.median(counts, axis=0)\n",
    "\n",
    "for i in range(3):\n",
    "    plt.subplot(1,3,i+1)\n",
    "    plt.hist(counts[:,i], bins=bins[i],edgecolor=\"k\",color=\"w\", linewidth=0.5, align='left')\n",
    "    plt.axvline(means[i],linestyle=\"dashed\", color=\"k\", linewidth=0.5)\n",
    "    plt.axvline(medians[i], color=\"r\", linewidth=0.5)\n",
    "    plt.title(names[i])\n",
    "    plt.xlabel(\"#UGA\")\n",
    "    plt.ylabel(\"#genes\")\n",
    "\n",
    "print(\"mean #inframe UGA:\", np.mean(counts[:,0]))\n",
    "print(\"median #inframe UGA:\", np.median(counts[:,0]))\n",
    "print(\"\\nmean #outframe UGA in CDS:\", np.mean(counts[:,1]))\n",
    "print(\"median #outframe UGA in CDS:\", np.median(counts[:,1]))\n",
    "print(\"\\nmean #UGA in 3'UTR:\", np.mean(counts[:,2]))\n",
    "print(\"median #UGA in 3'UTR:\", np.median(counts[:,2]))\n",
    "\n",
    "uga_counts = counts\n",
    "\n",
    "plt.tight_layout(h_pad=-30)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_lengths = []\n",
    "exon_lengths = []\n",
    "utr_lengths = []\n",
    "intron_counts = []\n",
    "for ann in annotations:\n",
    "    for gene in ann.genes:\n",
    "        el, il, ul = gene.get_exon_intron_utr_stats()\n",
    "        exon_lengths += el\n",
    "        intron_lengths += il\n",
    "        if (len(il) > 0 and np.max(il) > 30) or ul > 100:\n",
    "            print(ann.name, il, ul)\n",
    "            # Can see that the two largest ones (58, 85) are from 615, \n",
    "            # and are not very trusted (coverage-signal is ambiguous)\n",
    "        utr_lengths += [ul]\n",
    "        intron_counts += [len(il)]\n",
    "\n",
    "datasets = [exon_lengths, utr_lengths, intron_counts, intron_lengths]\n",
    "names = [\"exon lengths\", \"3'UTR lengths\", \"intron counts\", \"intron lengths\"]\n",
    "bins = [30, 30, 5, 10]\n",
    "        \n",
    "for i in range(4):        \n",
    "    plt.subplot(2,2,i+1) \n",
    "    plt.hist(datasets[i], bins=bins[i], edgecolor=\"k\",color=\"w\", linewidth=0.5, align='left')\n",
    "    \n",
    "    mean = np.mean(datasets[i])\n",
    "    median = np.median(datasets[i])\n",
    "    print(\"\\n%s:\\tmean=%s\\tmedian=%s\" % (names[i], mean, median))\n",
    "    plt.axvline(mean,linestyle=\"dashed\", color=\"k\", linewidth=0.5)\n",
    "    plt.axvline(median, color=\"r\", linewidth=0.5)\n",
    "    plt.title(names[i])\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "# intron lengths, exon lengths, utr lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_introns = len([g for g in intron_counts if g == 0])\n",
    "print(no_introns, \"of\", len(intron_counts), \"genes have no intron, that's\", 100*no_introns/len(intron_counts), \"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_length_distr = np.bincount(intron_lengths)[:30]\n",
    "print(\",\".join([str(p) for p in intron_length_distr / np.sum(intron_length_distr)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing into csv for plotting in LaTe$\\chi$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = [40, 40, 30, 5]\n",
    "datasets = [uga_counts.T[0], uga_counts.T[1], exon_lengths, intron_counts]\n",
    "filenames = [\"UGA_inframe\", \"UGA_outframe\", \"exonlengths\", \"intron_counts\"]\n",
    "for i in range(4):\n",
    "    with open(\"out/annotated_\" + filenames[i] + \".csv\", \"w\") as out:\n",
    "        result = \"xlower,y\\n\"\n",
    "        hist, bin_edges = np.histogram(datasets[i], bins=bins[i])\n",
    "        bar_width = [bin_edges[i+1] - bin_edges[i] for i in range(len(hist))][0]\n",
    "        print(filenames[i], len(hist), bar_width * 2)\n",
    "        result += \"\\n\".join([\"%s,%s\" % (l,c) for (c, l) in zip(hist, bin_edges[:-1])])\n",
    "        result += \"\\n%s,0\" % bin_edges[-1]\n",
    "        print(result)\n",
    "        out.write(result)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
