{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameter-file generation\n",
    "### Input: raw sequences (contigs.fasta)\n",
    "### Input: annotation (annotation.gff3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_filename = \"training_my_ghmm.gff\" # \"scaffolds.07-31.cleaned.gff\"\n",
    "sequence_filename = \"scaffolds.fasta\"\n",
    "output_filename = \"parameters.new.automatic.properties\"\n",
    "\n",
    "parameters = dict()\n",
    "parameters[\"start_region_size\"] = 6\n",
    "parameters[\"stop_region_size\"] = 24"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output: parameters.properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hyperparameters: To be chosen manually\n",
    " - `\"ncs_flank_size\"`: How many nt upstream of left and downstream of right end can be considered NCS?\n",
    " - `\"ncs_mean_length\"`: Manual estimate of how long an intergenic region is on average (in nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparameters = {}\n",
    "hyperparameters[\"ncs_flank_size\"] = 300\n",
    "hyperparameters[\"ncs_mean_length\"] = 1000\n",
    "hyperparameters[\"sds_size\"] = 5\n",
    "hyperparameters[\"sas_size\"] = 4\n",
    "\n",
    "hyperparameters[\"intron_length_tolerance\"] = 4 # how far to extend beyond the seen range of intron lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene format: CDS(begins with ATG) - intron - CDS - intron - CDS - stop_codon\n",
    "# i.e.: CDS[starts with ATG] - {intron - CDS}* - stop_codon\n",
    "# (thus, all genes have an even amount of features; start-codon is not treated separately)\n",
    "# The annotation may however also contain other features, especially (for me): three_prime_UTR \n",
    "# -> when reading the file, replace this! (might also have to ignore some features)\n",
    "\n",
    "complement = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\", \"N\": \"N\"}\n",
    "base_order = \"TCAG\"\n",
    "codon_list = [a + b + c for a in base_order for b in base_order for c in base_order]\n",
    "\n",
    "rna_codon_list = [c.replace(\"T\", \"U\") for c in codon_list]\n",
    "\n",
    "def codon2index(codon):\n",
    "    result = base_order.index(codon[0]) * 16\n",
    "    result += base_order.index(codon[1]) * 4\n",
    "    result += base_order.index(codon[2])\n",
    "    return result\n",
    "\n",
    "def positional_base_usage(fs):\n",
    "    positional_fs = np.zeros((3, 4))\n",
    "    for c in codon_list:\n",
    "        for i in range(3): # position in codon\n",
    "            positional_fs[i, base_order.index(c[i])] += fs[codon2index(c)]\n",
    "    return positional_fs\n",
    "\n",
    "# Simple approximation based on marginals\n",
    "def fit_codon_usage(counts):\n",
    "    positional_fs = positional_base_usage(counts)\n",
    "    independent_codon_approximation = np.zeros(64)\n",
    "    for c in codon_list:\n",
    "        independent_codon_approximation[codon2index(c)] = np.product([positional_fs[i, \n",
    "                                                                                    base_order.index(c[i])] \n",
    "                                                                      for i in range(3)])\n",
    "    \n",
    "    return independent_codon_approximation\n",
    "\n",
    "def rev_comp(sequence: str):\n",
    "    result = \"\"\n",
    "    for i in range(len(sequence)-1, -1, -1):\n",
    "        result += complement[sequence[i]]\n",
    "    return result\n",
    "\n",
    "class Feature:\n",
    "    def __init__(self, feature: str, start: int, end: int, strand: bool):\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "        self.feature = feature\n",
    "    \n",
    "class Gene:\n",
    "    def __init__(self, strand: bool):\n",
    "        self.strand = strand\n",
    "        self.features = [] # a sorted (!) list of CDS, introns and one UTR; sorted from 5' to 3', \n",
    "        # i.e. on reverse strand, the indices of subsequent features;\n",
    "        # Sorting is trusted to the outside world; maybe check in isValid!\n",
    "        \n",
    "    def append(self, feature: Feature):\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.features)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return \" \".join([\"%s%s %s %s%s\" % (\"[\" if f.strand else \"<\", \n",
    "                                           f.start, f.feature, f.end,\n",
    "                                           \">\" if f.strand else \"]\") \n",
    "                         for f in self.features])\n",
    "\n",
    "    def get_range(self):\n",
    "        return self.features[0].start, self.features[-1].end\n",
    "    \n",
    "    # sequence is nucleotide sequence from scaffolds.fasta\n",
    "    # this produces the concatenated cds -- potentially complementarily reversed (if strand is negative)\n",
    "    def get_contiguous_cds(self, sequence: str):\n",
    "        cds = \"\".join([sequence[(f.start-1):f.end] for f in self.features if f.feature == \"CDS\"])\n",
    "        if self.strand:\n",
    "            return cds\n",
    "        else:\n",
    "            return rev_comp(cds)\n",
    "        \n",
    "    def codon_usage(self, sequence: str):\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        general = np.zeros(64)\n",
    "        stop = np.zeros(64)\n",
    "        general_n = 0\n",
    "        for i in range(0, len(cds) - 21, 3):\n",
    "            general[codon2index(cds[i:i+3])] += 1\n",
    "            general_n += 1\n",
    "            \n",
    "        for i in range(len(cds) - 21, len(cds), 3):\n",
    "            stop[codon2index(cds[i:i+3])] += 1\n",
    "            \n",
    "        print(\"term: %s : %s*UGA\" % (cds[len(cds)-21:], stop[codon2index(\"TGA\")]))\n",
    "        return general, general_n, stop\n",
    "        \n",
    "    def count_UGA(self, sequence: str):\n",
    "        inframe = 0\n",
    "        outframe_in_cds = 0\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        for i in range(len(cds)-3):\n",
    "            if sequence[i:i+3] == \"TGA\":\n",
    "                if i % 3 == 0:\n",
    "                    inframe += 1\n",
    "                else:\n",
    "                    outframe_in_cds += 1\n",
    "        # REQUIRES this gene be valid\n",
    "        # Pure sanity check\n",
    "        if self.strand:\n",
    "            stop = sequence[(self.features[-1].start-1):self.features[-1].end].count(\"TGA\")\n",
    "        else:\n",
    "            stop = sequence[(self.features[0].start-1):self.features[0].end].count(\"TCA\")\n",
    "        \n",
    "        \n",
    "        return inframe, outframe_in_cds, stop\n",
    "        \n",
    "    def get_exon_intron_stats(self):\n",
    "        \"\"\"\n",
    "        returns the 'exon'-lengths and intron-lengths found in this gene. The intron-lengths are exactly that, \n",
    "        but the exon-lengths are cleaned of everything that is not modeled by the CDS-state in the GHMM: \n",
    "        start-codon, stop-region, intron-flanking interrupted codons (which makes little difference, but still)\n",
    "        \"\"\"\n",
    "        exon_lengths = []\n",
    "        intron_lengths = []\n",
    "        i = 0 # index of feature\n",
    "        after_intron = 0 # how many bases of a codon come after the last intron (codon interrupted)\n",
    "        \n",
    "        for f in self.features:\n",
    "            if f.feature == \"CDS\":\n",
    "                length = f.end - f.start + 1 - after_intron # +1 bc. gff is inclusive-end\n",
    "                if self.strand:\n",
    "                    if i==0:\n",
    "                        length -= 3 # cut off start-codon\n",
    "                    elif i==len(self.features):\n",
    "                        length -= parameters[\"stop_region_size\"] - 3 # cut off stop-region, but not stop-codon\n",
    "                else:\n",
    "                    if i==0:\n",
    "                        length -= parameters[\"stop_region_size\"] - 3\n",
    "                    elif i==len(self.features):\n",
    "                        length -= 3 # cut off start-codon\n",
    "                after_intron = (3 - (length % 3)) % 3 # from deviation from threevenness: get what needs to be cut off next\n",
    "                length -= (3-after_intron) % 3 # cut off the part before the intron of the current exon\n",
    "                exon_lengths += [length]\n",
    "            elif f.feature == \"intron\":\n",
    "                intron_lengths += [f.end - f.start + 1]\n",
    "            i += 1\n",
    "        \n",
    "        return exon_lengths, intron_lengths\n",
    "        \n",
    "    def is_valid(self):\n",
    "        if len(self.features) == 0:\n",
    "            print(\"! The gene is empty\")\n",
    "            return False\n",
    "        \n",
    "        stop_index = -1 if self.strand else 0\n",
    "        first_cds_index = 0 if self.strand else -1\n",
    "        \n",
    "        # a Gene must end in a stop_codon\n",
    "        if self.features[stop_index].feature != \"stop_codon\":\n",
    "            print(\"! 3' Terminal feature is not a stop_codon, but a %s\" % self.features[stop_index].feature)\n",
    "            return False\n",
    "        # a Gene must start with a CDS\n",
    "        if self.features[first_cds_index].feature != \"CDS\":\n",
    "            print(\"! 5' terminal feature is not a CDS, but a %s\" % self.features[first_cds_index].feature)\n",
    "            return False\n",
    "        \n",
    "        # Gene must start with CDS, end with UTR (len >= 2 so far), and can have [intron, CDS]-pairs in between \n",
    "        # (no two cds adjacent, nor two introns)\n",
    "        if len(self.features) % 2 != 0:\n",
    "            print(\"! uneven number of features\")\n",
    "            return False\n",
    "        \n",
    "        num_stops = 0\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].feature == \"stop_codon\":\n",
    "                num_stops += 1\n",
    "            if self.features[index+1].feature == \"stop_codon\":\n",
    "                num_stops += 1\n",
    "                \n",
    "            if self.features[index].strand != self.strand or self.features[index+1].strand != self.strand:\n",
    "                print(\"! Strand disagreement within the gene\")\n",
    "                return False\n",
    "            \n",
    "            if self.strand:\n",
    "                if not (self.features[index].feature == \"CDS\" \\\n",
    "                        and self.features[index+1].feature in [\"intron\", \"stop_codon\"]):\n",
    "                    print(\"! CDS-intron-CDS-stop_codon-pattern violated\")\n",
    "                    return False\n",
    "            else:\n",
    "                if not (self.features[index+1].feature == \"CDS\" \\\n",
    "                        and self.features[index].feature in [\"intron\", \"stop_codon\"]):\n",
    "                    print(\"! CDS-intron-CDS-stop_codon-pattern violated\")\n",
    "                    return False\n",
    "            \n",
    "            index +=2\n",
    "            \n",
    "        if num_stops != 1:\n",
    "            print(\"! Too many stops: %s\" % num_stops)\n",
    "            return False   \n",
    "        \n",
    "        return True\n",
    "\n",
    "class Annotation:\n",
    "    def __init__(self, name: str, start: int, end: int):\n",
    "        self.name = name\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.features = [] # The list of features; could also separate this by strand\n",
    "        self.genes = []\n",
    "    \n",
    "    def add(self, feature: Feature):\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].start > feature.end:\n",
    "                self.features.insert(index, feature)\n",
    "                return\n",
    "            index += 1\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def compile_genes(self) -> bool:\n",
    "        all_valid = True\n",
    "        self.genes = []\n",
    "        last_end = self.features[0].start - 1\n",
    "        current_gene = Gene(self.features[0].strand)\n",
    "        for f in self.features:\n",
    "            if f.start != last_end + 1: # there is a break\n",
    "                self.genes.append(current_gene)\n",
    "                all_valid = all_valid and current_gene.is_valid()\n",
    "                current_gene = Gene(f.strand)\n",
    "                \n",
    "            current_gene.append(f)\n",
    "            last_end = f.end\n",
    "        \n",
    "        if len(current_gene) > 0:\n",
    "            self.genes.append(current_gene)\n",
    "            all_valid = all_valid and current_gene.is_valid()\n",
    "        \n",
    "        return all_valid\n",
    "        \n",
    "        \n",
    "    def __str__(self):\n",
    "        return self.name + \":\\n\\t\" + (\"\\n\\t\".join([\"[%s, %s]%s%s\" % (f.start, f.end,\n",
    "                                                                     f.feature, \n",
    "                                                                     \"+\" if f.strand else \"-\") \n",
    "                                               for f in self.features]))\n",
    "    \n",
    "def format_matrix(matrix):\n",
    "    return \"{{%s}}\" % \"}\\\\\\n  {\".join([\", \".join([str(e) for e in row]) for row in matrix])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input: Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_64_length_178741_cov_1.661034:\n",
      "\t[129866, 131170]CDS+\n",
      "\t[131171, 131173]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[129866 CDS 131170> [131171 stop_codon 131173>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_65_length_177181_cov_1.879014:\n",
      "\t[59358, 59360]stop_codon-\n",
      "\t[59361, 59484]CDS-\n",
      "\t[59485, 59501]intron-\n",
      "\t[59502, 60271]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<59358 stop_codon 59360] <59361 CDS 59484] <59485 intron 59501] <59502 CDS 60271]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_69_length_175562_cov_1.893300:\n",
      "\t[4520, 6490]CDS+\n",
      "\t[6491, 6493]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[4520 CDS 6490> [6491 stop_codon 6493>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_90_length_157735_cov_1.661280:\n",
      "\t[90564, 91550]CDS+\n",
      "\t[91551, 91553]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[90564 CDS 91550> [91551 stop_codon 91553>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_103_length_148054_cov_1.718544:\n",
      "\t[45719, 47784]CDS+\n",
      "\t[47785, 47804]intron+\n",
      "\t[47805, 47907]CDS+\n",
      "\t[47908, 47910]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[45719 CDS 47784> [47785 intron 47804> [47805 CDS 47907> [47908 stop_codon 47910>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_127_length_132101_cov_1.744328:\n",
      "\t[31550, 31552]stop_codon-\n",
      "\t[31553, 32003]CDS-\n",
      "\t[32004, 32021]intron-\n",
      "\t[32022, 33388]CDS-\n",
      "\t[124200, 127031]CDS+\n",
      "\t[127032, 127034]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<31550 stop_codon 31552] <31553 CDS 32003] <32004 intron 32021] <32022 CDS 33388]\n",
      "\t[124200 CDS 127031> [127032 stop_codon 127034>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_134_length_130250_cov_1.655841:\n",
      "\t[40175, 40177]stop_codon-\n",
      "\t[40178, 41827]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<40175 stop_codon 40177] <40178 CDS 41827]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_135_length_129955_cov_1.831608:\n",
      "\t[129029, 129031]stop_codon-\n",
      "\t[129032, 129114]CDS-\n",
      "\t[129115, 129130]intron-\n",
      "\t[129131, 129905]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<129029 stop_codon 129031] <129032 CDS 129114] <129115 intron 129130] <129131 CDS 129905]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_150_length_122897_cov_1.750631:\n",
      "\t[69429, 70823]CDS+\n",
      "\t[70824, 70826]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[69429 CDS 70823> [70824 stop_codon 70826>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_155_length_121415_cov_1.724095:\n",
      "\t[114893, 114895]stop_codon-\n",
      "\t[114896, 116134]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<114893 stop_codon 114895] <114896 CDS 116134]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_160_length_120340_cov_1.623793:\n",
      "\t[66695, 66697]stop_codon-\n",
      "\t[66698, 68068]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<66695 stop_codon 66697] <66698 CDS 68068]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_176_length_115965_cov_1.774081:\n",
      "\t[53967, 53969]stop_codon-\n",
      "\t[53970, 54069]CDS-\n",
      "\t[54070, 54085]intron-\n",
      "\t[54086, 54184]CDS-\n",
      "\t[54185, 54200]intron-\n",
      "\t[54201, 56140]CDS-\n",
      "\t[56141, 56156]intron-\n",
      "\t[56157, 56276]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<53967 stop_codon 53969] <53970 CDS 54069] <54070 intron 54085] <54086 CDS 54184] <54185 intron 54200] <54201 CDS 56140] <56141 intron 56156] <56157 CDS 56276]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_177_length_115532_cov_1.858698:\n",
      "\t[22322, 23134]CDS+\n",
      "\t[23135, 23137]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[22322 CDS 23134> [23135 stop_codon 23137>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_206_length_105986_cov_1.546198:\n",
      "\t[53349, 54962]CDS+\n",
      "\t[54963, 54965]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[53349 CDS 54962> [54963 stop_codon 54965>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_210_length_104547_cov_1.706570:\n",
      "\t[19790, 20587]CDS+\n",
      "\t[20588, 20603]intron+\n",
      "\t[20604, 21023]CDS+\n",
      "\t[21024, 21026]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[19790 CDS 20587> [20588 intron 20603> [20604 CDS 21023> [21024 stop_codon 21026>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_217_length_103446_cov_1.848866:\n",
      "\t[27795, 29282]CDS+\n",
      "\t[29283, 29285]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[27795 CDS 29282> [29283 stop_codon 29285>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_227_length_100883_cov_2.022033:\n",
      "\t[11622, 11624]stop_codon-\n",
      "\t[11625, 11707]CDS-\n",
      "\t[11708, 11723]intron-\n",
      "\t[11724, 13119]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<11622 stop_codon 11624] <11625 CDS 11707] <11708 intron 11723] <11724 CDS 13119]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_250_length_96914_cov_1.797266:\n",
      "\t[42022, 42024]stop_codon-\n",
      "\t[42025, 43212]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<42022 stop_codon 42024] <42025 CDS 43212]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_265_length_94667_cov_1.884356:\n",
      "\t[19307, 19342]CDS+\n",
      "\t[19343, 19362]intron+\n",
      "\t[19363, 19789]CDS+\n",
      "\t[19790, 19792]stop_codon+\n",
      "\t[39745, 40593]CDS+\n",
      "\t[40594, 40613]intron+\n",
      "\t[40614, 40685]CDS+\n",
      "\t[40686, 40688]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[19307 CDS 19342> [19343 intron 19362> [19363 CDS 19789> [19790 stop_codon 19792>\n",
      "\t[39745 CDS 40593> [40594 intron 40613> [40614 CDS 40685> [40686 stop_codon 40688>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_267_length_94345_cov_1.719173:\n",
      "\t[12966, 12968]stop_codon-\n",
      "\t[12969, 14294]CDS-\n",
      "\t[14295, 14312]intron-\n",
      "\t[14313, 14717]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<12966 stop_codon 12968] <12969 CDS 14294] <14295 intron 14312] <14313 CDS 14717]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_273_length_93569_cov_1.730303:\n",
      "\t[65637, 65639]stop_codon-\n",
      "\t[65640, 71990]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<65637 stop_codon 65639] <65640 CDS 71990]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_279_length_92227_cov_1.518132:\n",
      "\t[19650, 19652]stop_codon-\n",
      "\t[19653, 20615]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<19650 stop_codon 19652] <19653 CDS 20615]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_301_length_87574_cov_1.953023:\n",
      "\t[63192, 63194]stop_codon-\n",
      "\t[63195, 65252]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<63192 stop_codon 63194] <63195 CDS 65252]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_324_length_81694_cov_1.607292:\n",
      "\t[58740, 59527]CDS+\n",
      "\t[59528, 59543]intron+\n",
      "\t[59544, 59667]CDS+\n",
      "\t[59668, 59670]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[58740 CDS 59527> [59528 intron 59543> [59544 CDS 59667> [59668 stop_codon 59670>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_358_length_77800_cov_1.675254:\n",
      "\t[26268, 30059]CDS+\n",
      "\t[30060, 30062]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[26268 CDS 30059> [30060 stop_codon 30062>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_379_length_75420_cov_1.946927:\n",
      "\t[25306, 25308]stop_codon-\n",
      "\t[25309, 26295]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25306 stop_codon 25308] <25309 CDS 26295]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_389_length_74397_cov_1.930066:\n",
      "\t[21604, 21606]stop_codon-\n",
      "\t[21607, 22785]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<21604 stop_codon 21606] <21607 CDS 22785]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_426_length_70982_cov_1.681053:\n",
      "\t[24670, 24672]stop_codon-\n",
      "\t[24673, 25131]CDS-\n",
      "\t[25132, 25147]intron-\n",
      "\t[25148, 26461]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<24670 stop_codon 24672] <24673 CDS 25131] <25132 intron 25147] <25148 CDS 26461]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_427_length_70843_cov_2.049663:\n",
      "\t[36168, 37112]CDS+\n",
      "\t[37113, 37115]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[36168 CDS 37112> [37113 stop_codon 37115>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_450_length_68074_cov_1.636967:\n",
      "\t[53953, 57489]CDS+\n",
      "\t[57490, 57492]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[53953 CDS 57489> [57490 stop_codon 57492>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_469_length_66432_cov_1.646482:\n",
      "\t[39237, 40199]CDS+\n",
      "\t[40200, 40215]intron+\n",
      "\t[40216, 40458]CDS+\n",
      "\t[40459, 40461]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[39237 CDS 40199> [40200 intron 40215> [40216 CDS 40458> [40459 stop_codon 40461>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_476_length_65419_cov_1.711113:\n",
      "\t[26840, 28468]CDS+\n",
      "\t[28469, 28471]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[26840 CDS 28468> [28469 stop_codon 28471>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_479_length_65262_cov_1.758118:\n",
      "\t[58933, 58935]stop_codon-\n",
      "\t[58936, 59031]CDS-\n",
      "\t[59032, 59048]intron-\n",
      "\t[59049, 59153]CDS-\n",
      "\t[59154, 59171]intron-\n",
      "\t[59172, 61685]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<58933 stop_codon 58935] <58936 CDS 59031] <59032 intron 59048] <59049 CDS 59153] <59154 intron 59171] <59172 CDS 61685]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_485_length_64803_cov_1.657740:\n",
      "\t[36500, 36502]stop_codon-\n",
      "\t[36503, 38739]CDS-\n",
      "\t[38740, 38758]intron-\n",
      "\t[38759, 39275]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<36500 stop_codon 36502] <36503 CDS 38739] <38740 intron 38758] <38759 CDS 39275]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_487_length_64315_cov_1.673226:\n",
      "\t[57491, 57493]stop_codon-\n",
      "\t[57494, 59107]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<57491 stop_codon 57493] <57494 CDS 59107]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_518_length_61243_cov_1.724180:\n",
      "\t[2921, 2923]stop_codon-\n",
      "\t[2924, 3550]CDS-\n",
      "\t[3551, 3569]intron-\n",
      "\t[3570, 3656]CDS-\n",
      "\t[3657, 3672]intron-\n",
      "\t[3673, 5586]CDS-\n",
      "\t[5587, 5603]intron-\n",
      "\t[5604, 6041]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2921 stop_codon 2923] <2924 CDS 3550] <3551 intron 3569] <3570 CDS 3656] <3657 intron 3672] <3673 CDS 5586] <5587 intron 5603] <5604 CDS 6041]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_547_length_58581_cov_1.583467:\n",
      "\t[52458, 52460]stop_codon-\n",
      "\t[52461, 53795]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<52458 stop_codon 52460] <52461 CDS 53795]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_555_length_57934_cov_1.702821:\n",
      "\t[16986, 18428]CDS+\n",
      "\t[18429, 18431]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[16986 CDS 18428> [18429 stop_codon 18431>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_585_length_55535_cov_1.790265:\n",
      "\t[3631, 3633]stop_codon-\n",
      "\t[3634, 5637]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<3631 stop_codon 3633] <3634 CDS 5637]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_632_length_52289_cov_1.832982:\n",
      "\t[17974, 19392]CDS+\n",
      "\t[19393, 19395]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[17974 CDS 19392> [19393 stop_codon 19395>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_712_length_47909_cov_1.617806:\n",
      "\t[7720, 7722]stop_codon-\n",
      "\t[7723, 7869]CDS-\n",
      "\t[7870, 7885]intron-\n",
      "\t[7886, 8755]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<7720 stop_codon 7722] <7723 CDS 7869] <7870 intron 7885] <7886 CDS 8755]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_721_length_47096_cov_1.882050:\n",
      "\t[2161, 2163]stop_codon-\n",
      "\t[2164, 2232]CDS-\n",
      "\t[2233, 2249]intron-\n",
      "\t[2250, 3122]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2161 stop_codon 2163] <2164 CDS 2232] <2233 intron 2249] <2250 CDS 3122]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_725_length_46739_cov_1.861538:\n",
      "\t[11122, 11124]stop_codon-\n",
      "\t[11125, 12450]CDS-\n",
      "\t[12451, 12467]intron-\n",
      "\t[12468, 12872]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<11122 stop_codon 11124] <11125 CDS 12450] <12451 intron 12467] <12468 CDS 12872]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_726_length_46702_cov_1.652088:\n",
      "\t[26570, 26572]stop_codon-\n",
      "\t[26573, 26671]CDS-\n",
      "\t[26672, 26693]intron-\n",
      "\t[26694, 26813]CDS-\n",
      "\t[26814, 26832]intron-\n",
      "\t[26833, 29655]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<26570 stop_codon 26572] <26573 CDS 26671] <26672 intron 26693] <26694 CDS 26813] <26814 intron 26832] <26833 CDS 29655]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_762_length_45064_cov_1.971004:\n",
      "\t[40763, 42133]CDS+\n",
      "\t[42134, 42136]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[40763 CDS 42133> [42134 stop_codon 42136>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_893_length_39578_cov_1.648298:\n",
      "\t[14064, 15803]CDS+\n",
      "\t[15804, 15806]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[14064 CDS 15803> [15804 stop_codon 15806>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_896_length_39504_cov_1.683648:\n",
      "\t[25489, 25491]stop_codon-\n",
      "\t[25492, 25942]CDS-\n",
      "\t[25943, 25959]intron-\n",
      "\t[25960, 27338]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25489 stop_codon 25491] <25492 CDS 25942] <25943 intron 25959] <25960 CDS 27338]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_926_length_38670_cov_1.606881:\n",
      "\t[16573, 16575]stop_codon-\n",
      "\t[16576, 17703]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<16573 stop_codon 16575] <16576 CDS 17703]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1083_length_33926_cov_1.664753:\n",
      "\t[25285, 25287]stop_codon-\n",
      "\t[25288, 25349]CDS-\n",
      "\t[25350, 25365]intron-\n",
      "\t[25366, 27673]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25285 stop_codon 25287] <25288 CDS 25349] <25350 intron 25365] <25366 CDS 27673]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1087_length_33881_cov_1.716982:\n",
      "\t[13033, 13035]stop_codon-\n",
      "\t[13036, 13080]CDS-\n",
      "\t[13081, 13106]intron-\n",
      "\t[13107, 13565]CDS-\n",
      "\t[13566, 13588]intron-\n",
      "\t[13589, 25447]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<13033 stop_codon 13035] <13036 CDS 13080] <13081 intron 13106] <13107 CDS 13565] <13566 intron 13588] <13589 CDS 25447]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1178_length_31386_cov_1.635561:\n",
      "\t[9831, 10628]CDS+\n",
      "\t[10629, 10644]intron+\n",
      "\t[10645, 11064]CDS+\n",
      "\t[11065, 11067]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[9831 CDS 10628> [10629 intron 10644> [10645 CDS 11064> [11065 stop_codon 11067>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1721_length_22983_cov_1.829424:\n",
      "\t[15310, 16314]CDS+\n",
      "\t[16315, 16317]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[15310 CDS 16314> [16315 stop_codon 16317>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1864_length_21573_cov_1.733703:\n",
      "\t[10945, 12819]CDS+\n",
      "\t[12820, 12822]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[10945 CDS 12819> [12820 stop_codon 12822>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_2222_length_18295_cov_4.113496:\n",
      "\t[358, 360]stop_codon-\n",
      "\t[361, 3405]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<358 stop_codon 360] <361 CDS 3405]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_3449_length_12448_cov_1.887347:\n",
      "\t[8528, 8530]stop_codon-\n",
      "\t[8531, 10114]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<8528 stop_codon 8530] <8531 CDS 10114]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "True\n",
      "Genes:\n",
      "\t[57 CDS 1380> [1381 stop_codon 1383>\n",
      "56 = 56 of which 0 were faulty\n",
      "There are 58 genes annotated\n"
     ]
    }
   ],
   "source": [
    "annotations = []\n",
    "with open(annotation_filename) as annotation_file:\n",
    "    c = 0\n",
    "    f = 0\n",
    "    line = annotation_file.readline()\n",
    "    current = None\n",
    "    while line:\n",
    "        line = annotation_file.readline()[:-1]\n",
    "        content = line.split(\"\\t\")\n",
    "        if line.startswith(\"##sequence-region\"):\n",
    "            print(current)\n",
    "            if current is not None:\n",
    "                correct = current.compile_genes()\n",
    "                if not correct:\n",
    "                    f += 1\n",
    "                print(\"+ valid gene structure\" if correct else \"/!\\\\ invalid gene structure\")\n",
    "                print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "                annotations.append(current)\n",
    "            print(\"-  \" * 36)\n",
    "            \n",
    "            current = Annotation(content[1], int(content[2]), int(content[3]))\n",
    "            c += 1\n",
    "        elif len(line) > 0 and not line.startswith(\"#\"):\n",
    "            # Transform 3'UTR-annotation into stop-codons!\n",
    "            feature = content[2]\n",
    "            forward = content[6] == \"+\"\n",
    "            start = int(content[3])\n",
    "            end = int(content[4])\n",
    "            if feature == \"three_prime_UTR\":\n",
    "                if forward:\n",
    "                    end = start+2\n",
    "                else:\n",
    "                    start = end-2\n",
    "                feature = \"stop_codon\"\n",
    "                \n",
    "            current.add(Feature(feature, start, end, forward))\n",
    "\n",
    "print(current.compile_genes())\n",
    "print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "annotations.append(current)\n",
    "print(c, \"=\", len(annotations), \"of which\", f, \"were faulty\")\n",
    "\n",
    "annotation_dict = {a.name:a for a in annotations}\n",
    "num_of_genes = np.sum([len(a.genes) for a in annotations])\n",
    "print(\"There are %s genes annotated\" % num_of_genes)\n",
    "\n",
    "\n",
    "# Get information that does not require sequences:\n",
    "intron_lengths = []\n",
    "exon_lengths = []\n",
    "intron_counts = []\n",
    "for ann in annotations:\n",
    "    for gene in ann.genes:\n",
    "        el, il = gene.get_exon_intron_stats()\n",
    "        exon_lengths += el\n",
    "        intron_lengths += il\n",
    "        if (len(il) > 0 and np.max(il) > 30):\n",
    "            print(ann.name, il)\n",
    "            # Can see that the two largest ones (58, 85) are from 615, \n",
    "            # and are not very trusted (coverage-signal is ambiguous)\n",
    "        intron_counts += [len(il)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input: Raw sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NODE_64_length_178741_cov_1.661034\n",
      "\t [ 4. 12.  1.]\n",
      "term: AGAGAATAGATGCTTTCCAAG : 0.0*UGA\n",
      "NODE_65_length_177181_cov_1.879014\n",
      "\t [9. 9. 1.]\n",
      "term: TCAGTTGTTCCGTATTTCCTT : 0.0*UGA\n",
      "intron:  GTAATTTCATTATGCAG\n",
      "NODE_69_length_175562_cov_1.893300\n",
      "\t [18. 17.  1.]\n",
      "term: CTGTATTCAAGAGCTAGATTA : 0.0*UGA\n",
      "NODE_90_length_157735_cov_1.661280\n",
      "\t [2. 8. 1.]\n",
      "term: CTCTATGAGCAGCTTAAGTGT : 0.0*UGA\n",
      "NODE_103_length_148054_cov_1.718544\n",
      "\t [ 6. 18.  1.]\n",
      "term: ATTATGTAGTGGAAAAGAGCT : 0.0*UGA\n",
      "intron:  GTAATATTTATTAATTTTAG\n",
      "NODE_127_length_132101_cov_1.744328\n",
      "\t [11. 18.  1.]\n",
      "term: ATGGTCAATAAGCCCAAACAA : 0.0*UGA\n",
      "intron:  GTAAATTTTAATTTTTAG\n",
      "\t [18. 28.  1.]\n",
      "term: TTTGGAAGAGGACTTATTGAT : 0.0*UGA\n",
      "NODE_134_length_130250_cov_1.655841\n",
      "\t [12. 16.  1.]\n",
      "term: AATCCAGAATACCTTTTGCTC : 0.0*UGA\n",
      "NODE_135_length_129955_cov_1.831608\n",
      "\t [ 0. 10.  1.]\n",
      "term: GCAAATTTAATCCCTCCTATT : 0.0*UGA\n",
      "intron:  GTAAGCTTTTATATAG\n",
      "NODE_150_length_122897_cov_1.750631\n",
      "\t [ 7. 15.  1.]\n",
      "term: GGAGCTGAAAATTTCCTAAGT : 0.0*UGA\n",
      "NODE_155_length_121415_cov_1.724095\n",
      "\t [6. 8. 1.]\n",
      "term: GTTGCAGCTCTTTCTGATATT : 0.0*UGA\n",
      "NODE_160_length_120340_cov_1.623793\n",
      "\t [6. 8. 1.]\n",
      "term: TTAAGAACAATAGAGAAAATA : 0.0*UGA\n",
      "NODE_176_length_115965_cov_1.774081\n",
      "\t [15. 29.  1.]\n",
      "term: AGATATTCTGAGTTGTTTTCA : 0.0*UGA\n",
      "intron:  GTAAAATTTTATGCAG\n",
      "intron:  GTAAGTATTAATATAG\n",
      "intron:  GTAAGCTTTAACTTAG\n",
      "NODE_177_length_115532_cov_1.858698\n",
      "\t [ 3. 12.  1.]\n",
      "term: AAGCAAATATTTGTGATGTCT : 0.0*UGA\n",
      "NODE_206_length_105986_cov_1.546198\n",
      "\t [12. 26.  1.]\n",
      "term: GCAGGCTTGAAAATGAAGGCT : 0.0*UGA\n",
      "NODE_210_length_104547_cov_1.706570\n",
      "\t [ 2. 30.  1.]\n",
      "term: GACACAGCTCTTTACAACTAG : 0.0*UGA\n",
      "intron:  GTAAACCATAATACAG\n",
      "NODE_217_length_103446_cov_1.848866\n",
      "\t [15. 10.  1.]\n",
      "term: CTAAGGCAATTTACAAGTATT : 0.0*UGA\n",
      "NODE_227_length_100883_cov_2.022033\n",
      "\t [14.  8.  1.]\n",
      "term: GCAAATTTAATCCCTCCCATT : 0.0*UGA\n",
      "intron:  GTAAGTACTAATATAG\n",
      "NODE_250_length_96914_cov_1.797266\n",
      "\t [3. 9. 1.]\n",
      "term: CAAGATTTGTCTGAAATTCTT : 0.0*UGA\n",
      "NODE_265_length_94667_cov_1.884356\n",
      "\t [1. 6. 1.]\n",
      "term: AAAAACGTCTTGTTCTATCTG : 0.0*UGA\n",
      "intron:  GTAAGGATTTTAATATATAG\n",
      "\t [3. 8. 1.]\n",
      "term: AAGCATCCCTACTTACTATCT : 0.0*UGA\n",
      "intron:  GTAAGATTTTTTATAAAAAG\n",
      "NODE_267_length_94345_cov_1.719173\n",
      "\t [10. 28.  1.]\n",
      "term: TTTCTAGATGTTGCTTTAGGG : 0.0*UGA\n",
      "intron:  GTAATCTTTACTATACAG\n",
      "NODE_273_length_93569_cov_1.730303\n",
      "\t [27. 76.  1.]\n",
      "term: CTAGACTAGGAGTATGATCTA : 0.0*UGA\n",
      "NODE_279_length_92227_cov_1.518132\n",
      "\t [6. 8. 1.]\n",
      "term: GAAAACCCATTCAGAGATTGG : 0.0*UGA\n",
      "NODE_301_length_87574_cov_1.953023\n",
      "\t [11. 19.  1.]\n",
      "term: GTAAGACAAAGCAATTTTATT : 0.0*UGA\n",
      "NODE_324_length_81694_cov_1.607292\n",
      "\t [ 5. 10.  1.]\n",
      "term: TCCGTTGTTCCATACTTCTTA : 0.0*UGA\n",
      "intron:  GTAATTATTAATATAG\n",
      "NODE_358_length_77800_cov_1.675254\n",
      "\t [17. 56.  1.]\n",
      "term: ATAGAAAACTATACCTACTGG : 0.0*UGA\n",
      "NODE_379_length_75420_cov_1.946927\n",
      "\t [4. 9. 1.]\n",
      "term: CTCTATGCCTCACTAAAATGC : 0.0*UGA\n",
      "NODE_389_length_74397_cov_1.930066\n",
      "\t [12. 18.  1.]\n",
      "term: ACAGTCAAAATTTGGGTATGC : 0.0*UGA\n",
      "NODE_426_length_70982_cov_1.681053\n",
      "\t [14. 13.  1.]\n",
      "term: TCAGAGGCGATACAAAAATAT : 0.0*UGA\n",
      "intron:  GTAATTTATTATACAG\n",
      "NODE_427_length_70843_cov_2.049663\n",
      "\t [3. 9. 1.]\n",
      "term: AGAATCTTCTTATATCGCCGT : 0.0*UGA\n",
      "NODE_450_length_68074_cov_1.636967\n",
      "\t [16. 35.  1.]\n",
      "term: AGAATAGCCACTTCTCCTACT : 0.0*UGA\n",
      "NODE_469_length_66432_cov_1.646482\n",
      "\t [10. 11.  1.]\n",
      "term: TTGTTATCTCTTATTAAGACT : 0.0*UGA\n",
      "intron:  GTAAGTTTTTATGTAG\n",
      "NODE_476_length_65419_cov_1.711113\n",
      "\t [12. 11.  1.]\n",
      "term: ATTTTCAAATAGCAGTTATCT : 0.0*UGA\n",
      "NODE_479_length_65262_cov_1.758118\n",
      "\t [15. 24.  1.]\n",
      "term: GCTAAATGGGACGAAGACTTT : 0.0*UGA\n",
      "intron:  GTAAGGCTATAACACAG\n",
      "intron:  GTAAGTTAATAACTTTAG\n",
      "NODE_485_length_64803_cov_1.657740\n",
      "\t [14. 42.  1.]\n",
      "term: TTTAGGGAGCTCACAGAAAAG : 0.0*UGA\n",
      "intron:  GTAATTTTAATAATTGTAG\n",
      "NODE_487_length_64315_cov_1.673226\n",
      "\t [13. 16.  1.]\n",
      "term: GCTGGATTAAAAATGAAGGCT : 0.0*UGA\n",
      "NODE_518_length_61243_cov_1.724180\n",
      "\t [12. 25.  1.]\n",
      "term: TATAGAGAGCTCACTGAAAAA : 0.0*UGA\n",
      "intron:  GTAATTTTTTAATTTGTAG\n",
      "intron:  GTAAGTATTTATGTAG\n",
      "intron:  GTAAGTTAATTACACAG\n",
      "NODE_547_length_58581_cov_1.583467\n",
      "\t [ 4. 16.  1.]\n",
      "term: GTTGATAGATTCAAACAACTA : 0.0*UGA\n",
      "NODE_555_length_57934_cov_1.702821\n",
      "\t [ 8. 13.  1.]\n",
      "term: AGAGAGCCCCTTAAGACTTTC : 0.0*UGA\n",
      "NODE_585_length_55535_cov_1.790265\n",
      "\t [ 7. 19.  1.]\n",
      "term: GCAAAAGTAGTTACTGTTGAT : 0.0*UGA\n",
      "NODE_632_length_52289_cov_1.832982\n",
      "\t [ 6. 15.  1.]\n",
      "term: GAAGAAATGCTTGGCCCAAAA : 0.0*UGA\n",
      "NODE_712_length_47909_cov_1.617806\n",
      "\t [7. 4. 1.]\n",
      "term: TAGGCAGTCATCAATAACTTG : 0.0*UGA\n",
      "intron:  GTAATTCTTAATTTAG\n",
      "NODE_721_length_47096_cov_1.882050\n",
      "\t [2. 6. 1.]\n",
      "term: ATGCTAAACACTGTCATCTTC : 0.0*UGA\n",
      "intron:  GTAATTTTTAATTGTAG\n",
      "NODE_725_length_46739_cov_1.861538\n",
      "\t [14. 21.  1.]\n",
      "term: TTTCTTGATGTTGCACTTGGT : 0.0*UGA\n",
      "intron:  GTAAGTTTTAACATTAG\n",
      "NODE_726_length_46702_cov_1.652088\n",
      "\t [19. 18.  1.]\n",
      "term: CCAAACTGGGACCAAGACTTT : 0.0*UGA\n",
      "intron:  GTAAGATCTCATTTCTTTACAG\n",
      "intron:  GTAAATTTTTTAACTTCAG\n",
      "NODE_762_length_45064_cov_1.971004\n",
      "\t [ 7. 19.  1.]\n",
      "term: ATAAAAACTTTAGAAGAAAAT : 0.0*UGA\n",
      "NODE_893_length_39578_cov_1.648298\n",
      "\t [12. 28.  1.]\n",
      "term: GCAAAGCAATCCTGCTGCCAC : 0.0*UGA\n",
      "NODE_896_length_39504_cov_1.683648\n",
      "\t [ 7. 30.  1.]\n",
      "term: ATGGTCAATAAACCAAAACAA : 0.0*UGA\n",
      "intron:  GTAATTTTTAATTATAG\n",
      "NODE_926_length_38670_cov_1.606881\n",
      "\t [ 6. 11.  1.]\n",
      "term: ATCATACACAAAAAATGCATT : 0.0*UGA\n",
      "NODE_1083_length_33926_cov_1.664753\n",
      "\t [12. 14.  1.]\n",
      "term: CCAAACAACGAAAGAACTTTG : 0.0*UGA\n",
      "intron:  GTAATTATTAATACAG\n",
      "NODE_1087_length_33881_cov_1.716982\n",
      "\t [ 78. 169.   1.]\n",
      "term: GAAGGAGGTTTTGAGGAACAA : 0.0*UGA\n",
      "intron:  GTAAGGTTAAATTTAACCATTTATAG\n",
      "intron:  GTAAGGGCCATTTATATATTTAG\n",
      "NODE_1178_length_31386_cov_1.635561\n",
      "\t [10. 16.  1.]\n",
      "term: GACGCTGCTCTATACAATTAG : 0.0*UGA\n",
      "intron:  GTAAGTTATAATTCAG\n",
      "NODE_1721_length_22983_cov_1.829424\n",
      "\t [ 6. 14.  1.]\n",
      "term: AAAGAGGTGGAAAAAATGCTT : 0.0*UGA\n",
      "NODE_1864_length_21573_cov_1.733703\n",
      "\t [13. 47.  1.]\n",
      "term: CCTCCTAAGGAAAGAGTTTAT : 0.0*UGA\n",
      "NODE_2222_length_18295_cov_4.113496\n",
      "\t [12. 14.  1.]\n",
      "term: GCAGTTAAGTTAATTTTCGCT : 0.0*UGA\n",
      "NODE_3449_length_12448_cov_1.887347\n",
      "\t [ 9. 17.  1.]\n",
      "term: CAGTCAGAGTTGGACGATATG : 0.0*UGA\n",
      "NODE_13726_length_2657_cov_6.379447\n",
      "\t [ 1. 32.  1.]\n",
      "term: ATCGAAAAAAATACAAGTTGG : 0.0*UGA\n"
     ]
    }
   ],
   "source": [
    "counts = np.zeros((num_of_genes,3))\n",
    "codon_usage_general = np.zeros(64)\n",
    "codon_usage_stop = np.zeros(64)\n",
    "codon_number = 0\n",
    "stop_codon_number = 0\n",
    "\n",
    "start_upstream = 3\n",
    "start_regions = []\n",
    "introns_by_length = {}\n",
    "ncs_base_counts = np.zeros(4)\n",
    "\n",
    "current_header = None\n",
    "current_sequence = \"\"\n",
    "gene_index = 0\n",
    "with open(sequence_filename) as scaffolds:\n",
    "    for row, line in enumerate(scaffolds):\n",
    "        if line.startswith(\">\"):\n",
    "            # Here query the analysis\n",
    "            if current_header is not None:\n",
    "                if current_header in annotation_dict:\n",
    "                    print(current_header)\n",
    "                    ann = annotation_dict[current_header]\n",
    "                    for gene in ann.genes:\n",
    "                        # Collect base-frequencies in flanking NCS\n",
    "                        \n",
    "                        start, end = gene.get_range()\n",
    "                        ncs_left = current_sequence[max(0, start -1 -hyperparameters[\"ncs_flank_size\"]): start]\n",
    "                        ncs_right = current_sequence[end: min(end+hyperparameters[\"ncs_flank_size\"], len(current_sequence))]\n",
    "                        ncs = ncs_left + ncs_right\n",
    "                        \n",
    "                        for i in range(4):\n",
    "                            ncs_base_counts[i] += ncs.count(base_order[i])\n",
    "                        \n",
    "                        # Collect number of inframe-stops, outframe-stops, and check for presence of stop as annotated\n",
    "                        counts[gene_index, :] = np.array(gene.count_UGA(current_sequence))\n",
    "                        print(\"\\t\", counts[gene_index,:])\n",
    "                        if counts[gene_index,2] == 0:\n",
    "                            print(\"\\t\\t\\t\\t!>!>!>missing STOP\")\n",
    "                            \n",
    "                        # Collect codon usage in CDS and stop regions\n",
    "                        cds_codon_counts, cds_num_of_codons, stop_region = gene.codon_usage(current_sequence)\n",
    "                        codon_number += cds_num_of_codons\n",
    "                        stop_codon_number += 7\n",
    "                        codon_usage_general += cds_codon_counts\n",
    "                        codon_usage_stop += stop_region\n",
    "                        \n",
    "                        gene_index += 1\n",
    "                        \n",
    "                        # Collect intron sequences:\n",
    "                        for f in [f for f in gene.features if f.feature == \"intron\"]:\n",
    "                            curr_intron = current_sequence[(f.start-1):f.end]\n",
    "                            curr_intron_length = f.end - f.start + 1\n",
    "                            if not f.strand:\n",
    "                                curr_intron = rev_comp(curr_intron)\n",
    "                            if curr_intron_length not in introns_by_length:\n",
    "                                introns_by_length[curr_intron_length] = [curr_intron]\n",
    "                            else:\n",
    "                                introns_by_length[curr_intron_length] += [curr_intron]\n",
    "                            \n",
    "                            print(\"intron: \", curr_intron)\n",
    "                            \n",
    "                        # Collect start-region:\n",
    "                        g_start, g_end = gene.get_range()\n",
    "                        if gene.strand:\n",
    "                            start_regions += [current_sequence[max(0, g_start-1-start_upstream):\n",
    "                                                               min(len(current_sequence), g_start+2)]]\n",
    "                        else:\n",
    "                            start_regions += [rev_comp(current_sequence[max(0, g_end-3):\n",
    "                                                                        min(len(current_sequence), g_end+start_upstream)])]\n",
    "                        if not start_regions[-1].endswith(\"ATG\"):\n",
    "                            print(\"! weird Start-region\", start_regions[-1], \"\\n\")\n",
    "            \n",
    "            current_header = line[1:-1]\n",
    "            current_sequence = \"\"\n",
    "        else:\n",
    "            current_sequence += line[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NCS-parameters\n",
    " - Estimate NCS-base-frequency from flanking NCS\n",
    " - How estimate NCS-length? Have user specify mean NCS-length estimate (allowed to be manual), thence geometric estimate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters[\"base_frequencies_NCS\"] = \"{%s}\" % \", \".join([str(f) for f in ncs_base_counts / np.sum(ncs_base_counts)])\n",
    "parameters[\"transition_probability_of_staying_in_NCS\"] = 1-1/hyperparameters[\"ncs_mean_length\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDS-parameters\n",
    " - Estimate probability to stay in CDS by assuming geometric **distr. of cds/exon length** (mean)\n",
    " - Estimate probability of going to stop by assuming geometric model of **number of introns** (mean)\n",
    " - Compute empirical **positional base frequency marginals**\n",
    " - For Stop-region: compute empirical **codon usage** outside stop-regions\n",
    " \n",
    "For the probability of staying in CDS (with mean length $\\mu$ in nucleotides, thus mean length in codons $\\frac{\\mu}{3}$): $p_{leave} = \\frac{1}{\\frac{\\mu}{3}} = \\frac{3}{\\mu}$ and thus $p_{stay} = 1-\\frac{3}{\\mu}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There should not be any exon-length not divisible by three, and there are 0 which are not divisible by 3\n"
     ]
    }
   ],
   "source": [
    "print(\"There should not be any exon-length not divisible by three, and there are\", \n",
    "      np.sum([l % 3 for l in exon_lengths]), \"which are not divisible by 3\")\n",
    "\n",
    "parameters[\"transition_probability_of_staying_in_CDS\"] = 1 - 3 / np.mean(exon_lengths)\n",
    "parameters[\"transition_probability_of_CDS_to_stop_given_that_CDS_is_being_left\"] = 1/(1+np.mean(intron_counts))\n",
    "\n",
    "cds_positional_marginals = positional_base_usage(codon_usage_general / codon_number)\n",
    "\n",
    "parameters[\"base_frequency_marginals_CDS\"] = format_matrix(cds_positional_marginals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start region parameters\n",
    "These are difficult to infer automatically (especially the length of the pattern); can infer the base-probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sanity check: the last 3 rows should show deterministic ATG\n",
      "[[0.10344828 0.         0.89655172 0.        ]\n",
      " [0.12068966 0.         0.79310345 0.0862069 ]\n",
      " [0.18965517 0.         0.74137931 0.06896552]\n",
      " [0.         0.         1.         0.        ]\n",
      " [1.         0.         0.         0.        ]\n",
      " [0.         0.         0.         1.        ]]\n"
     ]
    }
   ],
   "source": [
    "start_region_bases = np.zeros((start_upstream + 3, 4))\n",
    "# print(\"\\n\".join(start_regions))\n",
    "for r in [r for r in start_regions if len(r) == start_upstream + 3]:\n",
    "    # print(r)\n",
    "    for i in range(start_upstream + 3):\n",
    "        start_region_bases[i, base_order.index(r[i])] += 1\n",
    "        \n",
    "actual_start_region = start_region_bases / num_of_genes\n",
    "print(\"Sanity check: the last 3 rows should show deterministic ATG\")\n",
    "print(actual_start_region)\n",
    "actual_start_region = np.mean(actual_start_region[:3,:], axis=0)\n",
    "parameters[\"base_frequencies_start_region_upstream\"] = \"{%s}\" % \", \".join([str(p) for p in actual_start_region])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stop region parameters\n",
    "Size is fixed; like for start-region, this has to be done manually (_could_ try some sort of automation, but likely not very good, and not worth the effort)\n",
    " - Compute empirical **positional base frequency marginals**\n",
    " - Compute empirical **codon usage** in and outside stop region\n",
    " - Force UGA to 0\n",
    " - Force UAA to empirical (low) frequency\n",
    " - Find the codon which maximises the difference between codon-wise $stopregion - cds$ and base-wise $stopregion - cds$ (cf `StopCodons.ipynb`) and correct its probability\n",
    " - Find a compensatory codon, and correct its probability (4 codons explicit)\n",
    " \n",
    "The latter part can be extended: Could also pick the top 3 most devious codons, and then 2 compensatory ones etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0.027093596059113302, 0.027093596059113302, 0.03201970443349754, 0.022167487684729065, 0.019704433497536946, 0.007389162561576354, 0.014778325123152709, 0.0, 0.029556650246305417, 0.014778325123152709, 0.0, 0.017241379310344827, 0.0024630541871921183, 0.012315270935960592, 0.0, 0.017241379310344827, 0.034482758620689655, 0.012315270935960592, 0.024630541871921183, 0.0049261083743842365, 0.014778325123152709, 0.009852216748768473, 0.017241379310344827, 0.0024630541871921183, 0.0024630541871921183, 0.0049261083743842365, 0.027093596059113302, 0.007389162561576354, 0.0024630541871921183, 0.0024630541871921183, 0.0, 0.0, 0.03201970443349754, 0.017241379310344827, 0.019704433497536946, 0.027093596059113302, 0.022167487684729065, 0.0024630541871921183, 0.014778325123152709, 0.0, 0.027093596059113302, 0.022167487684729065, 0.05665024630541872, 0.03201970443349754, 0.007389162561576354, 0.0024630541871921183, 0.03940886699507389, 0.0049261083743842365, 0.029556650246305417, 0.014778325123152709, 0.007389162561576354, 0.0049261083743842365, 0.03201970443349754, 0.0049261083743842365, 0.022167487684729065, 0.0024630541871921183, 0.024630541871921183, 0.019704433497536946, 0.046798029556650245, 0.027093596059113302, 0.0049261083743842365, 0.0049261083743842365, 0.012315270935960592, 0.0024630541871921183}\n"
     ]
    }
   ],
   "source": [
    "empirical_codon_frequencies_stop = codon_usage_stop / stop_codon_number\n",
    "empirical_codon_frequencies_general = codon_usage_general / codon_number\n",
    "\n",
    "print(\"{%s}\" % \", \".join([str(i) for i in empirical_codon_frequencies_stop]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support. ' +\n",
       "              'Please try Chrome, Safari or Firefox  6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            if (mpl.ratio != 1) {\n",
       "                fig.send_message(\"set_dpi_ratio\", {'dpi_ratio': mpl.ratio});\n",
       "            }\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        fig.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var backingStore = this.context.backingStorePixelRatio ||\n",
       "\tthis.context.webkitBackingStorePixelRatio ||\n",
       "\tthis.context.mozBackingStorePixelRatio ||\n",
       "\tthis.context.msBackingStorePixelRatio ||\n",
       "\tthis.context.oBackingStorePixelRatio ||\n",
       "\tthis.context.backingStorePixelRatio || 1;\n",
       "\n",
       "    mpl.ratio = (window.devicePixelRatio || 1) / backingStore;\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width * mpl.ratio);\n",
       "        canvas.attr('height', height * mpl.ratio);\n",
       "        canvas.attr('style', 'width: ' + width + 'px; height: ' + height + 'px;');\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option);\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'] / mpl.ratio;\n",
       "    var y0 = (fig.canvas.height - msg['y0']) / mpl.ratio;\n",
       "    var x1 = msg['x1'] / mpl.ratio;\n",
       "    var y1 = (fig.canvas.height - msg['y1']) / mpl.ratio;\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width / mpl.ratio, fig.canvas.height / mpl.ratio);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x * mpl.ratio;\n",
       "    var y = canvas_pos.y * mpl.ratio;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overridden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    var width = fig.canvas.width/mpl.ratio\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\" width=\"' + width + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var width = this.canvas.width/mpl.ratio\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\" width=\"' + width + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        // select the cell after this one\n",
       "        var index = IPython.notebook.find_cell_index(this.cell_info[0]);\n",
       "        IPython.notebook.select(index + 1);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAAAHgCAYAAAA10dzkAAAgAElEQVR4nOzde3gU9aH/8dlsdjc3knC/5obNZREvRA20ykWwhbQq4q0qReQ5WhJ/moBVTLUttTUpoRWlyqEaU/W0eI1RvLQ9rdbAIUCgraXUxVpFMBXUFk2wyEXg8/sDdprZ3ZnZ3fnOzkzm83qeeR7dbyYzs8t+972T7EQCEREREbmKZPUOEBEREVFqMQCJiIiIXIYBSEREROQyDEAiIiIil2EAEhEREbkMA5CIiIjIZRiARERERC7DACQiIiJyGQYgERERkcswAImIiIhchgFIRERE5DIMQCIiIiKXYQASERERuQwD8KRjx46hu7sbPT096O3t5cKFCxcuXLhwse3S09OD7u5uHDt2LKnuYQCe1N3dDUmSuHDhwoULFy5cHLN0d3cn1T0MwJN6enrkO9LqqufChQsXLly4cNFawieuenp6kuoeBuBJvb29kCQJvb29Vu8KERERkSaj3cIAPIkBSERERE7BABSEAUhkf0W3v4Si21+yejeIiCzHABSEAUhkfwxAIqITGICCMACJ7I8BSER0AgNQEAYgpUI4YBgxyeF9R0R0AgNQEAYgpQID0Bjed0QUixvnBgagIAxASgUGoDG874goFjfODQxAQRiAlAoMQGN435FV+Ny1Nzc+NgxAQewUgG78h+wWfBExhvcdWYXPXXMZvW/d+NgwAAVhAJIIei8SfBExhvcdWYXPXXMxABPHABSEAUgiMADNxfuOrGL2c9ftcwMDMHEMQEEYgCSC2QHIFwn3HjtZiwFoLgZg4hiAgjAASQSjAcgziNpcc+xLc/+zkC0wAM3FAEwcA1AQBiCJ4PQAtPu/PbvvnzAMwIQ5PdAYgAzARDEABWEAkggMQHPZff+Mko+PAZgwpwcaA5ABmCgGoCC2DMAkXwDs/ETo75McA9Bcdt8/oxiAJyVx7E4PtP4+N+phACaOASgIAzA1+vskxwA0l933zyi3B6CRuc/pgdbf50Y9DMDEMQAFYQCmRkKB48AXQKGBF+P4GYD23j+jhAWgA587AAPQyXOfUQzAxDEABWEApgYDkAEYF5UAss3+mYQByAB06txnFAMwcQxAQfplANpwEmEAMgDjwgBkACb4b58B6GwMwMQxAAVhAKaG3QPQ6CRvqwA0cBbFcgxABiAD0JTt2JXRf/v9fW6IhQEoCAMwNRiAAgNQ7/5x8iRqNAAd+iEKYS+CRo/dovtOWACasP8MQHMxABPHABSEARhjfROkdBK34PeIzA48BiADMKH1k8UA1P7+JrB67rMaAzBxDEBBGIDi1o/nezMAbR6AVv8IlgHIAOxnAWj3uc9qDMDEMQAFcVMA6j1RGIAMQAagNRiADEBbPH4WPP4MwMQ5NgBXrVqF4uJiBAIBVFZWYv369Zpf39HRgcrKSgQCAZSUlGD16tVRX9PW1oZgMAi/349gMIj29va494cBKG77MqsncQagdQFoNMD0tp/kuN31mwA0Ye5iACYgiftPb32zMQAT58gAfPLJJ+Hz+dDS0oJQKIT6+npkZ2dj9+7dMb9+586dyMrKQn19PUKhEFpaWuDz+dDW1iZ/zcaNG+H1etHU1IQdO3agqakJ6enp2Lx5c1z7xAAUt32Z1ZM4AzDlASgswCwKQKtfRBiADEAGoKB/+w5785cMRwZgVVUVampqFLdVVFSgoaEh5tcvWbIEFRUVitsWLlyISZMmyf9/5ZVXYtasWYqvmTlzJq666qq49okBKG77MqsncaMBaEJAmh2Aovbf9ABLcjxlAWjRiwgDkAFo1wCMe32LnntWP3et4LgAPHz4MLxeb9SPZ+vq6jBlypSY60yePBl1dXWK29rb25Geno4jR44AAAoKCrBixQrF16xYsQKFhYVx7ZeTAlDYeJLbj5sFk7it1k/i+6c0AE0ILAag9v6bvX0GoE0CSm//9Mat3n8D979t/u0zAHWlPADff/99SJKEzs5Oxe2NjY0oKyuLuU5paSkaGxsVt3V2dkKSJOzZswcA4PP5sGbNGsXXrFmzBn6/P+b3PHToEHp7e+Wlu7ubARjn+nFzewBaEHgMQJu8iNjlRTBZDEBj2zc6zgB03HPXCo4NwI0bNypuv/vuu1FeXh5zndLSUjQ1NSlu27BhAyRJwt69ewGcCMDHH39c8TW//OUvEQgEYn7PpUuXQpKkqCUVAWjkXaRQFkzSRrdv98CzO6NnMURt3ypWv8hYHXhmvzm08s2lnoTePJmwfaPr95f9t81zx6z1UxiQjgtAu/wI2MozgFa/CMoYgAxAi7ZvFce8iMS7foIYgM4PKKvuP6Pr2+65Y9UZRoEcF4DAiQ+B1NbWKm4LBoOaHwIJBoOK22pqaqI+BFJdXa34mlmzZtnyQyBWvwjKGIAMQIu271RG958ByABMdn0GoLPXN4MjAzB8GZjW1laEQiEsWrQI2dnZ2LVrFwCgoaEB8+bNk78+fBmYxYsXIxQKobW1NeoyMJ2dnfB6vVi2bBl27NiBZcuW2fYyMLZ5ETT6IqIxxgC0JwagMbYJQFHbF7w+AzBOVs89DEAGICy+EHRRURH8fj8qKyuxbt06eWz+/PmYOnWq4us7OjowYcIE+P1+FBcXx7wQ9DPPPIPy8nL4fD5UVFTg2WefjXt/7PQpYCdjANobA9AYBiADMNntG12fAejs9c3g2AC0GwagBRwYcAxAMdt3K2EBZ3R9BiADkAGY0vXNwAAUhAFoAacHnE0mgUQwAK1ldQDKTApAs9eXMQAZgA5b3wwMQEEYgBZgAKYcA9Ba/SYAk8QATH59W81dDgwwq9c3AwNQEAagBYxOgklw+o9wjWIAWss2AWgRBmDy6zMAnb2+GRiAgjAAnUFoAFLKuf2+ZwAyAJNdX+jcZUHAWB1gVq9vBgagIAxAZ2AAOpvb73sePwMwWU6fu6wOMKvXNwMDUBAGoDM4fRJ0O7c/djx+BmCynD73WR1gVq9vBgagIAxAZ3D6JOh2bn/sePwMwGQ5fe6zOsCsXt8MDEBBGIDO4PRJ0O3c/tjx+BmAyXL63Gd1gFm9vhkYgIIwAJ3B6ZOg27n9sePx2yQAHai/7L/lAZckO973DEBBGIDO4PRJ0O3c/tjx+BmAyeov+88AFIcBKAgD0BmcPgm6HR87dxMWgAa27dR/f/1l/w3/CJUBKGMACsIAdAanT4Jux8fO3RiAyesv+88AFIcBKAgD0BmcPgm6HR87d+Pjnzynz30MQPEYgIIwAJ3B6ZOg2/Gxczc+/slz+tzHABSPASgIA9AZnD4Juh0fO3fj4588p899DEDxGICCMACdwemToNvxsXM3Pv7Jc/rcJywAjW7fovXNwAAUhAHoDE6fBN2Ojx1Rcpw+91m97wzAaAzAkxiAzuD0SdDt+NgRJcfpc5/V+84AjMYAPIkB6AxOnwTdjo8dUXKcPvc5ed8Be+4/A1AQBqAzOH0SdDs+dkTJcfrc5+R9B+y5/wxAQRiAzuD0SdDt+NgRJcfpc5+T9x2w5/4zAAVhADqD0ydBt+NjR5Qcp899Tt53wJ77zwAUhAHoDE6fBImIkuH0uc/J+w7Yc/8ZgIIwAJ3B6ZMgEVEynD73OXnfAXvuPwNQEAagMzh9EiQiSobT5z4n7ztgz/1nAArCAHQGp0+CRETJcPrc5+R9B+y5/wxAQRiAzuD0SZCIKBmc+6xlx/ueASgIA9AZOAkSEVGq2fF1hwEoCAPQGRiARESUanZ83WEACsIAdAYGIBERpZodX3cYgIIwAJ2BAUhERKlmx9cdBqAgDEBnYAASEVGq2fF1hwEoCAPQGRiARESUanZ83WEACsIAdAYGIBERpZodX3ccF4Aff/wxvvGNbyA3Nxe5ubn4xje+gU8++URznePHj2Pp0qUYOXIkMjIyMHXqVPz1r3+Vx/ft24ebbroJZWVlyMzMREFBAW6++Wb09PTEvV8MQGdgABIRUarZ8XXHcQE4a9YsjB8/Hhs3bsTGjRsxfvx4XHjhhZrrLFu2DAMGDMCzzz6L7du34+tf/zpGjhyJ/fv3AwC2b9+OSy+9FC+88ALefvttvPrqqygtLcVll10W934xAJ2BAUhERKlmx9cdRwVgKBSCJEnYvHmzfNumTZsgSRLefPPNmOscP34cI0aMwLJly+TbDh06hLy8PPzsZz9T3dbTTz8Nv9+Pzz//PK59YwA6AwOQiIhSzY6vO44KwNbWVuTl5UXdnpeXh5///Ocx13nnnXcgSRL+9Kc/KW6/+OKLce2116puq6WlBUOGDIl73xiAzsAAJCKiVLPj646jArCxsRGlpaVRt5eWlqKpqSnmOp2dnZAkCe+//77i9htuuAFf+cpXYq7zr3/9C4WFhbjzzjtV9+XQoUPo7e2Vl+7ubgagAzAAiYgo1ez4umOLAFy6dCkkSdJctm7disbGRpSVlUWt/4UvfAE/+tGPYn7vcADu2bNHcfv111+PmTNnRn19b28vJk6ciFmzZuHIkSMJ7zMD0N4YgERElGp2fN2xRQD+85//xI4dOzSXgwcPmv4j4P379+OLX/wiZsyYgYMHD2ruM88AOhMDkIiIyCYBGK/wh0C6urrk2zZv3hzXh0Cam5vl2w4fPhz1IZDe3l5MmjQJU6dOxYEDBxLeN/4OoDMwAImIiBwWgMCJy8Ccfvrp2LRpEzZt2oTTTjst6jIw5eXlaG9vl/9/2bJlyMvLQ3t7O7Zv346rr75acRmY/fv3Y+LEiTjttNPw9ttvY+/evfJy9OjRuPaLAegMDEAiIiIHBuC+ffswd+5cDBgwAAMGDMDcuXOjLgQtSRIeeeQR+f/DF4IeMWIEAoEApkyZgu3bt8vjr732murvHr777rtx7RcD0BkYgERERA4MQLtiADoDA5CIiIgBKAwD0BkYgERERAxAYRiAzsAAJCIiYgAKwwB0BgYgERERA1AYBqAzMACJiIgYgMIwAJ2BAUhERMQAFIYB6AwMQCIiIgagMAxAZ2AAEhERMQCFYQA6AwOQiIiIASgMA9AZGIBEREQMQGEYgM7AACQiImIACsMAJCIiIqdgAArCACQiIiKnYAAKwgAkIiIip2AACtLT0wNJktDd3Y3e3l4uXLhw4cKFCxfbLt3d3ZAkCT09PUl1DwPwpPAdyYULFy5cuHDh4pSlu7s7qe5hAJ507NgxdHd3o6enJ6XlrnbGkeMct2rczvvGcY5z3L7jdt43O4yLXnp6etDd3Y1jx44l1T0MQIv09mr/7J7jHLdq3M77xnGOc9y+43beNzuM2w0D0CJW/0PkOMc5iXOc4xwXOW7nfbPDuN0wAC1i9T9EjnOckzjHOc5xkeN23jc7jNsNA9Aihw4dwtKlS3Ho0CGOc9xW43beN45znOP2Hbfzvtlh3G4YgEREREQuwwAkIiIichkGIBEREZHLMACJiIiIXIYBaLLHHnvM9F8I/eijj1THdu/ejePHj5u6fS133XUXDhw44NrtExER2RED0GRpaWn48MMPNb+mt7c35pW8jx49qvpx8uPHj+Pll1/GxRdfjPT0dPn2hoYGLF68WF48Hg92796tuu2PP/4YP/3pT2Nup6enR3UMOPHXU1544QXMnj1bvm3//v3o7f3Plcr1jv+tt97CVVddpbr9q6++Gu+8847q+n/5y19QU1Mj///LL7+MtWvXyove9j/77DOsXbsW+/fvjxrr7e3F2rVrNQP+L3/5C+rr61XHrWb1GwCrLViwIOZjmwpWv/lLxbG//vrrpn5/PZ9//rnqWCr+7Wsd/7p16zT3TwSrj1/L+eefj08++cSy7Vv53HcKBqDJPB6PZoC0t7ejtLQ05lmqAwcOoKysDC+88IJ82zvvvIM777wTY8aMQX5+PqqqqnD22WfL4zk5OZg4cSKmTZuGadOmQZIk3HXXXarb/8EPfoDLL79cdfyKK67A3XffrbjtrbfeQkNDA0aOHAm/34/hw4crtp+WliYvkiThN7/5jer3v+GGG3Dbbbepji9ZskQReMCJMPvZz36Gc845B5IkISMjQ7F9j8cjL5Ik4eGHH1b9/vfddx+mT5+uOj5jxgw88MADmtsfOHCgPFZQUICBAwfKi8/nw5YtW1S/v5EAB4B//etfuPfee1XH9QLY7DcAU6ZMMe1F4OOPP8ZPfvITlJSUyLfZ6Q1QWloadu7cqXkMVr75S1ZPTw9WrVqFM844Ax6PR3X7eXl5+Mc//qH5vZI5fgB44403sHjxYgwbNkz1a/T+7b/66qsIBoOqj/24ceOwfv36mGOrVq3ChAkTkJaWlvLtAyeOf9GiRRg0aJB82+rVq7Fy5Up58Xg82Lt3r+r2zX7u6732GTn+8L6fccYZqt/f6rnv008/Vd22XTAATebxeDTfpX/5y19GS0uL6nhraytmzJiBX/ziF5g6dSoCgQAuvPBCeL1ebN++HZMnT0Z7e7v89Tk5OYozZh6PB2eddZbq9z/jjDPwyiuvqI6/8sorOPPMM/HZZ5/h0UcfxeTJk+Hz+ZCWloaVK1di3rx5aGpqUmx/zZo16OjowGuvvQZJkjQDs7y8XDOQ/vCHP6CsrAwA0NHRgXnz5iErKwtpaWm4/fbbMX36dEXgxTp+rcA755xzFIEd6cUXX8Q555yjuv358+fj29/+tmL7y5cvx6OPPopHH30UkiTh2muvVf3+yQT48ePH8Zvf/AZXXHEFfD4fAoGAPBYZoJIkobOzU+j2Rb4BSOZF4He/+x2uuuoqZGRkID8/H8XFxYrt2+UNkCRJSEtLU/33bfWbv0Tv+1dffRVz585FZmYmKioqUF1djcmTJye9/USP/9NPP0VLSwsmTZoEr9eLiRMn4rLLLpPHr7nmGsyZM0deJEnCW2+9pbr9iy66CCtWrFAdX7lyJS655BLV41+yZAnuu+8+eTwygCVJ0gzwRLcfefxlZWUYO3asPJ6Tk4MxY8aguLgYxcXFkCRJ8/tb/dxP9PgB5XN/9OjRuOiii+SxRO9/s4/f6/VqvrbZAQPQZB6PB1/96lcVE1PfJSMjA1/+8pdV17/mmmvg8XgwadIkPPDAA/jXv/4FAEhPT8cbb7yB4cOH469//av89UOGDMG7776r2H4gEFC8M+y7BAIBLF26VHX7a9euhc/nQ25uLs4++2zcd999+OCDD+Ttl5eXK14kYgWYz+fDhAkTYi4ejwennnqq6va3bNmC9PR0nHLKKRgxYgQWL16MrVu3ytsvKirC1q1bNbfv9XpVt+/1enW3n5GRobr9U089Fb///e81t983UCLFG+AA8O677+K73/0uCgoKkJaWhnnz5mHOnDm4/fbbFdtPJECtfgMQ74vA7t278f3vfx9FRUUYPHgw0tLS0NbWZus3QB6PB1dccQW+8Y1vxPzeVr/5i+e+nzlzJn74wx+ipKQEw4YNw0033ST/2ze6/XiO/ytf+Qr+7//+D/Pnz0dOTg5OO+00eL1ebNiwAcuXL8fcuXMV27/ssstw3XXX4brrroMkSbj11ltVv39hYSFCoZDq+I4dOzBy5EjV4//Zz36GCy+8ULH9yAD+wQ9+YGj7BQUFqsd/wQUX4PHHH1dsP/L+/9KXvqT6/VPx3J89e7birFjfpaCgQDOQwsev9tw3ev+bffzz5s1Tfe7bBQPQZB6PB1//+tflSSly8Xq9Ue9y+vJ6vfB6vVG/yxCehDIyMvDmm29qbl+SJPldYeTi8XgwcuRIze37/f6obYS3n5WVhV27dsm3r1ixQnFGIRxg3//+92Mu2dnZmoHi8/mQkZGB3/zmN4ofFYW3HwgEFD9m27p1K44cOaLYflpamur2/X4/vvnNb6pu3+/3Iz09XXX7OTk5iuBetGiRHOnh7fv9ftVJMCcnRxHwkd566y1kZGRg+vTpyMjIwJw5c/DMM88kFKCDBg2y9A3AyJEjk34RWLFiBQKBALKysnD55Zfj+eefx+HDhxN6AxT+MX2sRZIk5OXlmXL84RfgzMxMy978aR17WloacnNzVbc/efJkeDweXH311XjppZdw9OjRhLeflZWFbdu2xVyGDh2KF198UXX7t912G7xeL0aPHo1bb70Vf/7znxXbr6qqwssvvxzzvg9vf+TIkar/9tPT0/Gd73xHdftTp06FJEmqxx9PAA8bNkxxVqrv4vV6sWDBAs3j93g8qsc/evRo+Ta17WdkZFj63A+fBY+1SJKk+BWCSCtXrkRaWprqcz+e+9/r9Vry3AeAzZs3o7CwUPX72wED0GR6vwdRUVGBX/ziF6rjNTU1yMrKQnZ2Nq688kq8+OKL+Pzzz+V/hF/4whfQ1tamuf2ioiLV8WnTpinOIEUqLi6G1+vFNddcg1//+tfyLxWHtz9w4EBs2LBBc/taT7IrrrhCM4Czs7ORlZWFO+64Azt27JBvD29/5MiR+N3vfqe5/aFDh6qOT5w4EcuWLVMdHzJkCAKBgOr2c3Nz0dXVpbl9I5NgXl4evF4vHnzwQXz88cdR248nQK1+A2Dk+EW8AfJ6vfIZ0cglMzNTMwKMHL/H48FFF12EtLQ0y978aR17eno6mpubdbcf+WPURN989v2d3PCSyGMfDq/I7Q8ePBh/+9vf5NvPOussdHd3K7YfPgMfa0lPT9f9HcLc3FzV448ngL1er3xGKnLJyMjA+PHjNbefl5enevyBQABvv/22fPtHH32keJNqh+d+eno6Ojo6Yi6jRo3CD3/4Q93jV/v3H+9Pv6x47gMnPoTT99dz7IgBaDK9ALzjjjtQWFiIDz74IGps7969KCwsxB133IF3330X3/ve91BYWIghQ4YgLS0NzzzzDOrq6jBu3DgcPHgwav3PPvsMkiTh+uuvV91+W1sb0tPTcf/99ysmmqNHj+KnP/0pfD4fVq9ejbvuugvFxcUYPnw46urqkJ6ejlAohOnTp2v+mMXj8eC8885THf/Tn/6EQCCAyy67DF1dXejp6UFPTw82b96MSy+9FIFAAK2trViwYAFycnJQWVmJFStWyNv/+te/rvg9kEiSJOHiiy9WHX/wwQeRnZ0d80zECy+8gOzsbNx2222q2//iF7+IxsZGzeMvKytTnQTPPPNMXH311arrBwIB5OXl4aGHHlJMLokEaHZ2tup4Kt4AZGdnJ/0icMEFFyAtLQ1f+tKXsHr1ajmCnfAGyOPx4MUXX1R8SKgvq9/8jR07VnEGJVJTUxNycnKQm5uLqqoq3H///fjoo48S2v7o0aOxa9eumMspp5yi+QGmyy+/HD6fDwUFBViyZAm2b98O4D/3fWZmpnyb2vb7fkAs0k033YTx48erzp1jx47FuHHjVI8/ngD2+/1Jb3/YsGHIz89XPf7CwkLFGdBY2x89erTquNVv/vWOf+DAgfD7/arPfaP3v9nHv2HDBtXnvl0wAE2mF4D79+/HqaeeigEDBqC2thb33XcfVq5ciZqaGgwYMADjxo1TvAM6fvw4fv3rX+OKK65AIBDAiBEjkJWVhcLCQixfvhzPP/881q5di+bmZhQUFECSJM1JEjgRoR6PB7m5uTjzzDMxYcIE5Obmyh906Ou3v/2t/Eu4paWlmDNnDrxeLx544AHFu89wQGp9Cjf8SaqJEydi6NChUWeHhg4dirVr18pf/+mnn+Khhx7CpEmT4PF4MG3aNHznO99BIBDA5Zdfji1btsgB2dXVhUsvvRSSJGmeIQSAuXPnwuPxIBgM4pJLLsGcOXNQUVGBtLQ0XHXVVZrbD38o5KWXXor6vi+88AIkScJPfvIT1W3rBXh6ejrq6upw/vnnIzMzE5deeina29vh8/nwxhtvxBWgfT8okOj2jb4BkCRJ8w2A3ovA+PHjUVtbi0cffRRTpkxBIBDAxRdfLP8enJ3fAHk8HtTU1Kh+CMnqN3/x3Pc333wzDhw4gNbWVpx77rny70Ddd999qKmpMbT9eI+/o6MD1157LbKzs3H66afLvwN36qmn4rHHHlP9/uE3X2o++OADjBo1CgUFBWhubpbnzmXLlqGgoACjRo3CBx98oHr8Y8eONRTg8W5f7fgXLFig+jt+x48fhyRJivkrktXP/XiO/91331V97ht9A2T28d9yyy2aH0C0AwagyTwej+LHErH09PSgtrYWgwYNkn9sMWjQINTW1mp+lH3fvn249957UVFRgZkzZyItLU3xI5aZM2eivr4+rgshd3V1oa6uDl/96ldRXV2N+vp6zTNL4Y/Qn3nmmfKPcmIF5A033BB1mYe+n6TKyMjA7Nmz8dlnn6G9vR3Lly9Hc3MznnvuOc39DoVC+Na3voVhw4YhLS1NfmHsuwwePBgPPvig4lpYf/3rXxW/hxT+EcJTTz2F2bNnY9y4cQgGg5g9ezaeeuqpuLYfvs/7BmQwGJR/zKV3Hch4A/ztt9+WPwXq8XhwzTXXYNGiRYYCNJHtA2LfAADxvwiGhf/tjBo1Crm5ubjkkkuQn59vyzdA4R+BPfPMMzG3a/Wbv0TvewB48803cdttt2HEiBEIBAIIBAJJb3///v2oqKiI+/j379+P1atXo6qqSv7dwPz8/JiXOtmzZw8kScKiRYtUtw8Au3btQnV1ddTcWV1drfhxYqzj93q9yMnJSTqAE91+5PFPmDABGRkZqKqqwtNPP40///nP2LZtG5566in5ElWbN2/W3L7Zz/2f//znMbcbvozP4MGD4z7+yOd+WVkZxowZY+j+N/P4fT6f6nPfLhiAJtM7A9g38I4fP46PPvoIH374oRwteuv3tW/fPnR1daGrqwv79u0DYPxCy/H44x//iE2bNqGurg7V1dWorq5GXV0dNm3aJH+N2ieptK6VFOtaS5E+//xzPPvsszhw4ADa29vR3NyM5uZmtLe349///jfWr18fdamMyMlG7wyhlvD2n3jiCcyePRvBYBDBYBAXX3wxnnjiCRQXF2teCy18sWC1AI91MeFjx47hV7/6FS677DL4/X74/X7VAB03bpylbwByc3MVv5PYVw1CC5AAACAASURBVDIvAn3vg/C/DZ/Pp/oG6J133lH824+8UHg4nM04fkmSsHDhQs373cw3f3/729+wbds2fPbZZ1HrHzhwANu2bcPOnTsTvu+BEy9yzz33HKZPn666/YkTJ8a8BmT4OoZz5syBz+dTPX6t60f+5S9/QW1tLbxeLwYMGIAbb7xRDsja2loMGDAAQ4YMiXl2MZaPP/4YW7ZsQVdXl+J3bbWO/5FHHtEMYK/Xq3mh6JKSEvm5EWv7fcdjHX99fT3y8/MRDAbl+z38OASDQVx//fWWPvcjz45FXsbm3HPPlT+Frnb/x7rQdfi5P3PmTEiSpHr/jxw5En//+9/l9VL53I91/HbEADSZXsAZHde72OU3v/lNzQstX3/99fjCF76gGogXXnghLrzwwqQDsqurCzfccIPqJ6liiXWGMBa9QDx27BgmT56s+EXrnJwcrFu3Drt27cK7776LxYsX49JLL5XH1c4QxjseSe/xMTr+0Ucf4Z577lENULu8AQgT9SLQV/j+ifUG6MUXX1RcLDbyQuHhH6eacfxWv/l75JFHcNZZZ0V9iAA4ETBnnXWW/DuIse77o0ePqgbkZ599hm3btsnfO9b2I0Vex3Du3Lny7yDGOn5A+y+dACfmioULF2LgwIHyYzpw4EAsXLgQb7zxBu644w75a88991zFJaDOPvts3QtV69m5c6dqAOs9d43O/QDkKx68/vrreOqpp/DUU0/hT3/6EwD7vPlXu4yNlngu9A2cuEyX2v3f0tIiX0ILSO1zP9bcZ0cMQJN5PB78z//8j+KdR9/F4/Ho/h6LkfXLyso0L7MxZ84czV9UPf300zWvtq4XkB6PBwsWLFD9JFVYImcI9QKx77jH44m6WG3fSe/hhx+Gz+dTjPedTCJ/j0jvDGJkIJr9BkDvz41Z/QYgzIwXgTfeeAO33HKL5nhJSQkGDBgg3xb5+Dc3N+OCCy6Q/z/WWQKtM4h6v4Rv5Zu/c889F0888YTq+BNPPIHKykrVM4Q/+MEPNAOysrISjY2NmmcYDxw4oHodQzV9zxDG+iX+8Pill14qjx8/fhwffvihIiC/+93vKi6zkpOTg7q6OvkSUBMmTMDEiRMV91ffQDz99NNx8803q46fc845ckDGCmCzH3+tM4SA9c/95uZmlJeXq17GJpLWm8NIkc/9WPf/hRdeiAcffFBeJ5XPfadgAJos1iUQ+i7hU8hagRfP+moyMjIUH1WPNHbsWM2PqhcVFWley0gvIIuLi+Hz+VQ/SRXvGUK9QFQbjzz+8I+Lwy666CJ4vV75/yPPEJaXl+P0009XHb/iiiuQn5+vGO8biJIkab7L9Hi0/1KM3rheBFj9BuCMM87A2LFjhb0IJDoeCASwePFieTzyReD+++9X/Dm1yLMEHo/yQt6xziL0fQPW90XC4/HglltusezN39ChQzV/jLt8+XKkp6erBl5OTg5uvPFG1fVra2uRk5Ojuv7QoUORnZ2teh3DSFpnCOMZDwsHYl5eXtRj2/exv/rqqxX/9iMDcdSoUVFnj/uOT5w4Ed/61rdU7x+Px4PXXntN9TqIHo8HDz/8sKHxvtcAjWT1c9/j8aC2tlb1MjZh8b45TDQQc3Nzox7fVD33+/6I2c4YgCaL512ekcDTe5HIz8/XvMyG3+/HkCFDVMfDv+itRi8g//CHP6CkpET1k1RerxeLFi1SPUOoF4h640OHDsVrr72mun+jRo1SXKogcpIYM2YMBg8erDo+c+ZMxcV0IwNRkiQMHz5c9S/BiBj/6le/qnp8Vr8B8Hg8mDBhguEXgWTH9S4UPmPGjKg3AH0f32AwGPUrBH3Hr7vuuqgzjH2fu2rXwUvFm7/wRZjVTJgwQfMyGbm5ufKfQYzl7LPPVhx7pLS0NBQUFKhexw0ADh48qHmGUG+8r8hA9Pl8+O///m95fM6cOYrfCQwGg4rjj3xsx44dq/i3Hzm+Zs0axdwZeYYwfP3LWI9d+ANietdJNHIdRauf+5MnT0ZeXp7qZWziPUOYbCB6PB7ceeed8ngqn/uifsRsNgagyYz+Hkg86+u9SEiS+sMcCAQwZcoU1fHwVdPV6AXkrl27kJmZKf9/5CepiouLkZ2drXqGUC8Q9cYvvPBCzavte71enH/++fL/R54hDAQCih8RR44XFRVpnmXweE5ci0ztL8GEL9RqZFzrUg/Dhw/Hq6++qjpu9huAG2+8UfNabnovAkbH9S4UPnz4cM3AHzVqVNQZ3r7jU6dORU5OTsxxj8eD7373u6iuro65bbPf/JWUlKCmpkZ1/ZycHASDQdXxzMxMxbFHGjhwoOZ19u699174fD7V6xjW1tZi4MCBqmcI9cYB7UDMzs6Wfx9O7fj7zk2RgThgwADFv+3I8ZtuuikqIPqeIZQkCXPnzlW9DqLH48ELL7xgaPyJJ55QPUM4ePBgPPTQQ6rHn4o3/+FroMa6jI3X68Udd9yh+ubQaCBa+dwHTvyIWe25bxcMQJOJCDwj43oXWk5LS8O0adNU1x81ahRGjRqlOq4XkDNnzoz5lzj6fpLK4/GoniH88pe/jAEDBqgGot7473//e/kTWX3vpw8//BC33HILJEnCPffco7r/+fn5mn8uy+/3KwI5MhA9Ho8iICOJ+D0hrb81PXr0aM2r/afqDUCyLwJGx/UuFJ6WlqY4gxp5lsDv9ysev8jx0aNHq74BSEtLw2uvvab6l2g8Ho9mnIt68xfrLOCf//xn3cuklJeXa/7b9fv9mtfZ27ZtG7KyslSvYxh+7NTOEOqN6wViZWUlHnjgAc39Ly8vVx3PzMzUDNyysjLFeKw3f1r3j4jnfjxnGNWk+s1/b28vVq9ejXPOOUc+OzxkyBDVN4dGA9HK5z5w4pPaWn+Fyg4YgCa77rrroiawvvSe5Hrr671IACc+Cal2oeV7771XMxB9Ph/8fn/SASlJkuIPdsfyxz/+Uf7vyDOE3/72t/Hyyy+rBiIAvPfee5rjq1atgt/vR1paGvLz8+W/g+r3+zF+/HjNM4RjxozBmDFjVMczMjIwadIk1fG0tDQMGjRIc9zoGwStvzUd/lNkVr0BeOWVVzB8+HD5/yOvZab3ItDY2IjS0tKkx8NvgLQuFN73x4SRBg0apPki6PP5FC+CfV8kPB4PtmzZEvVj1p6eHqxatSquM3yxAjHe9Y8cOYJp06YhPT0ds2bNwqJFi7B48WLMmjUL6enpyM7O1gykSy65BF6vVzUgvV4v5syZo7r+qlWrFL9DFnkdw4EDB6KgoED1DOGaNWtwwQUXqI7rBeLy5csxaNAgzf3X+lN8es/9zMxMjBs3Tv7/yDOE4bP/akQE4NatW1XPEL700kuG5m7Rz/2+tm/fjkWLFmHo0KGqbw71ntt6gWjlcx848QlxrV+xsAMGoMn0XsCNBp7eJLFgwQLs379f80LLWoG4du1aQwGp95c4wvsXqe8ZwrS0NPn2WIGoF5B//OMf8d5772HFihWora1FbW0tVqxYgffee0/3DGH4WLXOIM6aNUv1+CRJUnzSLJLHY+xDIHqPP2DtG4CLL74Yl19+ecyxeF4EwoyMP//886oXCj/vvPM0zxKMHDlS90Vw6tSpMceuu+46PPfccxgxYgQA4NVXX8XcuXORmZmJiooKSJKEV155JWq9cOCFr40XFrm+x+OJuX5fW7ZsQXNzM8444wxkZWUhMzMTZ5xxBpqbm9HU1IT8/HzVQBo0aBDGjh2rGpAlJSUYPHiw6vqDBw9W/VvD+/btw3333Ye8vDzVM4RhauN6gXjkyBFMmTIF6enpqK6ulve/uroa6enpKC4u1gzEzMxM1d+jDJ9Bra+vV73vx48fr/gRaUNDAxYvXiwvI0aMUFwoOdFxSZKwe/du1e0DsZ/7Ho8nJc/9r33ta4q/VBK5/7feeqti7o88Q/jFL34R99xzT9KBCMR+7ns8HtOf+wDwv//7v/Jz364YgCYz+0fARgNSLxCNBqRewMSzf+vXr4+6PRyIkS+S8Y73pXWG8P7779ccv/nmm3UDctSoUaqflCsqKsLo0aOTHpckSfHn8tRY9QbA5/OhpKREdf+DwaDiDULkGcLwi4DRcbULheudJQi/CKqNp6WlaZ4FmT59OsaPH4+SkhIMGzYMN910k/wCFfncjQy8O++8E7/+9a/xwx/+MOb6anNDOCAnTJigePMU9sknn8jjkiSpBt6UKVNw4MAB1YA8cOAAzj77bM31+54R0du/yDOEo0ePVlyGRW1cKyAPHz6MH/3oRzjjjDOQmZmJzMxMnH766fjRj36Ef//736isrFQNxC996UuYPHmy6nh2djZWrlyp+thfeeWVUR8QmzhxIqZNm4Zp06YhNzdXcQYx0XFJknDXXXepbj8s0ed+OBDVnvvxBmR6enrUJbj67v/IkSNVP8Xb981hWKKBGGbVc/9rX/sarrzySt3Hx0oMQJMZPYNjNKCMBqjRgPR4PHI0xVokSdL8g+Hx7J/WGUa98TC1M4TxjGsF4mmnnaY6yQEnzhL0/aRZouMej0fzF41fffVVBINB1QAbN24c1q9fr/un+JINyHPOOUfz+FeuXKn6Y7hYLwKJjl922WXwer2ax9/Y2Kh6hvC5557TPIN4zz33qL6IDB8+HJIkYebMmXjppZfkH1X1PUPR3d2tGnjV1dUYMGAArr766pjrxxOQfT8EEWu8q6tLNfCOHDkS8y9ZhAOysrISHo8n7vX19q+vvmcI4x3XC0i1/VcLxMOHD2PLli2q401NTcjLy1M9Q5ieno5rr71Wvi3yd8TKy8ujrjGayPjYsWNxyimnyP8feYatvr4ezz77rOr4rbfeioMHDwp/7ocDcdy4cYrL9ETuf2trq+J3MPXOEPYVTyBWVVXhv/7rvzS//9NPP616hlDtuR8e13ruX3rppQgEAoqfTtkRA9BkRgNIxPpGA9ToGcyVK1fi0Ucfjbl4PB789Kc/NW3/4v0RebLjYWqBWFBQIP8uYiwjR47U/DGB3vgjjzyC9PR01cDJycnBTTfdpLr+//t//w8DBgzQPANZVFSUdEAWFhZqHr/e/hcWFqKwsDDp8ZycHM1PwoYDVO0sQZjWuFogSpKEiy66CG+99ZZim/EGntfrxeLFi1XXB7QDMp7xWPoGUt8zdPEGXGRgJbp9vTOYeuNA7EBMZv/jPYMa6wyhz+dTBPCQIUMU12UcMmQIsrOzkx6/6667oj6EkMgZxOHDh2sGktGAHD58uOIvJSW6/yNGjNANOK1AnD59etRlfiK//4oVK4Q/9/sGpN0xAE0WTwBlZWVpBl5mZqahgLJzQKZi/4ycQY33DKiaQCCg+HuUkcJ/yzfZ8RkzZigm0UgDBw7EjBkzVMfPP/98zUt96J2B1AtIj8eDxx9/XHX96dOna+6/0TOkesevF6DxBnCsF4lXX30V119/PXJzc1FVVYX7778fH330UdyBt3HjRs319QJSbzxSoj+C1lv/lFNOQXZ2tqHt653BjBVwfQMxkQBN9gyq2hnCjIwMzQiIvMRUouNnnXWW5qdQ9c4gLliwIOoyJyIDMj09XfM6lHr7r7d/eoE4ZswYrFmzRvX7652BjDeA9QLSzhiAJosngIyeIdMLKL2A1AtQj8dj2o9wRQSy3v4Z2X+jZ0DHjh0b868VhA0fPlzz713qjQ8ZMgSjR49WHff7/ZpnEEeNGqUZmHpnIPUCcvDgwaiqqlId19t/o2dI9Y7faIDqBfC4cePw29/+Fq2trTj33HPlv1Jz33334Xe/+51m4IUdOHAg5vp6AWn0DGI8Aam1fqrPYCYToGaeQQ0EApq/fzxixAjVT8nGMx6+yHJYomcQzz77bMVlWkQHpNfrxfz585Pef73909t+WloalixZovr9zT6DGg5EO2MAmszs39ETEZB668+dO9dQoBkJLBH7pzeud/8YOQN60003Yfz48TEngs8++wyDBg3C4MGDkx73eDyYN2+e6vYLCgo0zyL4fD7NS13onYHUC8i5c+fC5/Mlvf9Gz5DqHb/RANUL4MjfcXzzzTdx2223YcSIEcjIyMBFF12kGnixziz3Xd/v96OwsFA1II2eQdQLOL31U3EG00iAmn0G9ZxzzsHYsWM1n9uDBg1KelzvuaN3BnHIkCGKcdEB+ZWvfAWZmZlJ77/e/ult3+fzaf4pO7PPoIZ/xGxnDECTWX0haLO/v16ghRcr98/oj6CNnEH84IMPMGrUKBQUFKC5uRnPP/881q5di2XLlqGgoADDhw/HiBEjkh73er2KS0VEqq6u1gwwn8+n+afk9M5A6gXkpk2bIElS0vtv9Ayp3vEbDVC9AN6xY4fiLEfY0aNH8dxzz0VdhiJWIMbSd329gEz2DKJewMVzhs/I9s0OULPPoH7wwQcYMWIECgsLsXz5cvnffnNzMwoKCjBs2DAMGzYs6XGv14uHH3445r8PQP8MYvgv9KgxGpCdnZ3weDxJ77/e/ultv6ioSBGIkcw+g/qLX/xC8xqxdsAAtFg813HTYvWngI3uv9n7J+IMqpEzoAMHDkRubi6qq6vlXxAOX6m/uroa7777Lnbt2pX0+LXXXqt5hrGiogLZ2dmqAZadnY2Kioqkz0DqBeSzzz6LMWPGJL3/Rs+Q6h2/0QDVC+C///3vmhcDVqMWiHr0AjKRM4hhagEX74+wk92+2QGqt/9Gz2ACJy4GPHPmzKh/+zNnzsQ777xjaHz+/PkYN25c0mcQ/X6/5nPXaEDu2LEDfr8/6f3X2z+97c+bNw8ej8eyM6h/+9vfNP+KlB0wAB3O7AAzGmhGpeIMYKrOgH788cfYsmULurq68PHHH0d9r2TG9c4wjho1Clu3blUNsC1bthg6Q6kXkOPHj5cvxZHM/hs9Q6p3/EYDNJ4ALikpUR03i15AJnIGsS+jP8JOZvtmBWi8+5/sGcy+9u3bh66uLnR1dWHfvn1R+5XMuNEzjFlZWSgrKzMtIJ966in5MjXJ7L/e/ult/8EHH4TX67XsDOqOHTs0/1ayHTAAyRCjAWo1q8+AiqB3BjFMLTCNnIHUC8hRo0Yp/jxWMvtvdFzr+I0GaCIB7ATJ/Ag6mfWT3b4ZAZrs9uM9g2omI2cQu7q6TA3IcePGoa6uLun919u/eLY/f/58y86g9g1gu2IAkqvZ/QxoIvTOIBpdP9mANHv78Y6rsTqA7SjZH0Gnan2zAjSZ7ScToKIle4bRzIAcOXJk3P/2k9m/RLZvxRnUeALYagxAIgPscAbQLowGqNWsDmASx2iAJktUgKaaWQFp9v4Z3b5dAtgqDEAiIgGcHsAkjlUBahW9gLT79u0ewGZhABIRERElyeoAThYDkIiIiMhlGIBERERELsMAJCIiInIZBiARERGRyzAAiYiIiFyGAUhERETkMgxAIiIiIpdhABIRERG5DAOQiIiIyGUYgEREREQuwwAkIiIichkGIBEREZHLMACJiIiIXIYBeNKxY8fQ3d2Nnp4e9Pb2cuHChQsXLly42Hbp6elBd3c3jh07llT3MABP6u7uhiRJXLhw4cKFCxcujlm6u7uT6h4G4Ek9PT3yHWl11XPhwoULFy5cuGgt4RNXPT09SXUPA/Ck3t5eSJKE3t5eq3eFiIiISJPRbmEAnsQAJCIiIqdgAArCACQiIiKnYAAKwgAkIrKvottfkhciYgAKwwAkIrIvBiCREgNQEAYgEZF9MQCJlBiAgjAAiYjsiwFIpMQAFIQBSERkXwxAIiUGoCAMQCIi+2IAEikxAAVhABIR2RcDkEiJASgIA5CIyL4YgERKDEBBGIBERPbFACRSYgAKwgAkIrIvBiCREgNQEAYgEZF9MQCJlBiAgjAAiYjsiwFIpMQAFIQBSERkXwxAIiUGoCAMQCIi+2IAEikxAAVhABIR2RcDkEiJASgIA5CIyL4YgERKDEBBGIBERPbFAHQ3PvbRGICCMACJiOyLAehufOyjMQAFYQASEdkXA9Dd+NhHYwAKwgAkIrIvBqC78bGPxgAUhAFIRGRfDEB342MfjQEoCAOQiMi+GIDuxsc+GgNQEAYgEZF9MQDdjY99NMcG4KpVq1BcXIxAIIDKykqsX79e8+s7OjpQWVmJQCCAkpISrF69WvVrn3jiCUiShNmzZ8e9PwxAIiL7YgC6Gx/7aI4MwCeffBI+nw8tLS0IhUKor69HdnY2du/eHfPrd+7ciaysLNTX1yMUCqGlpQU+nw9tbW1RX7tr1y6MHj0akydPZgASEfUTDEB342MfzZEBWFVVhZqaGsVtFRUVaGhoiPn1S5YsQUVFheK2hQsXYtKkSYrbjh49inPPPRcPP/ww5s+fzwAkIuonGIDuxsc+muMC8PDhw/B6vWhvb1fcXldXhylTpsRcZ/Lkyairq1Pc1t7ejvT0dBw5ckS+7Xvf+x4uueQSANANwEOHDqG3t1deuru7GYBERDbFAHQ3PvbRHBeA77//PiRJQmdnp+L2xsZGlJWVxVyntLQUjY2Nits6OzshSRL27NkDANiwYQNGjx6Nf/7znwD0A3Dp0qWQJClqYQASEdkPA9Dd+NhHc2wAbty4UXH73XffjfLy8pjrlJaWoqmpSXHbhg0bIEkS9u7di/3796O4uBi/+tWv5HGeASQi6j8UAbg098Ri1fYp5XjfR3NcAJrxI+DXX38dkiTB6/XKi8fjgcfjgdfrxdtvv627X/wdQCIi+2IAuhvv+2iOC0DgxIdAamtrFbcFg0HND4EEg0HFbTU1NfKHQA4ePIjt27crltmzZ2P69OnYvn07Dh8+rLtPDEAiIvtiALob7/tojgzA8GVgWltbEQqFsGjRImRnZ2PXrl0AgIaGBsybN0/++vBlYBYvXoxQKITW1lbVy8CE8VPARET9BwPQ3XjfR3NkAAInLgRdVFQEv9+PyspKrFu3Th6bP38+pk6dqvj6jo4OTJgwAX6/H8XFxZoXgg5/DwYgEfVLFgSQUEnsPwPQ3XjfR3NsANoNA5CIHIMByAB0Gd730RiAgjAAicgxGIAMQJfhfR+NASgIA5CIHIMByAB0Gd730RiAgjAAKRX4IkIA/hMwyUYMA5AB6DK876MxAAVhAFIq8EWEADAAGYCUIN730RiAgjAAKRX4IkIAGIAMQEoQ7/toDEBBGICUCnwRIQAMQAYgJYj3fTQGoCBOCkA+EZyLLyLuFhUwDMC46QWg2c8tpz93+8v+038wAAVhAFIqOH0SJmMYgCcxAFOuv+y/4edOP8IAFIQBSKng9EmYjGEAnsQATLn+sv8MwP9gAArCAKRUcPokTMYwAE9iAKZcf9l/BuB/MAAFYQCmhtMnIaPcfvxuxwA8yYUBKHT/jN5/DsQAjMYAFIQBmBpWT4JWc/okTMYwAE9iABrj5ABM8t8+AzAaA1AQBmBqWD0JWs02kzBZggF4EgPQGAags//9C8IAFIQBmBpWT4JWs80kTJZgAJ7EADSGAejsf/+CMAAFSWUA6j0J4x534JNAbxJPiNOPnxLm9PvO7AC0+/1j5LnfrwLQgrnPNnOPSQFoi2NLMQagIAzA1LB6ErSabSZhh3L6fccAZAAyABmAojAABWEApobVk6DVbDMJO5TT7zsGIAOQAcgAFIUBKIglAZjkJM4APMnpx08Jc/p9xwBkADIAGYCiMAAFYQCmhtWToNWEvoi4kNOPnQHIAHRrABr9t88AjMYAFIQBmBpWT4JWYwAa4/RjZwAyABmADEBRGICCMABTw+pJ0GoMQGOcfuwMQAYgA5ABKAoDUBAGYGpYPQlajQFojNOPnQHIAGQAMgBFcWwArlq1CsXFxQgEAqisrMT69es1v76jowOVlZUIBAIoKSnB6tWrFeMPPfQQzjvvPOTn5yM/Px8zZsxAV1dX3PvTnwLQzk8EqydBqzEAjXH6sVsdgFbff0bmLgZgHxYEoKj1GYDiODIAn3zySfh8PrS0tCAUCqG+vh7Z2dnYvXt3zK/fuXMnsrKyUF9fj1AohJaWFvh8PrS1tclfc80112DVqlV4/fXXsWPHDixYsAB5eXn4xz/+Edc+MQBTw+pJ0GpWT8JO5/RjZwAyABmADEBRHBmAVVVVqKmpUdxWUVGBhoaGmF+/ZMkSVFRUKG5buHAhJk2apLqNo0ePYsCAAXjsscfi2icGYGpYPQlazepJ2OmcfuwMwDjnLr3AYwAyABmAzgvAw4cPw+v1or29XXF7XV0dpkyZEnOdyZMno66uTnFbe3s70tPTceTIkZjr7N+/HxkZGXjxxRdjjh86dAi9vb3y0t3dzQBMAasnQatZPQk7ndOPvd8EYJL7zgBkADIAxXFcAL7//vuQJAmdnZ2K2xsbG1FWVhZzndLSUjQ2Nipu6+zshCRJ2LNnT8x1brzxRpxyyik4ePBgzPGlS5dCkqSohQFoLqsnQatZPQk7nbCAskjc+x/vuNr319u+UXYPQBP+bTAAGYB249gA3Lhxo+L2u+++G+Xl5THXKS0tRVNTk+K2DRs2QJIk7N27N+rrm5ubMXDgQGzbtk11P9x8BtDKJ4rVk6DVrJ6Enc7tAWj23BE3BqCx9RmACQccAzCa4wLQ7B8B//jHP0ZeXh62bt2a0H656XcAGYDWsdWLiJVMOgtgdykLwCTH48YANLY+A5ABKIDjAhA48SGQ2tpaxW3BYFDzQyDBYFBxW01NTdSHQJYvX47c3Fxs2rQp4X1iAKaG1ZOg1Wz1ImIlBiADkAGY/M4wABmAcGgAhi8D09railAohEWLFiE7Oxu7du0CADQ0NGDevHny14cvA7N48WKEQiG0trZGXQamubkZfr8fbW1t2Lt3r7x8+umnce0TAzA1rJ4ErWb0+BmADEAGIAOQAcgABBwagMCJC0EXFRXB7/ejsrIS69atk8fmz5+PqVOnKr6+o6MDEyZMgN/vR3FxcdSFoIuKimJ+qGPp0qVx7Q8DMDXMngSNTlJGu7Bk/gAAIABJREFU6W3f6QEo7L5lAMYXeAzAqDEGIPp3ABoddxHHBqDdMABTgwHIAATAAGQAMgCdHIAmPHfNHu+PGICCMABTI6FJJIlxBqC5GIDGMAAZgGbNfQltPwkMQPthAArCAEyC2QFjNAAtDCQGoA4GIAOQAaj+xTacGxmA9sMAFMSWAWh0PMntG93/eNYVMokzAC07C2AYA9CRAWj2+m4OQLu/OWYA2g8DUJB+GYA2fBFgANpjEmYAWoMByABkADIARWEACsIAFLf/8azLAGQAMgAZgImMMwD7YAAyAMEAFIYBKG7/41mXAcgAZAAyABMZt1UA2nDu0xu3y9zDABSHASgIA1Dc/sezrl0DUOi7fL1xvf1PYn2j+xfv+oYxABmAVgag1W++GICOC0Bhc59ADEBBGIDi9j+edRmADEAGIAMwkXEGYAL7zwBkAMaBAXgSA1Dc/sezLgOQAai5/xZN8mZzTABaNHcwABmADMD4MQAFYQCK2/941mUApuBFLon9i3d9w5Os2QFo00B0TQCasH8MwAT2nwFoXgDaaG5hAArCABS3//Gs2y8CUG/7et/fghe5eL+/GgagMQxABiAD0OIATHJuYAD2YwxAcfsfz7q2CMAkAs7o+gzAiH1jABob1/v+CY7LGICq+2bXuU9vnAF4EgNQxgA8iQEobv/jWZcByABkADIAExlnACaw/1bOrRYFHgMwcQzAkxiAMZiwvtWBk9D6ehiADMAEMQAZgAxABqAoDEBBGIDixmU2DJyE1tfDALRdANpxku6LAWhdAJodcEbXt/q5L2x9BmDKMAAFYQCKG5c5fZLTwwB0bwCa9CLCALRJAFr43HPq/ketywA0HQNQEAaguHGZ0yc5PQxA5wag2fuvggHIAOyv+x+1LgPQdAxAQRiA4sZlTp/k9PTnADR7kk1y3OoXAaPrMwAZgP11/6PWFfzctXpuStnckgAGoCAMQHHjsv48yZl0fHYZt+skywBkANo1oNw+HnXfpfi5bfW43vGbgQEoCANQ3LjMhpO01es7Zdz0SdDpk3SS6zMAGYD9dTzqvkvxc9vqcb3jNwMDUBAGoLhxmQ0naavXd8q46ZOg3SdpkyZ5BiADsL+OR913KX5uWz2ud/xmYAAKwgAUNy6z4SRt9fpOGTd9ErT7JO2U44vAAGQAWn7/WfTctnpc7/jN4NgAXLVqFYqLixEIBFBZWYn169drfn1HRwcqKysRCARQUlKC1atXR31NW1sbgsEg/H4/gsEg2tvb494fBqC4cVk/nKSt3r7dA7C/j8uSHBe+fxGsHo86foHf3y7PDY5z7jA0dwjkyAB88skn4fP50NLSglAohPr6emRnZ2P37t0xv37nzp3IyspCfX09QqEQWlpa4PP50NbWJn/Nxo0b4fV60dTUhB07dqCpqQnp6enYvHlzXPvEABQ3LjNhkjH7+9t9+5zEbTKJJzkufP8iWD0edfwCv79dnhsc59xhaO4QyJEBWFVVhZqaGsVtFRUVaGhoiPn1S5YsQUVFheK2hQsXYtKkSfL/X3nllZg1a5bia2bOnImrrroqrn1iAIobl5kwyZj9/e2+fU7iNpnEkxwXvn8RrB6POn6B398uzw2Oc+4wNHcI5LgAPHz4MLxeb9SPZ+vq6jBlypSY60yePBl1dXWK29rb25Geno4jR44AAAoKCrBixQrF16xYsQKFhYUxv+ehQ4fQ29srL++99x4kSUJ3d7fidjOWgkVPo2DR0+htGHBi6Wfj8hJjPLyu2vp642Z/f7tvP1Xj8hjHtR/7BMeF75+DnvtGv79dnhsc59xhaO4QuHR3d0OSJPT09CTVYykPwPfffx+SJKGzs1Nxe2NjI8rKymKuU1paisbGRsVtnZ2dkCQJe/bsAQD4fD6sWbNG8TVr1qyB3++P+T2XLl0KSZK4cOHChQsXLlwcu3R3dyfVY5YF4MaNGxW333333SgvL4+5TmlpKZqamhS3bdiwAZIkYe/evQBOBODjjz+u+Jpf/vKXCAQCMb9n5BnATz75BO+88w56enpMr/a+5a52xpHjHLdq3M77xnGOc9y+43beNzuMi156enrQ3d2NY8eOJdVjrv0RsNV6e7V/ds9xjls1bud94zjHOW7fcTvvmx3G7cayD4HU1tYqbgsGg5ofAgkGg4rbampqoj4EUl1drfiaWbNmxf0hkFSz+h8ixznOSZzjHOe4yHE775sdxu3G0svAtLa2IhQKYdGiRcjOzsauXbsAAA0NDZg3b5789eHLwCxevBihUAitra1Rl4Hp7OyE1+vFsmXLsGPHDixbtiyhy8CkmtX/EDnOcU7iHOc4x0WO23nf7DBuN5ZeCLqoqAh+vx+VlZVYt26dPDZ//nxMnTpV8fUdHR2YMGEC/H4/iouLY14I+plnnkF5eTl8Ph8qKirw7LPPmn0YSTt06BCWLl2KQ4cOcZzjthq3875xnOMct++4nffNDuN2wz8FR0REROQyDEAiIiIil2EAEhEREbkMA5CIiIjIZRiARERERC7DACQiIiJyGQYgERERkcswAImIiIhchgFIRERE5DIMQCIiIiKXYQASERERuQwDkIiIiMhlGIBERERELsMAJCIiInIZBiARERGRyzAAiYiIiFyGAUhERETkMgxAIiIiIpdhABIRERG5DAOQiIiIyGUYgEREREQuwwAkIiIichkGIBEREZHLMACJiIiIXIYBSEREROQyDEAiIiIil2EAEhEREbkMA5CIiIjIZRiARERERC7DACQiIiJyGQYgERERkcswAImIiIhchgFIRERE5DIMQCIiIiKXYQASERERuQwDkIiIiMhlGIBERERELsMAJCIiInIZBiARERGRyzAAiYiIiFyGAUhERETkMgxAIiIiIpdhABIRERG5DAOQiIiIyGUYgEREREQuwwAkIiIichkGIBEREZHLMACJiIiIXIYBSEREROQyDEAiIiIil2EAnnTs2DF0d3ejp6cHvb29XLhw4cKFCxcutl16enrQ3d2NY8eOJdU9DMCTuru7IUkSFy5cuHDhwoWLY5bu7u6kuocBeFJPT498R1pd9Vy4cOHChQsXLlpL+MRVT09PUt3DADypt7cXkiSht7fX6l0hIiIi0mS0WxiAJzEAiYiIyCkYgIIwAImIiMgpGICCMACJiOyr6PaX5IWIGIDCMACJiOyLAUikxAAUhAFIRGRfDEAiJQagIAxAIiL7YgASKTEABWEAEhHZFwOQSIkBKAgDkIjIvhiAREoMQEEYgERE9sUAJFJiAArCACQisi8GIJESA1AQBiARkX0xAImUGICCMACJiOyLAUikxAAUhAFIRGRfDEAiJQagIAxAIiL7YgASKTEABWEAEhHZFwOQSIkBKAgDkIjIvhiAREoMQEEYgERE9sUAJFJiAArCACQisi8GIJESA1AQBiARkX0xAN2Nj300BqAgDEAiIvtiALobH/toDEBBGIBERPbFAHQ3PvbRGICCMACJiOyLAehufOyjMQAFYQASEdkXA9Dd+NhHYwAKwgAkIrIvBqC78bGPxgAUhAFIRGRfDEB342MfzbEBuGrVKhQXFyMQCKCyshLr16/X/PqOjg5UVlYiEAigpKQEq1evVv3aJ554ApIkYfbs2XHvDwOQiMi+GIDuxsc+miMD8Mknn4TP50NLSwtCoRDq6+uRnZ2N3bt3x/z6nTt3IisrC/X19QiFQmhpaYHP50NbW1vU1+7atQujR4/G5MmTGYBERP0EA9Dd+NhHc2QAVlVVoaamRnFbRUUFGhoaYn79kiVLUFFRobht4cKFmDRpkuK2o0eP4txzz8XDDz+M+fPnMwCJiPoJBqC78bGP5rgAPHz4MLxeL9rb2xW319XVYcqUKTHXmTx5Murq6hS3tbe3Iz09HUeOHJFv+973vodLLrkEAHQD8NChQ+jt7ZWX7u5uBiARkU0xAN2Nj300xwXg+++/D0mS0NnZqbi9sbERZWVlMdcpLS1FY2Oj4rbOzk5IkoQ9e/YAADZs2IDRo0fjn//8JwD9AFy6dCkkSYpaGIBERPbDAHQ3PvbRHBuAGzduVNx+9913o7y8POY6paWlaGpqUty2YcMGSJKEvXv3Yv/+/SguLsavfvUreZxnAImI+g9FAC7NPbG4aPtuxwCM5rgANONHwK+//jokSYLX65UXj8cDj8cDr9eLt99+W3e/+DuARET2ZXWAWb19t2MARnNcAAInPgRSW1uruC0YDGp+CCQYDCpuq6mpkT8EcvDgQWzfvl2xzJ49G9OnT8f27dtx+PBh3X1iABIR2ZfVAWb19t2OARjNkQEYvgxMa2srQqEQFi1ahOzsbOzatQsA0NDQgHnz5slfH74MzOLFixEKhdDa2qp6GZgwfgqYiKj/sDrArN6+2zEAozkyAIETF4IuKiqC3+9HZWUl1q1bJ4/Nnz8fU6dOVXx9R0cHJkyYAL/fj+LiYs0LQYe/BwOQiPolpwdIEvtvdYBZvX23YwBGc2wA2g0DkIgcw+kBwgCkBDEAozEABWEAEpFjOD1AGICUIAZgNAagIAxAInIMpwcIA5ASxACMxgAUhAFIqcCL2RKA/wREshHh9ABhAFKCOG9GYwAKwgCkVGAAEgAGIAOQEsR5MxoDUBAGIKUCA5AAMAAZgJQgzpvRGICCMAApFRiABIAByACkBHHejMYAFMRJAcgngnMxAN0tKiAYgHHTCzCzn1tOD0Cnzz1O3nezMAAFYQBSKjh9EiZjGIAnMQBTzulzj7DnTj/CABSEAUip4PRJmIxhAJ7EAEw5p889DMBoDEBBGICUCk6fhMkYBuBJDMCUc/rcwwCMxgAUhAGYGk6fhIxy+/G7HQPwJBcGoND9s3r7FmAARmMACsIATA2nT0JGuf343Y4BeBID0Birt29Ekv/2GYDRGICCMABTwzaTkEXcfvxuxwA8iQFojNXbN4IBKAwDUBAGYGrYZhKyiNuP3+0YgCcxAI2xevtGMACFYQAKksoA1HsSxj3uwCeB03+R2ijbTMIO5fT7zuwAtPv9IyygnB6ARuc+BqDq4+8mDEBBbBmAepO8AwOKAWiTSdihnH7fMQAZgAxABqAoDEBBGICpwQC0ySTsUE6/7xiADEAGIANQFAagIJYEYJKB5+SAYgDaZBJ2KKffdwxABiADkAEoCgNQEAZgajAAjU3CtpnELeL0Y2cAMgDdGoBG/+0zAKMxAAVhAKYGA5ABaITTj50ByABkADIARWEACsIATA0GIAPQCKcfOwOQAcgAZACKwgAUhAGYGgxABqARTj92BiADkAHIABTFsQG4atUqFBcXIxAIoLKyEuvXr9f8+o6ODlRWViIQCKCkpASrV69WjD/00EM477zzkJ+fj/z8fMyYMQNdXV1x709/CkA7PxEYgAxAI5x+7FYHoNX3n5G5iwHYhwUBKGp9BqA4jgzAJ598Ej6fDy0tLQiFQqivr0d2djZ2794d8+t37tyJrKws1NfXIxQKoaWlBT6fD21tbfLXXHPNNVi1ahVef/117NixAwsWLEBeXh7+8Y9/xLVPDMDUYAAyAI1w+rEzABmADEAGoCiODMCqqirU1NQobquoqEBDQ0PMr1+yZAkqKioUty1cuBCTJk1S3cbRo0cxYMAAPPbYY3HtEwMwNRiADEAjnH7sDMDk5zYGYB8MQAYgHBiAhw8fhtfrRXt7u+L2uro6TJkyJeY6kydPRl1dneK29vZ2pKen48iRIzHX2b9/PzIyMvDiiy/GHD906BB6e3vlpbu7mwGYAgxABqARTj/2fhOASe47A5AByAAUx3EB+P7770OSJHR2dipub2xsRFlZWcx1SktL0djYqLits7MTkiRhz549Mde58cYbccopp+DgwYMxx5cuXQpJkqIWBqC5GIAMQCOEBZRF4t7/eMfVvr/e9o2yewCa8G+DAcgAtBvHBuDGjRsVt999990oLy+PuU5paSmampoUt23YsAGSJGHv3r1RX9/c3IyBAwdi27Ztqvvh5jOAVj5RGIAMQCPcHoDCnvsWBEg822cAxqkfBmDC/3YZgM4LQLN/BPzjH/8YeXl52Lp1a0L75abfAWQAWkfoJOzk+8+kswB2l7IATHI8bgxABmCS6zMAxXFcAAInPgRSW1uruC0YDGp+CCQYDCpuq6mpifoQyPLly5Gbm4tNmzYlvE8MwNToNwGTJAbgSQxABiADMPmdYQAyAOHQAAxfBqa1tRWhUAiLFi1CdnY2du3aBQBoaGjAvHnz5K8PXwZm8eLFCIVCaG1tjboMTHNzM/x+P9ra2rB37155+fTTT+PaJwZgavSbgEmSrV5ErMQAZAAyAJPfGQYgAxAODUDgxIWgi4qK4Pf7UVlZiXXr1slj8+fPx9SpUxVf39HRgQkTJsDv96O4uDjqQtBFRUUxP9SxdOnSuPaHAZgaKZ2kLaC3fVu9iCRB2H3LAIwv8BiAUWMMQPTvADQ67iKODUC7YQCmBgPQRi8iSWAAGsMAZAD2iwA04blr9nh/xAAUhAGYGglNIhZMckYxAOPEAGQAMgDVv9iGcyMD0H4YgIIwAJNg9SRo9PubgAEYJwYgA5ABqP7FJoxbPfcwAMVjAAriygA0aRKPZ12zJnG7BJJdA1DUWQDDGICODECz13dzAJodiHaZexiA4jAABemXAWjDFwEGoD0mYQagNRiADEAGIANQFAagIAxAcfsfz7oMQAYgA5ABmMg4A7APBiADEAxAYRiA4vY/nnUZgAxABiADMJFxBmAfDEAGIBiAwjAAxe1/POvaNQBFBVKyAZjSgEwCA9AYBqBNAtDquYUB6LgAFDb3CcQAFIQBKG7/41mXAcgAZAAyABMZZwAmsP82nlsZgOIwAAVhAIrb/3jWZQAyADX336JJ3myOCUCL5g4GIAOQARg/BqAgDEBx+x/PugzAFLzIJbF/8a5veJI1OwBtGoiuCUAT9o8BmMD+23hudXwA2mhuYQAKwgAUt//xrMsAtOZFLt7vr3d8DMDkMAAZgAxAiwMwybmBAdiPMQDF7X8869oiAPW2nwQGYJwYgAxABqB5+2/13KqxPgNQHAagIAxAcfsfz7oMQAYgA5ABmMg4AzCB/bdg7ohaN8WBxwBMHAPwJAZgDCaszwBkACr2jQFobFzv+yc4LmMAqu4bA5ABaJe5hQEoCANQ3LiMAZjw/jEAtceFvQhYhAFoXQCmdG5JYn2rn/vC1mcApgwDUBAGoLhxmd0naQYgAzBZJr2IMABtEoAWPvecuv9R6zIATccAFIQBKG5cxgBMeP+sfhFwTQCavf8qGIAMwP66/1HrMgBNxwAUhAEoblzGAEx4/6x+EUjZJJvkuNUvAkbXZwAyAPvr/ketm+LAM3tuStnckgAGoCAMQHHjMrMnuRhSOkkmsX2njNt1kmUAMgDtGlBuH4+671L83LZ6XO/4zcAAFIQBKG5cZvYkHQMDkAHIAGQAGn5u2/i5adfxqPsuxc9tq8f1jt8MDEBBGIDixmVmT9IxMAAZgGbuX9zjKhiADMD+Oh5136X4uW31uN7xm4EBKAgDUNy4zOxJOgYGIAPQFgGYquOLwABkAFp+/1n03LZ6XO/4zeDYAFy1ahWKi4sRCARQWVmJ9evXa359R0cHKisrEQgEUFJSgtWrV0d9TVtbG4LBIPx+P4LBINrb2+PeHwaguHGZ2ZN0DCl9kbBg+3YPwP4+LktyXPj+RbB6POr4BX5/uzw3OM65w9DcIZAjA/DJJ5+Ez+dDS0sLQqEQ6uvrkZ2djd27d8f8+p07dyIrKwv19fUIhUJoaWmBz+dDW1ub/DUbN26E1+tFU1MTduzYgaamJqSnp2Pz5s1x7RMDUNy4zIRJRo/Zk5zV2+ckbpNJPMlx4fsXwerxqOMX+P3t8tzgOOcOQ3OHQI4MwKqqKtTU1Chuq6ioQENDQ8yvX7JkCSoqKv5/e2ceHUWRx/E+5kpmct/HJCFILiLhSoJiIHJsiMsVLoFsCHmLQlwICQIi6iogSPLWmAgsombFdZGHyBGOlVUQRFATvCICisq5q8ATDChREPjuH7zune50dU+mM8y4qc97/V7Sv6n+Vc10dX/r112/kuybOnUq+vTpI/4/btw4DBkyRPKZvLw8jB8/3qk6UQHYfnYRN1xktHD3Rc7T/ulF3Esu4i7a271+Mjxtb9X+djy+t/QNaqfXDl3XjnbkNycAr1y5Ap7nWz2eLSsrQ79+/RTL5OTkoKysTLJv48aNMBgMuHr1KgDAbrejurpa8pnq6mrExcUpHvOXX37BxYsXxe3UqVNgGAanT5+W7HfHZi9/Dfby13Bxnt/N7f/MLm4KdqEsqbyW3dnv1tXje7v/W2UXbdSufm630d7u9fsN9X29x/eWvkHt9Nqh69rRjtvp06fBMAyam5td0mO3XAD+5z//AcMw2L9/v2T/4sWLkZSUpFimS5cuWLx4sWTf/v37wTAMvv32WwCA0WjEmjVrJJ9Zs2YNTCaT4jEff/xxMAxDN7rRjW50oxvd6Pab3U6fPu2SHvOYAHzvvfck+5988kkkJycrlunSpQuWLFki2bdv3z4wDIPvvvsOwE0B+Oqrr0o+849//ANms1nxmPII4A8//IBvvvkGzc3NblftjsqdFHGkdmr3lN2b60bt1E7t3mv35rp5g729t+bmZpw+fRrXr193SY912EfAnubiRfVn99RO7Z6ye3PdqJ3aqd177d5cN2+wexsemwRSWloq2Zeamqo6CSQ1NVWyb9q0aa0mgeTn50s+M2TIEKcngdxqPH0iUju104s4tVM7tben3Zvr5g12b8OjaWDq6upw+PBhlJeXw2q14sSJEwCAefPmoaioSPy8kAamoqIChw8fRl1dXas0MPv37wfP81i6dCmOHDmCpUuXtikNzK3G0ycitVM7vYhTO7VTe3vavblu3mD3NjyaCDo+Ph4mkwk9e/bEO++8I9qKi4vRv39/yef37NmDHj16wGQyISEhQTER9Pr165GcnAyj0YiUlBRs2LDB3c1wmV9++QWPP/44fvnlF2qndq+ye3PdqJ3aqd177d5cN2+wext0KTgKhUKhUCiUDgYVgBQKhUKhUCgdDCoAKRQKhUKhUDoYVABSKBQKhUKhdDCoAHQzL7/8sttfCD137hzRdvLkSdy4ccOt/tVYsGABLl++3GH9UygUCoXijVAB6GY4jsPZs2dVP3Px4kXFTN7Xrl0jTie/ceMGtm/fjoKCAuJyd874v3DhAp599llFP83NzUQbAFy/fh1btmzBiBEjxH2XLl3CxYv/y1Su5f/o0aMYP3480f+ECRPwzTffEMt/9tlnmDZtmvj/9u3bUV9fL25a/ltaWlBfX49Lly61sl28eBH19fWqAv6zzz7DzJkziXZP4+kBgKcpKSlR/G1vBZ4e/N2Ktn/yySduPb4Wv/76K9F2K859tfa/8847qvVrDzzdfm/Gk33/twIVgG6GZVlVAbJx40Z06dJFMUp1+fJlJCUlYcuWLeK+b775Bo888ghiY2MRGBiIe++9F/Pnzxft8+bNQ0VFhbgxDIOTJ08S/S9cuBBjxowh2seOHYsnn3xSsu/o0aOYN28eoqKiYDKZEBERIdpsNhs4jhM3hmGwY8cO4vHvu+8+zJkzh2ifO3euROABN4XZc889h8zMTDAMA4vFIvHPsqy4MQyDF198kXj8mpoaDBgwgGgfOHAgli9fruo/KChItNntdgQFBYmb0WhEY2Mj8fh6BDgAfP/993jmmWeIdk8PAPr164cffviB6F8PQt0zMjKIn/Fk+zmOw7Fjx1Tb4MnBn6s0NzdjxYoV6NGjBziOI36uU6dO+P7771WP5Ur7AeDQoUOoqKhAeHg48TOebr+7/AM3219eXo7g4GBx38qVK1FbWytuLMuKS6UqQfu+e9v/448/tqFFnoEKQDfDsqzqKH3w4MF44YUXiPa6ujoMHDgQr7zyCvr37w+z2YyhQ4eC53kcPHgQzz33HIYOHSp+3mazITs7G7m5ucjNzQXDMFiwYAHx+BkZGdi5cyfRvnPnTnTv3h0tLS1YvXo1cnJyYDQawXEcamtrUVRUJFmn2WazYc2aNdizZw92794NhmFUBWZycrKqQPrwww+RlJQE4GYuyKKiIvj6+oLjODz00EMYMGCARODZbDZJxJBlWVWBl5mZKRHYcrZu3YrMzEyi/+LiYjz88MMS/1VVVVi9ejVWr14NhmEwadIk4vFdEeA3btzAjh07MHbsWBiNRsl613IByjAM9u/f367+23MAsGvXLqSmphIvwmlpadi7d69k/1tvvYXx48fDYrEgJiYGw4YNE23eNABiGAYcxxHPb08P/tr63e/atQuFhYXw8fFBSkoK5s6di5qaGpf9t7X9P/74I1544QX06dMHPM8jOzsbo0ePFu0TJ05EQUGBuDEMg6NHj3pt+9vqX97+pKQkJCYminabzYbY2FgkJCQgISEBDMO0Wh7VEdr33dt+nudV723eABWAboZlWdxzzz2SC5PjZrFYMHjwYGL5iRMngmVZ9OnTB8uXLxdH1AaDAYcOHUJOTo5kXWUlAWS32yUjQ8fNbDbj8ccfJ/qvr6+H0WiEv78/evfujZqaGpw5c0b0n5ycLOmkSv6NRiN69OihuLEsi65duxL9NzY2wmAwoHPnzoiMjERFRQUOHDgg+o+Pj8eBAwdU/fM8T/TP87ymf4vFQvTftWtXvP3226r+ExISiMd3VoADwPHjx/HYY4/BbreD4zgUFRWhoKAADz30kMR/WwSopwcAw4YNU71J1dbWYuTIkTh58iSeeOIJxMfHIyQkBBzH4fXXX3dqALRw4UKPtJ9lWYwdOxZ/+MMfFI99KwZ/am135rvPy8vDokWL0KlTJ4SHh2P69Oniua938OlM+3/3u9/h3XffRXFxMWw2G26//XbwPI99+/ahqqoKhYWFEv+jR4/G5MmTMXknfyTOAAAdpklEQVTyZDAMg9mzZ3u0/Xq//5EjRxLbP2jQILz66qsS//Jrz5133kk8Pu377m1/UVERse97C1QAuhmWZXHvvfeKFyX5xvM8Ro4cSSzP8zx4nm/1LoNwEYqIiMDnn38u7g8NDcXx48cl/gURorSxLIuoqChV/yaTCV988YWif19fX3EJPwCorq6WjOgEAfbEE08oblarVVWgGI1GWCwW7NixQ/KoSPBvNpslj9kOHDiAq1evSvxzHEf0bzKZcP/99xP9m0wmGAwGon+bzSb5vsvLyyWPvViWhclkkjwacNxsNpvk95Nz9OhRWCwWDBgwABaLBQUFBVi/fn2bBGhwcLBHBwBRUVHE9tvtdtVRcnV1NcxmM3x9fTFmzBhs3rwZV65cadMAiOd5SVRUHiENCAhwS/uFG7CPj4/HBn9qbec4Dv7+/kT/OTk5YFkWEyZMwLZt23Dt2rU2+09NTUVTU5PiFhYWhq1btxL9z5kzBzzPIyYmBrNnz8ann34q8Z+VlYXt27er+o+KiiKe+0FBQZLovTvaHx4eLolKOW5+fn6q174HH3wQBoOB2P6YmBhxH8m/xWKhfd8DfR8APvjgA8TFxRGP7w1QAehmtN4BTElJwSuvvEK0T5s2Db6+vrBarRg3bhy2bt2KX3/9VTwJLRZLK3Em96/2nlBubq4kgiQnISEBPM9j4sSJeOONN8SXigX/QUFB2Ldvn6p/tU42duxYVQFstVrh6+uL+fPn48iRI+J+wX9UVBTeeustVf9hYWFEe3Z2NpYuXUq0h4aGwmw2E/37+/ujoaFB1b/wKFBpYxgGLMsSywcEBIDneaxatQoXLlxo5d8ZAcowjEcHAHra3x4DILPZLEZE5ZuPjw8effRRt7SfZVkMGzYMHMd5bPCn1naDwYDKykpN//LHqG3xL/y+8q0tv70gvOT+Q0JC8OWXX4r7e/XqhdOnT0v8qw1+GYZBbGysW9vP87wYkZJvLMsiOzvb5fabzWZ8/fXX4v5z585JBqm073uu7wM3J+E4vp7jjVAB6Ga0BOD8+fMRFxeHM2fOtLJ99913iIuLw/z583H8+HH8+c9/RlxcHEJDQ8FxHNavX4/bbrsNr7/+uqr/+Ph4ov3111+HwWDAsmXLJBeaa9eu4dlnn4XRaMTKlSuxYMECJCQkICIiAmVlZTAYDDh8+DAGDBig+piFZVncddddRPvHH38Ms9mM0aNHo6GhAc3NzWhubsYHH3yAUaNGwWw2o66uDiUlJbDZbOjZsyeqq6tF//fee6/kPRA5DMNg+PDhRPuqVatgtVoVIxFbtmyB1WrFnDlziP7vuOMOLF68WLX9SUlJ2LNnj+LWvXt3TJgwgVjebDYjICAAzz//vOTi0hYBarVaifZbMQCwWq3E9kdHR2PRokXE8oMGDQLHcbjzzjuxcuVKUQT/FgZALMti69atkklCjnh68JeYmCiJoMhZsmQJbDYb/P39kZWVhWXLluHcuXNt8m80GnHixAnFrXPnzqoTmMaMGQOj0Qi73Y65c+fi4MGDAP733fv4+Ij7SP4dJ4h5ov16vv+JEyfCYDAQ2x8XFyeJgCr5j4mJIdpp33dv+/ft20fs+94CFYBuRksAXrp0CV27doWfnx9KS0tRU1OD2tpaTJs2DX5+fkhLS5OMgG7cuIE33ngDY8eOhdlshtVqRXBwMH7++edWx25paQHDMJgyZYpqHefPnw+WZeHv74/u3bujR48e8Pf3Fyc6OPLmm2+KL+F26dIFBQUF4Hkey5cvl4w+BQGpNgtXmEmVnZ2NsLCwViPEsLAw1NfXi5//8ccf8fzzz6NPnz5gWRa5ubl49NFHYTabMWbMGDQ2NooCsqGhAaNGjQLDMKoRQgAoLCwUH1eNHDkSBQUFSElJAcdxGD9+vKp/YVLItm3bWh13y5YtYBgGf/nLX4i+tQS4wWBAWVkZ7r77bvj4+GDUqFHYuHEjjEYjDh065JQA7d27t8v+9Q4AGIZRHQBMnz4d6enpxPM3PT0dpaWlWL16Nfr16wez2Yzhw4eL78F58wCIZVlMmzaNOAnJ04M/Z777GTNm4PLly6irq0Pfvn3Fd6BqamqQmJioy7+z7d+zZw8mTZoEq9WKbt26ie/Ade3aFS+//LKqf2ECmTe231n/pPaXlJQQ3/G7ceMGGIaRXL/k0L7v3vbPmjVLdQKiN0AFoJthWVbyWEKJ5uZmlJaWIjg4WHxsERwcjNLSUtWp7OfPn8fChQthMBgQFxeHqqoqbN68GfX19aisrITdbofNZtNMRQEADQ0NKCsrwz333IP8/HzMnDlTNbIkTKHv3r27GMpXEpD33XdfqzQPjjOpLBYLRowYgZaWFmzcuBFVVVWorKzEpk2bVBM4Hz58GA8++CDCw8PBcZx4Y3TcQkJCsGrVKkkurM8//1zyHpLwCGHdunUYMWIE0tLSkJqaihEjRmDdunVO+Rd+M0cBmZqaKj7m0EoF4awA//rrr8VZoCzLYuLEiSgvL9clQNviH2jfAQAAnDlzBtHR0bDb7aisrBTP36VLl8JutyM6OloiEIRzJzo6Gv7+/khKSkJsbKxXDoBYloXBYMD69esV/Xp68NfW7x4AvvjiC8yZMweRkZHgeR42m81l/5cuXUJKSorT7b906RJWrlyJrKws8d3AwMBAxVQn3377LRiGQXl5ude2v63+5e3v0aMHLBYLsrKy8Nprr+HTTz9FU1MT1q1bJ6ao+uCDD4j+Adr33dl+o9FI7PveAhWAbkYrAugo8G7cuIFz587h7NmzomjRKg8Ax44dQ15eHjiOk7xjk5eXh507d+pKtOwMH330Ed5//32UlZUhPz8f+fn5KCsrw/vvvy9+hjSTSi1XklKuJTm//vorNmzYgMuXL2Pjxo2orKxEZWUlNm7ciJ9++gl79+6VRMCEqfqO35NWhFANwf/atWsxYsQIpKamIjU1FcOHD8fatWuRkJCAf//738TyQrJgkgBXSiZ8/fp1/POf/8To0aNhMplgMpmIAjQtLc2plVDcNQDw9/cn5oITcrmFhIQgPz+/1e+Sn58veadH/h1s2bIFeXl5YBiGOACKiorCV199JZaTJwoXhLM72s8wDKZOnar6vbtz8BcZGYm3334bLS0trcpfvnwZTU1NOHbsWJu/e+DmTe6ll16C2Wwm+jeZTJJ39AQc8xgajUZi+9VyyH322WcoLS0Fz/Pw8/PDAw88IArI0tJS+Pn5ITQ0VDG66MiJEyfc1n6e5zUTZav5t9vtxL4jJKAPDAxEamqqWE44TmpqKqZMmUL7vof6PsdxqtFBb4EKQDejJeD02h05f/48Ghoa0NDQgPPnzwMA7r//ftVEy1OmTMFtt91GFIhDhw7F0KFDXRaQDQ0NuO+++4gzqZRQihAqoSUQr1+/jpycHKSnp4v7bDYb3nnnHZw4cQLHjx9HRUUFRo0aJdpJEUJn7XK0kpHqtZ87dw5PP/00UYDqXWnFGZwZAAjIc5n17dtXTAVx4cIFNDY2oqGhQTLhRWs1hcbGRuIA6IUXXhDT6ACtE4ULj1Pd0X5PD/4qKyvRq1evVpMIgJsCplevXuI7iErf/bVr19DU1KQoIFtaWtDU1ISjR48S/cvPK3kew8LCQvEdOKX2A+ornQA3rxVTp05FUFCQ6D8oKAhTp07FoUOHJHkS+/btK0kB1bt3b3FwRjr3tFD7/rX6rmOibCX/zvz+QsaDTz75BOvWrcO6devw8ccfi98N7fue6ftK7fdGqAB0MyzL4u9//7tk5OG4sSyr+R6LWnmO41TLJyUlqU61LygoUH1RtVu3bqrZ1rUEJMuyKCkpIc6kEmhLhFBLIDraWZZFTk6OaJNP1X/xxRdhNBoldseLifw9Iq0IolwgunsAoLXcmKcHAAKkXGZqaK32cOjQIcyaNUu0Kw2Ahg4dilWrVoll5L9/ZWUlBg0aJP6vFCVwbLvcrvUSvicHf3379sXatWuJZdauXYuePXsSI4QLFy50WkAq+QeAn3/+mZjHkITWSieCfdSoUaL9xo0bOHv2rERAPvbYYygpKRHL2Ww2lJWViSmgsrOz8eCDDxLr8e2336oKyMzMTFFAKrXf3b+/1kortO97ru//VqAC0M0opUBw3IQQsppAdKY8CYvFIpmqLicxMVF1qnp8fLxqLiMtAZmQkACj0UicSeVshFBLIJLs8vYLj4sFhg0bBp7nxf/lEcLk5GR069aNaB87diwCAwMldkeByDCM6iiTZdVXitGya0UZPD0AyMjIQGJiIjGXmRy1KIEzdkeEWdKO7ZPfBJYtWwaDwSCxy/uYYyJvpSiC4wDM8SbBsixmzZrlscFfWFiY6mPMqqoqGAwGosCz2Wx44IEHiOW1BOS4ceMQFBREzGMoRy1C6IxdQBCIAQEBrX5bx99+zZo1CA0NFf+XC7yoqCgUFxdLyrdFQLIsi927dxPzILIsK8nh2d7lad/3XN93fMTszVAB6GacGeXpEXhax4+IiMCuXbuIdpPJJLkIyjGbzaoCUUtAfvjhh+jUqRNxJhXP8ygvLydGCLUEopY9LCwMu3fvJtYvOjpakqdQfpGIjY1FSEgI0Z6XlydJpisXiAzDICIigrgSTHvY77nnHmL7PD0AYFkWPXr0IOYyE9CKEjgbRZDfJFiWxSOPPCLa5YnCBw4c2GoA4Pj7pqamtnqFwNE+efJk+Pn5SeyOfZeUB+9WDP58fX3R1NREtPfo0UM1TYa/v7+4DKISWgJSSMFEyuMGaEcI2xJBlAtEo9GIv/71r6K9oKBA8k7g9OnTW/32jgLP19dXstRcWwWkkANP6bdzzIOnJvCcKU+C9n3P9f32esTsbqgAdDNaEZq2vCdEKq8WJejbt69qslGz2Yx+/foR7ULWdBJaAvLEiRPw8fER/5fPpEpISIDVaiVGCLUEopZ96NChksdAcniex9133y3+L48Qms1mySNiuT0+Pl41ysCyN3ORkVaCERK16rGrpXrw9ADggQceUM3lVllZieTkZGKUQMsuQLpJaCUKj4iIUBX40dHRrSK8jvb+/fvDZrMp2lmWxWOPPYb8/HxF3+4e/GVkZGDlypVEu81mQ2pqKtHu4+MjabscLQE5Y8YMBAYGEvMYlpaWqkYIteyAukC0Wq3i+3BKJCUlSfIEyn9bq9WK2267Tfy/rQKSYRgUFhYS8yA6K/DUyq9du5YoIENCQvD8888T20/7vvv6PnDzETOp73sLVAC6mfYQeHojiAzDEBMtcxyH3Nxc4vGjo6MRHR1NtGsJyLy8PMWVOBxnUrEsS4wQDh48GH5+fkSBqGV/++23xRlZjt/j2bNnMWvWLDAMg6effppY/8DAQNXlskwmk0QgywUiy7ISASlH73tALKu+1nRMTIxqtv9bNQAg5TLjeR7z588nRgm07Fo3Ca1E4RzHSSKo8iiByWSS/H5ye0xMDHEAwHEcdu/eTVyJhmVZVXHuzLVh0aJFxMFfcXEx/Pz8FKOAn376qWaalOTkZNVzV0tAHjt2TFydQSmPofDbkiKEWnYtgdizZ08sX76cWD8fHx+kpaWJ/8sFnq+vr6rA0RKQLKueh5BlWRw4cEBV4H344Yeq5bUEJMOQb/G077uv7wM3Z2qrrULlDVAB6GYmT57c6gLmiNYNXm95ANi6dSsx0fIzzzyjuhKH0WiEyWRyWUAyDCNZsFuJjz76SPxbHiF8+OGHsX37dqJABIBTp06p2lesWAGTyQSO4xAYGCiug2oymZCenq4aIYyNjVVdLspisaBPnz5EO8dxCA4OVrXrHSCorTUtLEXmqQHAzp07ERERIf4vz2Vmt9sRGhpKjBIsXrwYXbp0Idq1bhLCSjNqicIdHxPKCQ4OVr0JGo1GyU3Q8SbBsiwaGxtbRcmam5uxYsUKpyJ8SgLRsbzWI2aGYWAwGDBkyBCUl5ejoqICQ4YMgcFggNVqVRVII0eOBM/zLgvIpqYm+Pr6iv/L8xgGBQXBbrcTI4Rr1qzBoEGDiHYtgVhVVYXg4GDV+s+cOZNY/5SUFNUIp5aAFKL/JNpj8KcmILdt26br2t3eff/ixYtYuXIlMjMzwXHc/3XfB24OgNTOH2+ACkA3o3UD1xJ4egVCSUkJLl26pJpoWU0g1tfX6xKQWitxCPWT4xgh5DhO3K8kELUE5EcffYRTp06huroapaWlKC0tRXV1NU6dOqUZIRTaqhZBHDJkCLF9DMNIZprJYVl9k0C8fQAwfPhwjBkzRtF28OBBlJeXIywsjBglECDZtW4SALB582ZiovC77rpLNUoQFRWleRPs37+/om3y5MnYtGkTIiMjAQC7du1CYWEhfHx8kJKSAoZhsHPnzlblBIEn5MYTcLa8I1evXkVlZSUyMjLg6+sLHx8fZGRkoLKyEkuWLEFgYCBRIAUHByMxMdFlAblixQriJILz58+jpqYGAQEBxAihAMmuJRCvXr2Kfv36wWAwID8/X6x/fn6+WP/a2lpi/YVVJ1wVkOnp6ZII4rx581BRUSFuDMPg5MmTxPJZWVmSpQLbWh5Q7vssy3bovs+yrNv7PgD861//Evu+t0IFoJu5FY+A3SkQ9QpILQHjTP327t3bar8gEOU3SWftjqhFCJctW6ZqnzFjhqaAjI6OJs6Ui4+PR0xMjMt2hmEky+WR8NQAwGg0olOnTsT6p6amSgYI8gjhHXfcIXlET7Jr3URIicK1ogTCTZBk5zhONQoyYMAApKeno1OnTggPD8f06dPFG5R88CcXeI888gjeeOMNLFq0SLG8Vt8BoJiI+IcffsCKFSvEiQokgdevXz9cvnzZZQEZEhKCyspKxd9d8O84uJNHCGNiYjBjxgxNu5qAvHLlCp566ilkZGTAx8cHPj4+6NatG5566iksWbIEAQEBxPoHBQWhU6dOLgvIcePGtZoglp2djdzcXOTm5orZEQTkAm/w4MGSR5Ty8gzDYMGCBUT/Am3t+4JAJPV9ZwWkyWSSrPUsb9/s2bMl5788Qijv+yS7t/b93//+9xg3bpzm7+NJqAB0M3ojOFrl9T4idneEkWVZUTQpbQzDSGbhulI/tQijll2AFCF0xq4mEG+//XZiqgLgZpTAcaZZW+0sy6q+aLxr1y6kpqYSBVhaWhr27t2ruRSfqwIyMzNTtf21tbUYOXKkos0xSuCsXS4Q09LSEBERodr+xYsXEyOEmzZtUo0gPv3008SbSEREBBiGQV5eHrZt2yY+qnKMUJw+fZoo8PLz8+Hn54cJEyYolif1bZLAUhKYDQ0NRIF39epV3QLS8ZGYkn/SJA3HCKGzdi0BqVZ/JYF311134aeffnJZQBoMBkyaNEncJ39HrKSkpNUkA7lAnDx5MrF8YmIiOnfuLP4vF1gzZ87Ehg0biPbZs2fj559/bve+LwjE0tJSyes/8vZFRUURrw2u9H25QMzKysIf//hH1fa/9tprxAghqe8LdrW+P2rUKJjNZsnTKW+ECkA3o1cAtUd5vQJUbwSztrYWq1evVtxYlsWzzz7rtvpp2UmPoJ21C5AEot1uF99FVCIqKkr1MYGW/aWXXoLBYCAKHJvNhunTpxPL/+lPf4Kfn59qBDI+Pt5lARkXF6fafq36x8XFIS4uzukIoiMHDx5EYmIirFYr0b8gQElRAgE1O0kgMgyDYcOG4ejRoxKfzgo8nudRUVFBLK83gkhCEEg9e/bULSB37NjRZv8kAeusHVAWiKT6kwTelStXnBbASgLSaDRKygsTYgR69+4tyZAgF3gBAQGSHKTy8gsWLGg1CcFRYPn7+0veUZTbIyIiVAWSXgGZk5MjydMob19dXR2Sk5NV/Tuux64VQXTk4MGDGDBggOQdPHn7IyMjUV1d3e5931FAejtUALoZZwSQr6+vqsDz8fHRJaC8WUDeivrpeQTtbASUhNlslqxHKUdYy9dV+8CBAyU3ATlBQUEYOHAg0X733XerpvrQikBqCUiWZfHqq68Syw8YMEC1/nr9GwwG1QXptQSoswJY6Saxa9cuTJkyBf7+/sjKysKyZctw7tw5pwXee++9p1oe0BdBlNPeAlJYD1mPf8cIobMRREeBKMyUdqcAJglIi8WiKgJCQ0Mls0zlAs9sNqsOXnr16qU6CzU5ObnVKkZtiUDqFZC+vr6SCFh7C9jIyEhV/7GxsVizZg2x/c4IUGcEsJaA9GaoAHQzzgggvREyLQGkV4CyLOu2R7i3on566q83ApqYmKi4WoFAREQEcbkjZ+yhoaGIiYkh2k0mk2oEMTo6WlVgakUgtQRkSEgIsrKyiHat+uv1bzAYVCfhuFuApqWl4c0330RdXR369u0rrlJTU1ODt956S1PgATcjEErlhRRIrkYQAfcKSJZlUVhY6LJ/Z+xq/jt37gyr1eoxAWw2m1XfPxZy5JGIi4tTPTcDAgIk5eUCKzQ0VCIg2xqB1CsgGYbBvHnziPV3t4DlOA5z584ltt/dEVRBIHozVAC6Gb2TPJwprzeCpiXACgsLdQk0PQKrPeqnZdcrsNXqP336dKSnpyteCFpaWhAcHIyQkBCX7SzLoqioiOjfbrer5nIzGo2qaW60IpBaArKwsBBGo9Hl+uv1b7fbJcle5bhbgMrfcfziiy8wZ84cREZGwmKxYNiwYUSBpxRZdizPMAwSExNdjiDqfQQNqAs0vf6dEaBq/j0tgDMzM5GYmEg8900mk+oqPvn5+br6jjyJvRytCKReARkREYEuXboQ/btbwBqNRtWl7NwtQIVHzN4MFYBu5lbMAvb0O3Zq/oWNhLtnQbfHI2g9EcQzZ84gOjoadrsdlZWV2Lx5M+rr67F06VLY7XZEREQgMjLSZTvP8/jb3/5G9K91EzEajao3Ia0IpJaAfP/998EwjMv11+u/qKgILMt6TIAeOXJEMcpz7do1bNq0qVUaCiWBqMS1a9fw1FNPIS4uzuUIot5H0M4KJFf9a9m1/HtaAJ85cwaRkZGIi4tDVVWVeO5XVlbCbrfD19cXSUlJxHMzKSkJvr6+xPI8z6u+3hAZGSnJwydHKwKpV0AWFRWB4ziPCdj4+HiJQJTjbgH6yiuvqOaI9QaoAPQwzuRx01Pe3bOA3V1/vfVrj0fQeiKgQUFB8Pf3R35+vviCsJCpPz8/H8ePH8eJEydctk+aNEk1wpiSkgKr1UoUYFarFSkpKS5HILUE5IYNGxAbG+ty/fX6r6urExNOe0KAfvXVV6rJgEmQBKISrkYQTSaTWwWkEm3xryXgnPXvSQF87Ngx5OXltTr38/Ly0NDQoCoQo6Ki0NDQQCxfXFyMtLQ01b4THBzscgRSr4B89913wTCMxwSs3sGfXgH65Zdfqq4i5Q1QAfgbx5lcYGq4W0Dq5VZEAG9VBPTChQtobGxEQ0MDLly40OpYrti1IozR0dE4cOAAUYA1NjbqilBqCcj09HQxFYcr9W8P/8XFxR4ToBs2bECnTp2I9vamLRFEQWC6W0Dq9U+yO/sOpd76t4cAPn/+PBoaGtDQ0IDz58+L+9UEomM0Sam8VoQxPDwc4eHhLkcg9QrIdevWIS4uzmMCdtWqVeB53mMC9MiRI6pLCXoDVAD+xtEbgfut+9eLpyOg7YFWBFGAJDD1RCC1BGR0dLRkeSxX6t9e/j0lQJVy0bmbtkQQHXG3gNTrX+87lHrr314CWA5JIGqhJSD1RCD1Csi0tDSUlZUR2+duAZuWlobi4mKPCdB169ZJ8jR6I1QAUjo03h4BbQtaEUS95V0VkN7u39MC2Btxt4DU61/vO5R66+8uAewqWgLS1Qik3kfYWuf+rfLvKQEqCGBvhQpACkUH3hAB9Bb0ClBP+/e0AP5/wlUB6S3+PS2AbzXuEpDe7t/TAtjTUAFIoVAo7YCnBTDFe/C0AL7VuPoI21v8e1oAewoqACkUCoVCoVBcxNMC2FWoAKRQKBQKhULpYFABSKFQKBQKhdLBoAKQQqFQKBQKpYNBBSCFQqFQKBRKB4MKQAqFQqFQKJQOBhWAFAqFQqFQKB0MKgApFAqFQqFQOhhUAFIoFAqFQqF0MKgApFAoFAqFQulgUAFIoVAoFAqF0sGgApBCoVAoFAqlg0EFIIVCoVAoFEoHgwpACoVCoVAolA4GFYAUCoVCoVAoHYz/AhAO5Ngn9ISbAAAAAElFTkSuQmCC\" width=\"640\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{TGA: 0, TAA: 0.0, AGA: 0.03940886699507389, TTC: 0.02691749925018771}\n",
      "blue: empirical\n",
      "orange: base-wise/approximate\n"
     ]
    }
   ],
   "source": [
    "stop_positional_marginals = positional_base_usage(codon_usage_stop / stop_codon_number)\n",
    "parameters[\"base_frequency_marginals_stop_region\"] = format_matrix(stop_positional_marginals)\n",
    "\n",
    "def difference_of_approximation(empirical_close, empirical_far, approx_close, approx_far):\n",
    "    return (empirical_close - empirical_far) - (approx_close - approx_far)\n",
    "\n",
    "explicit = {\"TGA\": 0, \"TAA\": empirical_codon_frequencies_stop[codon2index(\"TAA\")]}\n",
    "\n",
    "# Finding another two codons to be included:\n",
    "basewise_codon_frequencies_stop = fit_codon_usage(empirical_codon_frequencies_stop)\n",
    "basewise_codon_frequencies_general = fit_codon_usage(empirical_codon_frequencies_general)\n",
    "\n",
    "# Plot: which codons are the most informative w.r.t. stop vs general:\n",
    "w = 0.3\n",
    "plt.subplot(3,1,1)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop - empirical_codon_frequencies_general, width=w)\n",
    "plt.bar(np.arange(64)+w/2, basewise_codon_frequencies_stop - basewise_codon_frequencies_general, width=w)\n",
    "plt.xticks(range(64), codon_list, rotation=90)\n",
    "\n",
    "# Plot before:\n",
    "plt.subplot(3,1,2)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop, width=w)\n",
    "plt.bar(np.arange(64)+w/2, basewise_codon_frequencies_stop, width=w)\n",
    "plt.xticks(range(64), [\"\"] * 64)\n",
    "\n",
    "# Corrections:\n",
    "diff = difference_of_approximation(empirical_codon_frequencies_stop, \n",
    "                                   empirical_codon_frequencies_general, \n",
    "                                   basewise_codon_frequencies_stop, \n",
    "                                   basewise_codon_frequencies_general)\n",
    "diff[codon2index(\"TGA\")] = 0\n",
    "diff[codon2index(\"TAA\")] = 0\n",
    "abs_diff = np.absolute(diff)\n",
    "\n",
    "most_prominent_deviant = codon_list[np.argmax(abs_diff)]\n",
    "explicit[most_prominent_deviant] = empirical_codon_frequencies_stop[np.argmax(abs_diff)]\n",
    "diff[np.argmax(abs_diff)] = 0\n",
    "\n",
    "corrected_sum = 0\n",
    "basewise_sum = 0\n",
    "for codon in explicit:\n",
    "    corrected_sum += explicit[codon]\n",
    "    basewise_sum += basewise_codon_frequencies_stop[codon2index(codon)]\n",
    "    # print(\"+\", codon, corrected_sum, basewise_sum)\n",
    "desired_deviation = basewise_sum - corrected_sum\n",
    "# print(desired_deviation)\n",
    "\n",
    "# Find compensator:\n",
    "diff = empirical_codon_frequencies_stop - basewise_codon_frequencies_stop\n",
    "candidates = sorted([c for c in codon_list if c not in [\"TGA\", \"TAA\", most_prominent_deviant]], \n",
    "                    key=lambda codon: abs(diff[codon2index(codon)] - desired_deviation))\n",
    "\n",
    "explicit[candidates[0]] = empirical_codon_frequencies_stop[codon2index(candidates[0])]\n",
    "corrected_sum = 0\n",
    "basewise_sum = 0\n",
    "for codon in explicit:\n",
    "    corrected_sum += explicit[codon]\n",
    "    basewise_sum += basewise_codon_frequencies_stop[codon2index(codon)]\n",
    "    # print(\"+\", codon, corrected_sum, basewise_sum)\n",
    "\n",
    "explicit[candidates[0]] += basewise_sum - corrected_sum\n",
    "\n",
    "parameters[\"explicit_codon_probabilities_stop_region\"] = \"{%s}\" % \", \".join([\"%s: %s\" % (k, explicit[k]) \n",
    "                                                                             for k in explicit])\n",
    "print(parameters[\"explicit_codon_probabilities_stop_region\"])\n",
    "\n",
    "\n",
    "plt.subplot(3,1,3)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop, width=w)\n",
    "plt.bar(np.arange(64)+w/2, [explicit[codon] if codon in explicit else basewise_codon_frequencies_stop[codon2index(codon)] \n",
    "                            for codon in codon_list], width=w)\n",
    "plt.xticks(range(64), codon_list, rotation=90)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "print(\"blue: empirical\")\n",
    "print(\"orange: base-wise/approximate\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intron parameters\n",
    " - collect **intron lengths**\n",
    " - Compute intron length median/mean (for breakpoint)\n",
    " - find min and max intron length; find 5% and 95% quantiles; extend from quantiles (subsequent AUGUSTUS can still concern itself with long introns)\n",
    " - Maybe derive a maximum-likelihood approach for inferring the parameters (would be good)\n",
    " - manually set SDS- and SAS-size? Or try sliding a categorical model fit to the intron-middle onto the edges, and stop where they deviate significantly?\n",
    " - collect **empirical base-frequencies at SDS, SAS, and in between**\n",
    " \n",
    " - **For Poisson** only need the mean (+ min and max for truncation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intron lengths: 16 17.806451612903224 26\n",
      "0.5344827586206896\n"
     ]
    }
   ],
   "source": [
    "# TODO: remove the unreliable annotation with the big introns!\n",
    "intron_lengths = [l for l in intron_lengths if l < 40]\n",
    "print(\"intron lengths:\", np.min(intron_lengths), np.mean(intron_lengths), np.max(intron_lengths))\n",
    "print(np.mean(intron_counts))\n",
    "\n",
    "parameters[\"intron_minimum_length\"] = np.min(intron_lengths) - hyperparameters[\"intron_length_tolerance\"]\n",
    "parameters[\"intron_maximum_length\"] = np.max(intron_lengths) + hyperparameters[\"intron_length_tolerance\"]\n",
    "parameters[\"intron_mean_length\"] = np.mean(intron_lengths)\n",
    "\n",
    "# # v # SEQUENCE information # v # #\n",
    "\n",
    "sds_frequencies = np.zeros((hyperparameters[\"sds_size\"], 4))\n",
    "sas_frequencies = np.zeros((hyperparameters[\"sas_size\"], 4))\n",
    "intron_base_frequencies = np.zeros(4)\n",
    "\n",
    "# ADD PSEUDOCOUNTS!\n",
    "sds_frequencies[2:,:] = 1\n",
    "sas_frequencies[:-2,:] = 1\n",
    "\n",
    "for l in introns_by_length:\n",
    "    for intron in introns_by_length[l]:\n",
    "        sds = intron[:hyperparameters[\"sds_size\"]]\n",
    "        for i in range(len(sds)):\n",
    "            sds_frequencies[i, base_order.index(sds[i])] += 1\n",
    "        \n",
    "        middle = intron[hyperparameters[\"sds_size\"]:-hyperparameters[\"sas_size\"]]\n",
    "        for i in range(len(middle)):\n",
    "            intron_base_frequencies[base_order.index(middle[i])] += 1\n",
    "            \n",
    "        sas = intron[-hyperparameters[\"sas_size\"]:]\n",
    "        for i in range(len(sas)):\n",
    "            sas_frequencies[i, base_order.index(sas[i])] += 1\n",
    "        # print(\"%s %s %s\" % (sds, middle, sas))\n",
    "        \n",
    "parameters[\"intron_base_frequencies_splice_donor_site\"] = format_matrix((sds_frequencies.T / np.sum(sds_frequencies, \n",
    "                                                                                                    axis=1)).T)\n",
    "parameters[\"intron_base_frequencies_splice_acceptor_site\"] = format_matrix((sas_frequencies.T / np.sum(sas_frequencies, \n",
    "                                                                                                       axis=1)).T)\n",
    "parameters[\"intron_base_frequencies\"] = \"{%s}\" % \", \".join([str(f) \n",
    "                                                            for f \n",
    "                                                            in intron_base_frequencies / np.sum(intron_base_frequencies)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Writing the parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_filename, \"w\") as out:\n",
    "    result = \"# automatically generated parameter file\\n\"\n",
    "    result += \"# from annotation file: \" + annotation_filename + \"\\n\"\n",
    "    result += \"# and sequence file: \" + sequence_filename + \"\\n\"\n",
    "    result += \"### Hyperparameters: \\n\"\n",
    "    for key in hyperparameters:\n",
    "        result += \"# %s: %s\\n\" % (key, hyperparameters[key])\n",
    "    \n",
    "    for key in parameters:\n",
    "        result += \"%s: %s\" % (key, parameters[key])\n",
    "        result += \"\\n\"\n",
    "    \n",
    "    out.write(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
