{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameter-file generation\n",
    "### Input: raw sequences (contigs.fasta)\n",
    "### Input: annotation (annotation.gff3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_filename = \"scaffolds.07-31.cleaned.gff\"\n",
    "sequence_filename = \"scaffolds.fasta\"\n",
    "output_filename = \"parameters.automatic.properties\"\n",
    "\n",
    "parameters = dict()\n",
    "parameters[\"start_region_size\"] = 6\n",
    "parameters[\"stop_region_size\"] = 24"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output: parameters.properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hyperparameters: To be chosen manually\n",
    " - `\"ncs_flank_size\"`: How many nt upstream of left and downstream of right end can be considered NCS?\n",
    " - `\"ncs_mean_length\"`: Manual estimate of how long an intergenic region is on average (in nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparameters = {}\n",
    "hyperparameters[\"ncs_flank_size\"] = 300\n",
    "hyperparameters[\"ncs_mean_length\"] = 1000\n",
    "hyperparameters[\"sds_size\"] = 5\n",
    "hyperparameters[\"sas_size\"] = 4\n",
    "\n",
    "hyperparameters[\"intron_length_tolerance\"] = 4 # how far to extend beyond the seen range of intron lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene format: CDS(begins with ATG) - intron - CDS - intron - CDS - stop_codon\n",
    "# i.e.: CDS[starts with ATG] - {intron - CDS}* - stop_codon\n",
    "# (thus, all genes have an even amount of features; start-codon is not treated separately)\n",
    "# The annotation may however also contain other features, especially (for me): three_prime_UTR \n",
    "# -> when reading the file, replace this! (might also have to ignore some features)\n",
    "\n",
    "complement = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\", \"N\": \"N\"}\n",
    "base_order = \"TCAG\"\n",
    "codon_list = [a + b + c for a in base_order for b in base_order for c in base_order]\n",
    "\n",
    "rna_codon_list = [c.replace(\"T\", \"U\") for c in codon_list]\n",
    "\n",
    "def codon2index(codon):\n",
    "    result = base_order.index(codon[0]) * 16\n",
    "    result += base_order.index(codon[1]) * 4\n",
    "    result += base_order.index(codon[2])\n",
    "    return result\n",
    "\n",
    "def positional_base_usage(fs):\n",
    "    positional_fs = np.zeros((3, 4))\n",
    "    for c in codon_list:\n",
    "        for i in range(3): # position in codon\n",
    "            positional_fs[i, base_order.index(c[i])] += fs[codon2index(c)]\n",
    "    return positional_fs\n",
    "\n",
    "# Simple approximation based on marginals\n",
    "def fit_codon_usage(counts):\n",
    "    positional_fs = positional_base_usage(counts)\n",
    "    independent_codon_approximation = np.zeros(64)\n",
    "    for c in codon_list:\n",
    "        independent_codon_approximation[codon2index(c)] = np.product([positional_fs[i, base_order.index(c[i])] for i in range(3)])\n",
    "    \n",
    "    return independent_codon_approximation\n",
    "\n",
    "def rev_comp(sequence: str):\n",
    "    result = \"\"\n",
    "    for i in range(len(sequence)-1, -1, -1):\n",
    "        result += complement[sequence[i]]\n",
    "    return result\n",
    "\n",
    "class Feature:\n",
    "    def __init__(self, feature: str, start: int, end: int, strand: bool):\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "        self.feature = feature\n",
    "    \n",
    "class Gene:\n",
    "    def __init__(self, strand: bool):\n",
    "        self.strand = strand\n",
    "        self.features = [] # a sorted (!) list of CDS, introns and one UTR; sorted from 5' to 3', \n",
    "        # i.e. on reverse strand, the indices of subsequent features;\n",
    "        # Sorting is trusted to the outside world; maybe check in isValid!\n",
    "        \n",
    "    def append(self, feature: Feature):\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.features)\n",
    "    \n",
    "    def __str__(self):\n",
    "        return \" \".join([\"%s%s %s %s%s\" % (\"[\" if f.strand else \"<\", \n",
    "                                           f.start, f.feature, f.end,\n",
    "                                           \">\" if f.strand else \"]\") \n",
    "                         for f in self.features])\n",
    "\n",
    "    def get_range(self):\n",
    "        return self.features[0].start, self.features[-1].end\n",
    "    \n",
    "    # sequence is nucleotide sequence from scaffolds.fasta\n",
    "    # this produces the concatenated cds -- potentially complementarily reversed (if strand is negative)\n",
    "    def get_contiguous_cds(self, sequence: str):\n",
    "        cds = \"\".join([sequence[(f.start-1):f.end] for f in self.features if f.feature == \"CDS\"])\n",
    "        if self.strand:\n",
    "            return cds\n",
    "        else:\n",
    "            return rev_comp(cds)\n",
    "        \n",
    "    def codon_usage(self, sequence: str):\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        general = np.zeros(64)\n",
    "        stop = np.zeros(64)\n",
    "        general_n = 0\n",
    "        for i in range(0, len(cds) - 21, 3):\n",
    "            general[codon2index(cds[i:i+3])] += 1\n",
    "            general_n += 1\n",
    "            \n",
    "        for i in range(len(cds) - 21, len(cds), 3):\n",
    "            stop[codon2index(cds[i:i+3])] += 1\n",
    "            \n",
    "        print(\"term: %s : %s*UGA\" % (cds[len(cds)-21:], stop[codon2index(\"TGA\")]))\n",
    "        return general, general_n, stop\n",
    "        \n",
    "    def count_UGA(self, sequence: str):\n",
    "        inframe = 0\n",
    "        outframe_in_cds = 0\n",
    "        cds = self.get_contiguous_cds(sequence)\n",
    "        for i in range(len(cds)-3):\n",
    "            if sequence[i:i+3] == \"TGA\":\n",
    "                if i % 3 == 0:\n",
    "                    inframe += 1\n",
    "                else:\n",
    "                    outframe_in_cds += 1\n",
    "        # REQUIRES this gene be valid\n",
    "        # Pure sanity check\n",
    "        if self.strand:\n",
    "            stop = sequence[(self.features[-1].start-1):self.features[-1].end].count(\"TGA\")\n",
    "        else:\n",
    "            stop = sequence[(self.features[0].start-1):self.features[0].end].count(\"TCA\")\n",
    "        \n",
    "        \n",
    "        return inframe, outframe_in_cds, stop\n",
    "        \n",
    "    def get_exon_intron_stats(self):\n",
    "        exon_lengths = []\n",
    "        intron_lengths = []\n",
    "        for f in self.features:\n",
    "            if f.feature == \"CDS\":\n",
    "                exon_lengths += [f.end - f.start + 1] # +1 bc. gff is inclusive-end\n",
    "            elif f.feature == \"intron\":\n",
    "                intron_lengths += [f.end - f.start + 1]\n",
    "        \n",
    "        return exon_lengths, intron_lengths\n",
    "        \n",
    "    def is_valid(self):\n",
    "        if len(self.features) == 0:\n",
    "            print(\"! The gene is empty\")\n",
    "            return False\n",
    "        \n",
    "        stop_index = -1 if self.strand else 0\n",
    "        first_cds_index = 0 if self.strand else -1\n",
    "        \n",
    "        # a Gene must end in a stop_codon\n",
    "        if self.features[stop_index].feature != \"stop_codon\":\n",
    "            print(\"! 3' Terminal feature is not a stop_codon, but a %s\" % self.features[stop_index].feature)\n",
    "            return False\n",
    "        # a Gene must start with a CDS\n",
    "        if self.features[first_cds_index].feature != \"CDS\":\n",
    "            print(\"! 5' terminal feature is not a CDS, but a %s\" % self.features[first_cds_index].feature)\n",
    "            return False\n",
    "        \n",
    "        # Gene must start with CDS, end with UTR (len >= 2 so far), and can have [intron, CDS]-pairs in between \n",
    "        # (no two cds adjacent, nor two introns)\n",
    "        if len(self.features) % 2 != 0:\n",
    "            print(\"! uneven number of features\")\n",
    "            return False\n",
    "        \n",
    "        num_stops = 0\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].feature == \"stop_codon\":\n",
    "                num_stops += 1\n",
    "            if self.features[index+1].feature == \"stop_codon\":\n",
    "                num_stops += 1\n",
    "                \n",
    "            if self.features[index].strand != self.strand or self.features[index+1].strand != self.strand:\n",
    "                print(\"! Strand disagreement within the gene\")\n",
    "                return False\n",
    "            \n",
    "            if self.strand:\n",
    "                if not (self.features[index].feature == \"CDS\" \\\n",
    "                        and self.features[index+1].feature in [\"intron\", \"stop_codon\"]):\n",
    "                    print(\"! CDS-intron-CDS-stop_codon-pattern violated\")\n",
    "                    return False\n",
    "            else:\n",
    "                if not (self.features[index+1].feature == \"CDS\" \\\n",
    "                        and self.features[index].feature in [\"intron\", \"stop_codon\"]):\n",
    "                    print(\"! CDS-intron-CDS-stop_codon-pattern violated\")\n",
    "                    return False\n",
    "            \n",
    "            index +=2\n",
    "            \n",
    "        if num_stops != 1:\n",
    "            print(\"! Too many stops: %s\" % num_stops)\n",
    "            return False   \n",
    "        \n",
    "        return True\n",
    "\n",
    "class Annotation:\n",
    "    def __init__(self, name: str, start: int, end: int):\n",
    "        self.name = name\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.features = [] # The list of features; could also separate this by strand\n",
    "        self.genes = []\n",
    "    \n",
    "    def add(self, feature: Feature):\n",
    "        index = 0\n",
    "        while index < len(self.features):\n",
    "            if self.features[index].start > feature.end:\n",
    "                self.features.insert(index, feature)\n",
    "                return\n",
    "            index += 1\n",
    "        self.features.append(feature)\n",
    "        \n",
    "    def compile_genes(self) -> bool:\n",
    "        all_valid = True\n",
    "        self.genes = []\n",
    "        last_end = self.features[0].start - 1\n",
    "        current_gene = Gene(self.features[0].strand)\n",
    "        for f in self.features:\n",
    "            if f.start != last_end + 1: # there is a break\n",
    "                self.genes.append(current_gene)\n",
    "                all_valid = all_valid and current_gene.is_valid()\n",
    "                current_gene = Gene(f.strand)\n",
    "                \n",
    "            current_gene.append(f)\n",
    "            last_end = f.end\n",
    "        \n",
    "        if len(current_gene) > 0:\n",
    "            self.genes.append(current_gene)\n",
    "            all_valid = all_valid and current_gene.is_valid()\n",
    "        \n",
    "        return all_valid\n",
    "        \n",
    "        \n",
    "    def __str__(self):\n",
    "        return self.name + \":\\n\\t\" + (\"\\n\\t\".join([\"[%s, %s]%s%s\" % (f.start, f.end,\n",
    "                                                                     f.feature, \n",
    "                                                                     \"+\" if f.strand else \"-\") \n",
    "                                               for f in self.features]))\n",
    "    \n",
    "def format_matrix(matrix):\n",
    "    return \"{{%s}}\" % \"}\\\\\\n  {\".join([\", \".join([str(e) for e in row]) for row in matrix])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input: Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1_length_428989_cov_1.681242:\n",
      "\t[61087, 62502]CDS+\n",
      "\t[62503, 62505]stop_codon+\n",
      "\t[191281, 191283]stop_codon-\n",
      "\t[191284, 191307]CDS-\n",
      "\t[191308, 191324]intron-\n",
      "\t[191325, 192602]CDS-\n",
      "\t[276921, 278189]CDS+\n",
      "\t[278190, 278206]intron+\n",
      "\t[278207, 278707]CDS+\n",
      "\t[278708, 278710]stop_codon+\n",
      "\t[316584, 317810]CDS+\n",
      "\t[317811, 317813]stop_codon+\n",
      "\t[408260, 409213]CDS+\n",
      "\t[409214, 409216]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[61087 CDS 62502> [62503 stop_codon 62505>\n",
      "\t<191281 stop_codon 191283] <191284 CDS 191307] <191308 intron 191324] <191325 CDS 192602]\n",
      "\t[276921 CDS 278189> [278190 intron 278206> [278207 CDS 278707> [278708 stop_codon 278710>\n",
      "\t[316584 CDS 317810> [317811 stop_codon 317813>\n",
      "\t[408260 CDS 409213> [409214 stop_codon 409216>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_6_length_313002_cov_1.859273:\n",
      "\t[8843, 8845]stop_codon-\n",
      "\t[8846, 9832]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<8843 stop_codon 8845] <8846 CDS 9832]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_7_length_300196_cov_1.642899:\n",
      "\t[199239, 200060]CDS+\n",
      "\t[200061, 200076]intron+\n",
      "\t[200077, 200361]CDS+\n",
      "\t[200362, 200364]stop_codon+\n",
      "\t[289779, 289781]stop_codon-\n",
      "\t[289782, 290088]CDS-\n",
      "\t[290089, 290107]intron-\n",
      "\t[290108, 292038]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[199239 CDS 200060> [200061 intron 200076> [200077 CDS 200361> [200362 stop_codon 200364>\n",
      "\t<289779 stop_codon 289781] <289782 CDS 290088] <290089 intron 290107] <290108 CDS 292038]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_10_length_283447_cov_1.644007:\n",
      "\t[157388, 158140]CDS+\n",
      "\t[158141, 158157]intron+\n",
      "\t[158158, 158691]CDS+\n",
      "\t[158692, 158694]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[157388 CDS 158140> [158141 intron 158157> [158158 CDS 158691> [158692 stop_codon 158694>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_12_length_280749_cov_1.673450:\n",
      "\t[184887, 186518]CDS+\n",
      "\t[186519, 186521]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[184887 CDS 186518> [186519 stop_codon 186521>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_18_length_257280_cov_1.854729:\n",
      "\t[13286, 14369]CDS+\n",
      "\t[14370, 14388]intron+\n",
      "\t[14389, 14483]CDS+\n",
      "\t[14484, 14486]stop_codon+\n",
      "\t[194806, 196164]CDS+\n",
      "\t[196165, 196167]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[13286 CDS 14369> [14370 intron 14388> [14389 CDS 14483> [14484 stop_codon 14486>\n",
      "\t[194806 CDS 196164> [196165 stop_codon 196167>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_19_length_257256_cov_1.830618:\n",
      "\t[241310, 241312]stop_codon-\n",
      "\t[241313, 241553]CDS-\n",
      "\t[241554, 241572]intron-\n",
      "\t[241573, 243329]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<241310 stop_codon 241312] <241313 CDS 241553] <241554 intron 241572] <241573 CDS 243329]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_20_length_254023_cov_1.905639:\n",
      "\t[29075, 29077]stop_codon-\n",
      "\t[29078, 29318]CDS-\n",
      "\t[29319, 29337]intron-\n",
      "\t[29338, 31088]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<29075 stop_codon 29077] <29078 CDS 29318] <29319 intron 29337] <29338 CDS 31088]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_21_length_245355_cov_1.707925:\n",
      "\t[30162, 30164]stop_codon-\n",
      "\t[30165, 31301]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<30162 stop_codon 30164] <30165 CDS 31301]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_22_length_243293_cov_1.665932:\n",
      "\t[208057, 209688]CDS+\n",
      "\t[209689, 209691]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[208057 CDS 209688> [209689 stop_codon 209691>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_23_length_242384_cov_1.872846:\n",
      "\t[86935, 86937]stop_codon-\n",
      "\t[86938, 88263]CDS-\n",
      "\t[88264, 88280]intron-\n",
      "\t[88281, 88670]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<86935 stop_codon 86937] <86938 CDS 88263] <88264 intron 88280] <88281 CDS 88670]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_25_length_237540_cov_1.816270:\n",
      "\t[165820, 165822]stop_codon-\n",
      "\t[165823, 166064]CDS-\n",
      "\t[166065, 166084]intron-\n",
      "\t[166085, 166688]CDS-\n",
      "\t[166689, 166707]intron-\n",
      "\t[166708, 167631]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<165820 stop_codon 165822] <165823 CDS 166064] <166065 intron 166084] <166085 CDS 166688] <166689 intron 166707] <166708 CDS 167631]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_27_length_235310_cov_1.764511:\n",
      "\t[52181, 52183]stop_codon-\n",
      "\t[52184, 53362]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<52181 stop_codon 52183] <52184 CDS 53362]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_29_length_230800_cov_1.889965:\n",
      "\t[154753, 155553]CDS+\n",
      "\t[155554, 155570]intron+\n",
      "\t[155571, 155990]CDS+\n",
      "\t[155991, 155993]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[154753 CDS 155553> [155554 intron 155570> [155571 CDS 155990> [155991 stop_codon 155993>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_31_length_222000_cov_1.724356:\n",
      "\t[143901, 143903]stop_codon-\n",
      "\t[143904, 144003]CDS-\n",
      "\t[144004, 144020]intron-\n",
      "\t[144021, 146113]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<143901 stop_codon 143903] <143904 CDS 144003] <144004 intron 144020] <144021 CDS 146113]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_38_length_212895_cov_1.850617:\n",
      "\t[193965, 193967]stop_codon-\n",
      "\t[193968, 195989]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<193965 stop_codon 193967] <193968 CDS 195989]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_39_length_211856_cov_1.789892:\n",
      "\t[68426, 70051]CDS+\n",
      "\t[70052, 70054]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[68426 CDS 70051> [70052 stop_codon 70054>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_42_length_207565_cov_1.992215:\n",
      "\t[199545, 199547]stop_codon-\n",
      "\t[199548, 200726]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<199545 stop_codon 199547] <199548 CDS 200726]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_44_length_205421_cov_1.892588:\n",
      "\t[137353, 138192]CDS+\n",
      "\t[138193, 138211]intron+\n",
      "\t[138212, 138751]CDS+\n",
      "\t[138752, 138754]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[137353 CDS 138192> [138193 intron 138211> [138212 CDS 138751> [138752 stop_codon 138754>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_54_length_192099_cov_1.873872:\n",
      "\t[162604, 164472]CDS+\n",
      "\t[164473, 164475]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[162604 CDS 164472> [164473 stop_codon 164475>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_64_length_178741_cov_1.661034:\n",
      "\t[129866, 131170]CDS+\n",
      "\t[131171, 131173]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[129866 CDS 131170> [131171 stop_codon 131173>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_65_length_177181_cov_1.879014:\n",
      "\t[59358, 59360]stop_codon-\n",
      "\t[59361, 59484]CDS-\n",
      "\t[59485, 59501]intron-\n",
      "\t[59502, 60271]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<59358 stop_codon 59360] <59361 CDS 59484] <59485 intron 59501] <59502 CDS 60271]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_69_length_175562_cov_1.893300:\n",
      "\t[4520, 6490]CDS+\n",
      "\t[6491, 6493]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[4520 CDS 6490> [6491 stop_codon 6493>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_70_length_174851_cov_1.725001:\n",
      "\t[20807, 27142]CDS+\n",
      "\t[27143, 27145]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[20807 CDS 27142> [27143 stop_codon 27145>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_71_length_174712_cov_2.292786:\n",
      "\t[151652, 151654]stop_codon-\n",
      "\t[151655, 153922]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<151652 stop_codon 151654] <151655 CDS 153922]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_75_length_172946_cov_1.727212:\n",
      "\t[156122, 157861]CDS+\n",
      "\t[157862, 157864]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[156122 CDS 157861> [157862 stop_codon 157864>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_78_length_168807_cov_1.849484:\n",
      "\t[4233, 4235]stop_codon-\n",
      "\t[4236, 4308]CDS-\n",
      "\t[4309, 4324]intron-\n",
      "\t[4325, 5184]CDS-\n",
      "\t[104343, 104345]stop_codon-\n",
      "\t[104346, 105779]CDS-\n",
      "\t[114778, 114780]stop_codon-\n",
      "\t[114781, 116184]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<4233 stop_codon 4235] <4236 CDS 4308] <4309 intron 4324] <4325 CDS 5184]\n",
      "\t<104343 stop_codon 104345] <104346 CDS 105779]\n",
      "\t<114778 stop_codon 114780] <114781 CDS 116184]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_80_length_167342_cov_1.702640:\n",
      "\t[3749, 4789]CDS+\n",
      "\t[4790, 4792]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[3749 CDS 4789> [4790 stop_codon 4792>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_84_length_162193_cov_1.961824:\n",
      "\t[18322, 18324]stop_codon-\n",
      "\t[18325, 19683]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<18322 stop_codon 18324] <18325 CDS 19683]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_90_length_157735_cov_1.661280:\n",
      "\t[90564, 91550]CDS+\n",
      "\t[91551, 91553]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[90564 CDS 91550> [91551 stop_codon 91553>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_91_length_156069_cov_1.747919:\n",
      "\t[105405, 105809]CDS+\n",
      "\t[105810, 105828]intron+\n",
      "\t[105829, 107154]CDS+\n",
      "\t[107155, 107157]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[105405 CDS 105809> [105810 intron 105828> [105829 CDS 107154> [107155 stop_codon 107157>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_103_length_148054_cov_1.718544:\n",
      "\t[45719, 47784]CDS+\n",
      "\t[47785, 47804]intron+\n",
      "\t[47805, 47907]CDS+\n",
      "\t[47908, 47910]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[45719 CDS 47784> [47785 intron 47804> [47805 CDS 47907> [47908 stop_codon 47910>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_104_length_148052_cov_1.912118:\n",
      "\t[42293, 43741]CDS+\n",
      "\t[43742, 43744]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[42293 CDS 43741> [43742 stop_codon 43744>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_114_length_139872_cov_1.858449:\n",
      "\t[110525, 110527]stop_codon-\n",
      "\t[110528, 112402]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<110525 stop_codon 110527] <110528 CDS 112402]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_127_length_132101_cov_1.744328:\n",
      "\t[31550, 31552]stop_codon-\n",
      "\t[31553, 32003]CDS-\n",
      "\t[32004, 32021]intron-\n",
      "\t[32022, 33388]CDS-\n",
      "\t[124200, 127031]CDS+\n",
      "\t[127032, 127034]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<31550 stop_codon 31552] <31553 CDS 32003] <32004 intron 32021] <32022 CDS 33388]\n",
      "\t[124200 CDS 127031> [127032 stop_codon 127034>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_132_length_130847_cov_1.874350:\n",
      "\t[105430, 106824]CDS+\n",
      "\t[106825, 106827]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[105430 CDS 106824> [106825 stop_codon 106827>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_134_length_130250_cov_1.655841:\n",
      "\t[40175, 40177]stop_codon-\n",
      "\t[40178, 41827]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<40175 stop_codon 40177] <40178 CDS 41827]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_135_length_129955_cov_1.831608:\n",
      "\t[129029, 129031]stop_codon-\n",
      "\t[129032, 129114]CDS-\n",
      "\t[129115, 129130]intron-\n",
      "\t[129131, 129905]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<129029 stop_codon 129031] <129032 CDS 129114] <129115 intron 129130] <129131 CDS 129905]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_145_length_125098_cov_1.669611:\n",
      "\t[98689, 98691]stop_codon-\n",
      "\t[98692, 98995]CDS-\n",
      "\t[98996, 99012]intron-\n",
      "\t[99013, 100943]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<98689 stop_codon 98691] <98692 CDS 98995] <98996 intron 99012] <99013 CDS 100943]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_150_length_122897_cov_1.750631:\n",
      "\t[69429, 70823]CDS+\n",
      "\t[70824, 70826]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[69429 CDS 70823> [70824 stop_codon 70826>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_153_length_122011_cov_1.703448:\n",
      "\t[68948, 69741]CDS+\n",
      "\t[69742, 69757]intron+\n",
      "\t[69758, 69884]CDS+\n",
      "\t[69885, 69887]stop_codon+\n",
      "\t[96588, 97775]CDS+\n",
      "\t[97776, 97778]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[68948 CDS 69741> [69742 intron 69757> [69758 CDS 69884> [69885 stop_codon 69887>\n",
      "\t[96588 CDS 97775> [97776 stop_codon 97778>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_155_length_121415_cov_1.724095:\n",
      "\t[114893, 114895]stop_codon-\n",
      "\t[114896, 116134]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<114893 stop_codon 114895] <114896 CDS 116134]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_159_length_120583_cov_1.600319:\n",
      "\t[114198, 114200]stop_codon-\n",
      "\t[114201, 116495]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<114198 stop_codon 114200] <114201 CDS 116495]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_160_length_120340_cov_1.623793:\n",
      "\t[66695, 66697]stop_codon-\n",
      "\t[66698, 68068]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<66695 stop_codon 66697] <66698 CDS 68068]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_164_length_119121_cov_1.871162:\n",
      "\t[33232, 33234]stop_codon-\n",
      "\t[33235, 34878]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<33232 stop_codon 33234] <33235 CDS 34878]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_176_length_115965_cov_1.774081:\n",
      "\t[53967, 53969]stop_codon-\n",
      "\t[53970, 54069]CDS-\n",
      "\t[54070, 54085]intron-\n",
      "\t[54086, 54184]CDS-\n",
      "\t[54185, 54200]intron-\n",
      "\t[54201, 56140]CDS-\n",
      "\t[56141, 56156]intron-\n",
      "\t[56157, 56276]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<53967 stop_codon 53969] <53970 CDS 54069] <54070 intron 54085] <54086 CDS 54184] <54185 intron 54200] <54201 CDS 56140] <56141 intron 56156] <56157 CDS 56276]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_177_length_115532_cov_1.858698:\n",
      "\t[22322, 23134]CDS+\n",
      "\t[23135, 23137]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[22322 CDS 23134> [23135 stop_codon 23137>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_181_length_114511_cov_1.615750:\n",
      "\t[36863, 38884]CDS+\n",
      "\t[38885, 38887]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[36863 CDS 38884> [38885 stop_codon 38887>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_202_length_107880_cov_1.614990:\n",
      "\t[12750, 12752]stop_codon-\n",
      "\t[12753, 14624]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<12750 stop_codon 12752] <12753 CDS 14624]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_206_length_105986_cov_1.546198:\n",
      "\t[53349, 54962]CDS+\n",
      "\t[54963, 54965]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[53349 CDS 54962> [54963 stop_codon 54965>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_207_length_105357_cov_1.894403:\n",
      "\t[82927, 84537]CDS+\n",
      "\t[84538, 84540]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[82927 CDS 84537> [84538 stop_codon 84540>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_210_length_104547_cov_1.706570:\n",
      "\t[19790, 20587]CDS+\n",
      "\t[20588, 20603]intron+\n",
      "\t[20604, 21023]CDS+\n",
      "\t[21024, 21026]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[19790 CDS 20587> [20588 intron 20603> [20604 CDS 21023> [21024 stop_codon 21026>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_217_length_103446_cov_1.848866:\n",
      "\t[27795, 29282]CDS+\n",
      "\t[29283, 29285]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[27795 CDS 29282> [29283 stop_codon 29285>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_222_length_101554_cov_1.892070:\n",
      "\t[66796, 66798]stop_codon-\n",
      "\t[66799, 66871]CDS-\n",
      "\t[66872, 66887]intron-\n",
      "\t[66888, 67747]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<66796 stop_codon 66798] <66799 CDS 66871] <66872 intron 66887] <66888 CDS 67747]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_227_length_100883_cov_2.022033:\n",
      "\t[11622, 11624]stop_codon-\n",
      "\t[11625, 11707]CDS-\n",
      "\t[11708, 11723]intron-\n",
      "\t[11724, 13119]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<11622 stop_codon 11624] <11625 CDS 11707] <11708 intron 11723] <11724 CDS 13119]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_244_length_98187_cov_1.698399:\n",
      "\t[62237, 62239]stop_codon-\n",
      "\t[62240, 62480]CDS-\n",
      "\t[62481, 62500]intron-\n",
      "\t[62501, 64254]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<62237 stop_codon 62239] <62240 CDS 62480] <62481 intron 62500] <62501 CDS 64254]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_246_length_97678_cov_1.774323:\n",
      "\t[74378, 74380]stop_codon-\n",
      "\t[74381, 74687]CDS-\n",
      "\t[74688, 74706]intron-\n",
      "\t[74707, 76637]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<74378 stop_codon 74380] <74381 CDS 74687] <74688 intron 74706] <74707 CDS 76637]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_250_length_96914_cov_1.797266:\n",
      "\t[42022, 42024]stop_codon-\n",
      "\t[42025, 43212]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<42022 stop_codon 42024] <42025 CDS 43212]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_259_length_95643_cov_1.674882:\n",
      "\t[51629, 56044]CDS+\n",
      "\t[56045, 56065]intron+\n",
      "\t[56066, 56614]CDS+\n",
      "\t[56615, 56617]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[51629 CDS 56044> [56045 intron 56065> [56066 CDS 56614> [56615 stop_codon 56617>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_262_length_95179_cov_1.968112:\n",
      "\t[12581, 13282]CDS+\n",
      "\t[13283, 13299]intron+\n",
      "\t[13300, 13670]CDS+\n",
      "\t[13671, 13690]intron+\n",
      "\t[13691, 14537]CDS+\n",
      "\t[14538, 14555]intron+\n",
      "\t[14556, 14819]CDS+\n",
      "\t[14820, 14822]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[12581 CDS 13282> [13283 intron 13299> [13300 CDS 13670> [13671 intron 13690> [13691 CDS 14537> [14538 intron 14555> [14556 CDS 14819> [14820 stop_codon 14822>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_264_length_94820_cov_1.815794:\n",
      "\t[42211, 42213]stop_codon-\n",
      "\t[42214, 42942]CDS-\n",
      "\t[42943, 42958]intron-\n",
      "\t[42959, 43399]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<42211 stop_codon 42213] <42214 CDS 42942] <42943 intron 42958] <42959 CDS 43399]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_265_length_94667_cov_1.884356:\n",
      "\t[19307, 19342]CDS+\n",
      "\t[19343, 19362]intron+\n",
      "\t[19363, 19789]CDS+\n",
      "\t[19790, 19792]stop_codon+\n",
      "\t[39745, 40593]CDS+\n",
      "\t[40594, 40613]intron+\n",
      "\t[40614, 40685]CDS+\n",
      "\t[40686, 40688]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[19307 CDS 19342> [19343 intron 19362> [19363 CDS 19789> [19790 stop_codon 19792>\n",
      "\t[39745 CDS 40593> [40594 intron 40613> [40614 CDS 40685> [40686 stop_codon 40688>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_267_length_94345_cov_1.719173:\n",
      "\t[12966, 12968]stop_codon-\n",
      "\t[12969, 14294]CDS-\n",
      "\t[14295, 14312]intron-\n",
      "\t[14313, 14717]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<12966 stop_codon 12968] <12969 CDS 14294] <14295 intron 14312] <14313 CDS 14717]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_271_length_94123_cov_1.840504:\n",
      "\t[23799, 24479]CDS+\n",
      "\t[24480, 24496]intron+\n",
      "\t[24497, 25762]CDS+\n",
      "\t[25763, 25765]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[23799 CDS 24479> [24480 intron 24496> [24497 CDS 25762> [25763 stop_codon 25765>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_273_length_93569_cov_1.730303:\n",
      "\t[65637, 65639]stop_codon-\n",
      "\t[65640, 71990]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<65637 stop_codon 65639] <65640 CDS 71990]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_278_length_92397_cov_1.658773:\n",
      "\t[50676, 51629]CDS+\n",
      "\t[51630, 51632]stop_codon+\n",
      "\t[76967, 78361]CDS+\n",
      "\t[78362, 78364]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[50676 CDS 51629> [51630 stop_codon 51632>\n",
      "\t[76967 CDS 78361> [78362 stop_codon 78364>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_279_length_92227_cov_1.518132:\n",
      "\t[19650, 19652]stop_codon-\n",
      "\t[19653, 20615]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<19650 stop_codon 19652] <19653 CDS 20615]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_296_length_88345_cov_1.720182:\n",
      "\t[29944, 29946]stop_codon-\n",
      "\t[29947, 30016]CDS-\n",
      "\t[30017, 30033]intron-\n",
      "\t[30034, 30890]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<29944 stop_codon 29946] <29947 CDS 30016] <30017 intron 30033] <30034 CDS 30890]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_297_length_88055_cov_1.968429:\n",
      "\t[53198, 55245]CDS+\n",
      "\t[55246, 55265]intron+\n",
      "\t[55266, 55368]CDS+\n",
      "\t[55369, 55371]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[53198 CDS 55245> [55246 intron 55265> [55266 CDS 55368> [55369 stop_codon 55371>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_301_length_87574_cov_1.953023:\n",
      "\t[63192, 63194]stop_codon-\n",
      "\t[63195, 65252]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<63192 stop_codon 63194] <63195 CDS 65252]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_322_length_82363_cov_1.603434:\n",
      "\t[57959, 61450]CDS+\n",
      "\t[61451, 61453]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[57959 CDS 61450> [61451 stop_codon 61453>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_324_length_81694_cov_1.607292:\n",
      "\t[58740, 59527]CDS+\n",
      "\t[59528, 59543]intron+\n",
      "\t[59544, 59667]CDS+\n",
      "\t[59668, 59670]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[58740 CDS 59527> [59528 intron 59543> [59544 CDS 59667> [59668 stop_codon 59670>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_336_length_80118_cov_1.672351:\n",
      "\t[63703, 63705]stop_codon-\n",
      "\t[63706, 67188]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<63703 stop_codon 63705] <63706 CDS 67188]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_339_length_79629_cov_1.996868:\n",
      "\t[39147, 39149]stop_codon-\n",
      "\t[39150, 41729]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<39147 stop_codon 39149] <39150 CDS 41729]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_344_length_79198_cov_1.698916:\n",
      "\t[44241, 44243]stop_codon-\n",
      "\t[44244, 44322]CDS-\n",
      "\t[44323, 44338]intron-\n",
      "\t[44339, 44737]CDS-\n",
      "\t[44738, 44754]intron-\n",
      "\t[44755, 46973]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<44241 stop_codon 44243] <44244 CDS 44322] <44323 intron 44338] <44339 CDS 44737] <44738 intron 44754] <44755 CDS 46973]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_345_length_79091_cov_2.091434:\n",
      "\t[71411, 73164]CDS+\n",
      "\t[73165, 73183]intron+\n",
      "\t[73184, 73424]CDS+\n",
      "\t[73425, 73427]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[71411 CDS 73164> [73165 intron 73183> [73184 CDS 73424> [73425 stop_codon 73427>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_358_length_77800_cov_1.675254:\n",
      "\t[26268, 30059]CDS+\n",
      "\t[30060, 30062]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[26268 CDS 30059> [30060 stop_codon 30062>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_364_length_77330_cov_3.110864:\n",
      "\t[47325, 47327]stop_codon-\n",
      "\t[47328, 49928]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<47325 stop_codon 47327] <47328 CDS 49928]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_376_length_75967_cov_1.666139:\n",
      "\t[44450, 44452]stop_codon-\n",
      "\t[44453, 45397]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<44450 stop_codon 44452] <44453 CDS 45397]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_379_length_75420_cov_1.946927:\n",
      "\t[25306, 25308]stop_codon-\n",
      "\t[25309, 26295]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25306 stop_codon 25308] <25309 CDS 26295]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_386_length_74730_cov_1.852566:\n",
      "\t[31588, 31597]CDS+\n",
      "\t[31598, 31614]intron+\n",
      "\t[31615, 32789]CDS+\n",
      "\t[32790, 32792]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[31588 CDS 31597> [31598 intron 31614> [31615 CDS 32789> [32790 stop_codon 32792>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_389_length_74397_cov_1.930066:\n",
      "\t[21604, 21606]stop_codon-\n",
      "\t[21607, 22785]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<21604 stop_codon 21606] <21607 CDS 22785]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_410_length_72732_cov_1.702100:\n",
      "\t[25650, 25652]stop_codon-\n",
      "\t[25653, 25712]CDS-\n",
      "\t[25713, 25730]intron-\n",
      "\t[25731, 25868]CDS-\n",
      "\t[25869, 25884]intron-\n",
      "\t[25885, 26268]CDS-\n",
      "\t[26269, 26287]intron-\n",
      "\t[26288, 26365]CDS-\n",
      "\t[26366, 26384]intron-\n",
      "\t[26385, 27732]CDS-\n",
      "\t[27733, 27751]intron-\n",
      "\t[27752, 27816]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25650 stop_codon 25652] <25653 CDS 25712] <25713 intron 25730] <25731 CDS 25868] <25869 intron 25884] <25885 CDS 26268] <26269 intron 26287] <26288 CDS 26365] <26366 intron 26384] <26385 CDS 27732] <27733 intron 27751] <27752 CDS 27816]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_426_length_70982_cov_1.681053:\n",
      "\t[24670, 24672]stop_codon-\n",
      "\t[24673, 25131]CDS-\n",
      "\t[25132, 25147]intron-\n",
      "\t[25148, 26461]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<24670 stop_codon 24672] <24673 CDS 25131] <25132 intron 25147] <25148 CDS 26461]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_427_length_70843_cov_2.049663:\n",
      "\t[36168, 37112]CDS+\n",
      "\t[37113, 37115]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[36168 CDS 37112> [37113 stop_codon 37115>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_450_length_68074_cov_1.636967:\n",
      "\t[53953, 57489]CDS+\n",
      "\t[57490, 57492]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[53953 CDS 57489> [57490 stop_codon 57492>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_455_length_67931_cov_1.705902:\n",
      "\t[2074, 2076]stop_codon-\n",
      "\t[2077, 2241]CDS-\n",
      "\t[2242, 2258]intron-\n",
      "\t[2259, 2660]CDS-\n",
      "\t[2661, 2678]intron-\n",
      "\t[2679, 14747]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2074 stop_codon 2076] <2077 CDS 2241] <2242 intron 2258] <2259 CDS 2660] <2661 intron 2678] <2679 CDS 14747]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_469_length_66432_cov_1.646482:\n",
      "\t[39237, 40199]CDS+\n",
      "\t[40200, 40215]intron+\n",
      "\t[40216, 40458]CDS+\n",
      "\t[40459, 40461]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[39237 CDS 40199> [40200 intron 40215> [40216 CDS 40458> [40459 stop_codon 40461>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_476_length_65419_cov_1.711113:\n",
      "\t[26840, 28468]CDS+\n",
      "\t[28469, 28471]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[26840 CDS 28468> [28469 stop_codon 28471>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_479_length_65262_cov_1.758118:\n",
      "\t[58933, 58935]stop_codon-\n",
      "\t[58936, 59031]CDS-\n",
      "\t[59032, 59048]intron-\n",
      "\t[59049, 59153]CDS-\n",
      "\t[59154, 59171]intron-\n",
      "\t[59172, 61685]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<58933 stop_codon 58935] <58936 CDS 59031] <59032 intron 59048] <59049 CDS 59153] <59154 intron 59171] <59172 CDS 61685]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_485_length_64803_cov_1.657740:\n",
      "\t[36500, 36502]stop_codon-\n",
      "\t[36503, 38739]CDS-\n",
      "\t[38740, 38758]intron-\n",
      "\t[38759, 39275]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<36500 stop_codon 36502] <36503 CDS 38739] <38740 intron 38758] <38759 CDS 39275]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_487_length_64315_cov_1.673226:\n",
      "\t[57491, 57493]stop_codon-\n",
      "\t[57494, 59107]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<57491 stop_codon 57493] <57494 CDS 59107]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_502_length_62666_cov_1.675722:\n",
      "\t[57399, 57401]stop_codon-\n",
      "\t[57402, 58042]CDS-\n",
      "\t[58043, 58061]intron-\n",
      "\t[58062, 58875]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<57399 stop_codon 57401] <57402 CDS 58042] <58043 intron 58061] <58062 CDS 58875]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_508_length_61973_cov_1.679430:\n",
      "\t[15478, 15480]stop_codon-\n",
      "\t[15481, 16944]CDS-\n",
      "\t[56426, 60262]CDS+\n",
      "\t[60263, 60280]intron+\n",
      "\t[60281, 60355]CDS+\n",
      "\t[60356, 60358]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<15478 stop_codon 15480] <15481 CDS 16944]\n",
      "\t[56426 CDS 60262> [60263 intron 60280> [60281 CDS 60355> [60356 stop_codon 60358>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_518_length_61243_cov_1.724180:\n",
      "\t[2921, 2923]stop_codon-\n",
      "\t[2924, 3550]CDS-\n",
      "\t[3551, 3569]intron-\n",
      "\t[3570, 3656]CDS-\n",
      "\t[3657, 3672]intron-\n",
      "\t[3673, 5586]CDS-\n",
      "\t[5587, 5603]intron-\n",
      "\t[5604, 6041]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2921 stop_codon 2923] <2924 CDS 3550] <3551 intron 3569] <3570 CDS 3656] <3657 intron 3672] <3673 CDS 5586] <5587 intron 5603] <5604 CDS 6041]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_521_length_61204_cov_1.627896:\n",
      "\t[13178, 13180]stop_codon-\n",
      "\t[13181, 13540]CDS-\n",
      "\t[13541, 13557]intron-\n",
      "\t[13558, 14943]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<13178 stop_codon 13180] <13181 CDS 13540] <13541 intron 13557] <13558 CDS 14943]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_547_length_58581_cov_1.583467:\n",
      "\t[52458, 52460]stop_codon-\n",
      "\t[52461, 53795]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<52458 stop_codon 52460] <52461 CDS 53795]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_555_length_57934_cov_1.702821:\n",
      "\t[16986, 18428]CDS+\n",
      "\t[18429, 18431]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[16986 CDS 18428> [18429 stop_codon 18431>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_582_length_55810_cov_1.628343:\n",
      "\t[52960, 54075]CDS+\n",
      "\t[54076, 54091]intron+\n",
      "\t[54092, 54448]CDS+\n",
      "\t[54449, 54451]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[52960 CDS 54075> [54076 intron 54091> [54092 CDS 54448> [54449 stop_codon 54451>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_585_length_55535_cov_1.790265:\n",
      "\t[3631, 3633]stop_codon-\n",
      "\t[3634, 5637]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<3631 stop_codon 3633] <3634 CDS 5637]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_586_length_55532_cov_1.796896:\n",
      "\t[27482, 27539]CDS+\n",
      "\t[27540, 27556]intron+\n",
      "\t[27557, 27639]CDS+\n",
      "\t[27640, 27656]intron+\n",
      "\t[27657, 28669]CDS+\n",
      "\t[28670, 28686]intron+\n",
      "\t[28687, 28831]CDS+\n",
      "\t[28832, 28834]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[27482 CDS 27539> [27540 intron 27556> [27557 CDS 27639> [27640 intron 27656> [27657 CDS 28669> [28670 intron 28686> [28687 CDS 28831> [28832 stop_codon 28834>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_615_length_53271_cov_1.620145:\n",
      "\t[23436, 23438]stop_codon-\n",
      "\t[23439, 27308]CDS-\n",
      "\t[27309, 27393]intron-\n",
      "\t[27394, 29209]CDS-\n",
      "\t[29210, 29267]intron-\n",
      "\t[29268, 35851]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<23436 stop_codon 23438] <23439 CDS 27308] <27309 intron 27393] <27394 CDS 29209] <29210 intron 29267] <29268 CDS 35851]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_624_length_53013_cov_1.613678:\n",
      "\t[5315, 5317]stop_codon-\n",
      "\t[5318, 6262]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<5315 stop_codon 5317] <5318 CDS 6262]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_626_length_52975_cov_1.794202:\n",
      "\t[36980, 37864]CDS+\n",
      "\t[37865, 37867]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[36980 CDS 37864> [37865 stop_codon 37867>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_632_length_52289_cov_1.832982:\n",
      "\t[17974, 19392]CDS+\n",
      "\t[19393, 19395]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[17974 CDS 19392> [19393 stop_codon 19395>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_641_length_51465_cov_1.701001:\n",
      "\t[3895, 6625]CDS+\n",
      "\t[6626, 6644]intron+\n",
      "\t[6645, 7003]CDS+\n",
      "\t[7004, 7006]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[3895 CDS 6625> [6626 intron 6644> [6645 CDS 7003> [7004 stop_codon 7006>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_662_length_50310_cov_1.693741:\n",
      "\t[34788, 38243]CDS+\n",
      "\t[38244, 38246]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[34788 CDS 38243> [38244 stop_codon 38246>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_663_length_50272_cov_1.583887:\n",
      "\t[12451, 14853]CDS+\n",
      "\t[14854, 14872]intron+\n",
      "\t[14873, 14977]CDS+\n",
      "\t[14978, 14997]intron+\n",
      "\t[14998, 15855]CDS+\n",
      "\t[15856, 15858]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[12451 CDS 14853> [14854 intron 14872> [14873 CDS 14977> [14978 intron 14997> [14998 CDS 15855> [15856 stop_codon 15858>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_712_length_47909_cov_1.617806:\n",
      "\t[7720, 7722]stop_codon-\n",
      "\t[7723, 7869]CDS-\n",
      "\t[7870, 7885]intron-\n",
      "\t[7886, 8755]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<7720 stop_codon 7722] <7723 CDS 7869] <7870 intron 7885] <7886 CDS 8755]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_721_length_47096_cov_1.882050:\n",
      "\t[2161, 2163]stop_codon-\n",
      "\t[2164, 2232]CDS-\n",
      "\t[2233, 2249]intron-\n",
      "\t[2250, 3122]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2161 stop_codon 2163] <2164 CDS 2232] <2233 intron 2249] <2250 CDS 3122]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_725_length_46739_cov_1.861538:\n",
      "\t[11122, 11124]stop_codon-\n",
      "\t[11125, 12450]CDS-\n",
      "\t[12451, 12467]intron-\n",
      "\t[12468, 12872]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<11122 stop_codon 11124] <11125 CDS 12450] <12451 intron 12467] <12468 CDS 12872]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_726_length_46702_cov_1.652088:\n",
      "\t[26570, 26572]stop_codon-\n",
      "\t[26573, 26671]CDS-\n",
      "\t[26672, 26693]intron-\n",
      "\t[26694, 26813]CDS-\n",
      "\t[26814, 26832]intron-\n",
      "\t[26833, 29655]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<26570 stop_codon 26572] <26573 CDS 26671] <26672 intron 26693] <26694 CDS 26813] <26814 intron 26832] <26833 CDS 29655]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_748_length_45611_cov_1.600167:\n",
      "\t[11237, 11239]stop_codon-\n",
      "\t[11240, 11338]CDS-\n",
      "\t[11339, 11355]intron-\n",
      "\t[11356, 11659]CDS-\n",
      "\t[11660, 11678]intron-\n",
      "\t[11679, 12718]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<11237 stop_codon 11239] <11240 CDS 11338] <11339 intron 11355] <11356 CDS 11659] <11660 intron 11678] <11679 CDS 12718]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_762_length_45064_cov_1.971004:\n",
      "\t[40763, 42133]CDS+\n",
      "\t[42134, 42136]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[40763 CDS 42133> [42134 stop_codon 42136>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_774_length_44628_cov_1.753938:\n",
      "\t[31223, 32466]CDS+\n",
      "\t[32467, 32483]intron+\n",
      "\t[32484, 32581]CDS+\n",
      "\t[32582, 32601]intron+\n",
      "\t[32602, 32663]CDS+\n",
      "\t[32664, 32666]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[31223 CDS 32466> [32467 intron 32483> [32484 CDS 32581> [32582 intron 32601> [32602 CDS 32663> [32664 stop_codon 32666>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_810_length_43155_cov_1.672516:\n",
      "\t[26135, 28144]CDS+\n",
      "\t[28145, 28163]intron+\n",
      "\t[28164, 28253]CDS+\n",
      "\t[28254, 28271]intron+\n",
      "\t[28272, 28729]CDS+\n",
      "\t[28730, 28748]intron+\n",
      "\t[28749, 28872]CDS+\n",
      "\t[28873, 28875]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[26135 CDS 28144> [28145 intron 28163> [28164 CDS 28253> [28254 intron 28271> [28272 CDS 28729> [28730 intron 28748> [28749 CDS 28872> [28873 stop_codon 28875>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_847_length_41570_cov_1.563786:\n",
      "\t[23223, 25097]CDS+\n",
      "\t[25098, 25100]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[23223 CDS 25097> [25098 stop_codon 25100>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_889_length_39761_cov_1.904703:\n",
      "\t[33176, 33550]CDS+\n",
      "\t[33551, 33567]intron+\n",
      "\t[33568, 34539]CDS+\n",
      "\t[34540, 34559]intron+\n",
      "\t[34560, 34719]CDS+\n",
      "\t[34720, 34738]intron+\n",
      "\t[34739, 34746]CDS+\n",
      "\t[34747, 34749]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[33176 CDS 33550> [33551 intron 33567> [33568 CDS 34539> [34540 intron 34559> [34560 CDS 34719> [34720 intron 34738> [34739 CDS 34746> [34747 stop_codon 34749>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_893_length_39578_cov_1.648298:\n",
      "\t[14064, 15803]CDS+\n",
      "\t[15804, 15806]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[14064 CDS 15803> [15804 stop_codon 15806>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_896_length_39504_cov_1.683648:\n",
      "\t[25489, 25491]stop_codon-\n",
      "\t[25492, 25942]CDS-\n",
      "\t[25943, 25959]intron-\n",
      "\t[25960, 27338]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25489 stop_codon 25491] <25492 CDS 25942] <25943 intron 25959] <25960 CDS 27338]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_918_length_38891_cov_2.234522:\n",
      "\t[24965, 34442]CDS+\n",
      "\t[34443, 34460]intron+\n",
      "\t[34461, 34589]CDS+\n",
      "\t[34590, 34607]intron+\n",
      "\t[34608, 34725]CDS+\n",
      "\t[34726, 34746]intron+\n",
      "\t[34747, 34846]CDS+\n",
      "\t[34847, 34849]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[24965 CDS 34442> [34443 intron 34460> [34461 CDS 34589> [34590 intron 34607> [34608 CDS 34725> [34726 intron 34746> [34747 CDS 34846> [34847 stop_codon 34849>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_926_length_38670_cov_1.606881:\n",
      "\t[16573, 16575]stop_codon-\n",
      "\t[16576, 17703]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<16573 stop_codon 16575] <16576 CDS 17703]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_998_length_36220_cov_1.933976:\n",
      "\t[2939, 2941]stop_codon-\n",
      "\t[2942, 3011]CDS-\n",
      "\t[3012, 3031]intron-\n",
      "\t[3032, 3150]CDS-\n",
      "\t[3151, 3167]intron-\n",
      "\t[3168, 3914]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2939 stop_codon 2941] <2942 CDS 3011] <3012 intron 3031] <3032 CDS 3150] <3151 intron 3167] <3168 CDS 3914]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1067_length_34315_cov_1.894934:\n",
      "\t[24567, 24569]stop_codon-\n",
      "\t[24570, 26393]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<24567 stop_codon 24569] <24570 CDS 26393]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1083_length_33926_cov_1.664753:\n",
      "\t[25285, 25287]stop_codon-\n",
      "\t[25288, 25349]CDS-\n",
      "\t[25350, 25365]intron-\n",
      "\t[25366, 27673]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<25285 stop_codon 25287] <25288 CDS 25349] <25350 intron 25365] <25366 CDS 27673]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1087_length_33881_cov_1.716982:\n",
      "\t[13033, 13035]stop_codon-\n",
      "\t[13036, 13080]CDS-\n",
      "\t[13081, 13106]intron-\n",
      "\t[13107, 13565]CDS-\n",
      "\t[13566, 13588]intron-\n",
      "\t[13589, 25447]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<13033 stop_codon 13035] <13036 CDS 13080] <13081 intron 13106] <13107 CDS 13565] <13566 intron 13588] <13589 CDS 25447]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1148_length_32028_cov_1.539450:\n",
      "\t[13993, 15702]CDS+\n",
      "\t[15703, 15705]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[13993 CDS 15702> [15703 stop_codon 15705>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1178_length_31386_cov_1.635561:\n",
      "\t[9831, 10628]CDS+\n",
      "\t[10629, 10644]intron+\n",
      "\t[10645, 11064]CDS+\n",
      "\t[11065, 11067]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[9831 CDS 10628> [10629 intron 10644> [10645 CDS 11064> [11065 stop_codon 11067>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1295_length_28975_cov_2.083403:\n",
      "\t[24301, 25425]CDS+\n",
      "\t[25426, 25428]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[24301 CDS 25425> [25426 stop_codon 25428>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1401_length_27173_cov_1.713081:\n",
      "\t[462, 464]stop_codon-\n",
      "\t[465, 1703]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<462 stop_codon 464] <465 CDS 1703]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1404_length_27111_cov_1.507449:\n",
      "\t[17030, 18691]CDS+\n",
      "\t[18692, 18694]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[17030 CDS 18691> [18692 stop_codon 18694>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1482_length_26050_cov_1.559272:\n",
      "\t[15693, 18173]CDS+\n",
      "\t[18174, 18176]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[15693 CDS 18173> [18174 stop_codon 18176>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1721_length_22983_cov_1.829424:\n",
      "\t[15310, 16314]CDS+\n",
      "\t[16315, 16317]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[15310 CDS 16314> [16315 stop_codon 16317>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_1864_length_21573_cov_1.733703:\n",
      "\t[10945, 12819]CDS+\n",
      "\t[12820, 12822]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[10945 CDS 12819> [12820 stop_codon 12822>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_2222_length_18295_cov_4.113496:\n",
      "\t[358, 360]stop_codon-\n",
      "\t[361, 3405]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<358 stop_codon 360] <361 CDS 3405]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_3449_length_12448_cov_1.887347:\n",
      "\t[8528, 8530]stop_codon-\n",
      "\t[8531, 10114]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<8528 stop_codon 8530] <8531 CDS 10114]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_4094_length_10741_cov_1.831543:\n",
      "\t[5451, 5453]stop_codon-\n",
      "\t[5454, 7976]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<5451 stop_codon 5453] <5454 CDS 7976]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_4127_length_10657_cov_2.887179:\n",
      "\t[2063, 2065]stop_codon-\n",
      "\t[2066, 3739]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<2063 stop_codon 2065] <2066 CDS 3739]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_4354_length_10149_cov_3.036121:\n",
      "\t[4340, 4342]stop_codon-\n",
      "\t[4343, 8092]CDS-\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t<4340 stop_codon 4342] <4343 CDS 8092]\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "NODE_9798_length_4346_cov_32.044086:\n",
      "\t[2002, 3056]CDS+\n",
      "\t[3057, 3059]stop_codon+\n",
      "+ valid gene structure\n",
      "Genes:\n",
      "\t[2002 CDS 3056> [3057 stop_codon 3059>\n",
      "-  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  \n",
      "True\n",
      "Genes:\n",
      "\t[57 CDS 1380> [1381 stop_codon 1383>\n",
      "141 = 141 of which 0 were faulty\n",
      "There are 154 genes annotated\n",
      "NODE_615_length_53271_cov_1.620145 [85, 58]\n"
     ]
    }
   ],
   "source": [
    "annotations = []\n",
    "with open(annotation_filename) as annotation_file:\n",
    "    c = 0\n",
    "    f = 0\n",
    "    line = annotation_file.readline()\n",
    "    current = None\n",
    "    while line:\n",
    "        line = annotation_file.readline()[:-1]\n",
    "        content = line.split(\"\\t\")\n",
    "        if line.startswith(\"##sequence-region\"):\n",
    "            print(current)\n",
    "            if current is not None:\n",
    "                correct = current.compile_genes()\n",
    "                if not correct:\n",
    "                    f += 1\n",
    "                print(\"+ valid gene structure\" if correct else \"/!\\\\ invalid gene structure\")\n",
    "                print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "                annotations.append(current)\n",
    "            print(\"-  \" * 36)\n",
    "            \n",
    "            current = Annotation(content[1], int(content[2]), int(content[3]))\n",
    "            c += 1\n",
    "        elif len(line) > 0 and not line.startswith(\"#\"):\n",
    "            # Transform 3'UTR-annotation into stop-codons!\n",
    "            feature = content[2]\n",
    "            forward = content[6] == \"+\"\n",
    "            start = int(content[3])\n",
    "            end = int(content[4])\n",
    "            if feature == \"three_prime_UTR\":\n",
    "                if forward:\n",
    "                    end = start+2\n",
    "                else:\n",
    "                    start = end-2\n",
    "                feature = \"stop_codon\"\n",
    "                \n",
    "            current.add(Feature(feature, start, end, forward))\n",
    "\n",
    "print(current.compile_genes())\n",
    "print(\"Genes:\\n\\t\" + \"\\n\\t\".join([str(g) for g in current.genes]))\n",
    "annotations.append(current)\n",
    "print(c, \"=\", len(annotations), \"of which\", f, \"were faulty\")\n",
    "\n",
    "annotation_dict = {a.name:a for a in annotations}\n",
    "num_of_genes = np.sum([len(a.genes) for a in annotations])\n",
    "print(\"There are %s genes annotated\" % num_of_genes)\n",
    "\n",
    "\n",
    "# Get information that does not require sequences:\n",
    "intron_lengths = []\n",
    "exon_lengths = []\n",
    "intron_counts = []\n",
    "for ann in annotations:\n",
    "    for gene in ann.genes:\n",
    "        el, il = gene.get_exon_intron_stats()\n",
    "        exon_lengths += el\n",
    "        intron_lengths += il\n",
    "        if (len(il) > 0 and np.max(il) > 30):\n",
    "            print(ann.name, il)\n",
    "            # Can see that the two largest ones (58, 85) are from 615, \n",
    "            # and are not very trusted (coverage-signal is ambiguous)\n",
    "        intron_counts += [len(il)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input: Raw sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NODE_1_length_428989_cov_1.681242\n",
      "\t [ 4. 15.  1.]\n",
      "term: ATGAATTCTATATTTCAGGCT : 0.0*UGA\n",
      "\t [ 4. 11.  1.]\n",
      "term: TTATAAGAGTTTAATGGATTA : 0.0*UGA\n",
      "intron:  GTAAATTTTTATTGTAG\n",
      "\t [ 7. 17.  1.]\n",
      "term: TAGAAATACAAACTTAATATT : 0.0*UGA\n",
      "intron:  GTAAATTTTAATTGTAG\n",
      "\t [ 4. 11.  1.]\n",
      "term: AAGGGATATGTTCCTGAAAAT : 0.0*UGA\n",
      "\t [ 4. 10.  1.]\n",
      "term: GATGATCCTTTCATTGAATGG : 0.0*UGA\n",
      "NODE_6_length_313002_cov_1.859273\n",
      "\t [3. 8. 1.]\n",
      "term: CTATATGATTAGTTTAAATGT : 0.0*UGA\n",
      "NODE_7_length_300196_cov_1.642899\n",
      "\t [5. 5. 1.]\n",
      "term: CAGATAGTAAAAAGAGGTTCA : 0.0*UGA\n",
      "intron:  GTAAGTATTAATTCAG\n",
      "\t [12. 18.  1.]\n",
      "term: AAACTAGCAGTATCAAGCAAT : 0.0*UGA\n",
      "intron:  GTAATATTTTTAGATATAG\n",
      "NODE_10_length_283447_cov_1.644007\n",
      "\t [ 5. 18.  1.]\n",
      "term: CCATCTGCATTTGATTTAAGG : 0.0*UGA\n",
      "intron:  GTCAGTTTTTATTTTAG\n",
      "NODE_12_length_280749_cov_1.673450\n",
      "\t [20. 13.  1.]\n",
      "term: AAATTCGGTATCTTTAGAATG : 0.0*UGA\n",
      "NODE_18_length_257280_cov_1.854729\n",
      "\t [10. 15.  1.]\n",
      "term: TAGGAAGAGAAAGTTGCTAAT : 0.0*UGA\n",
      "intron:  GTAAATTTTTTTATATTAG\n",
      "\t [11. 15.  1.]\n",
      "term: AAGAACTTCAGAGATCCCAAG : 0.0*UGA\n",
      "NODE_19_length_257256_cov_1.830618\n",
      "\t [10. 14.  1.]\n",
      "term: TTTTCACTAATAAAAGTAGAC : 0.0*UGA\n",
      "intron:  GTAATTTTTTCATATACAG\n",
      "NODE_20_length_254023_cov_1.905639\n",
      "\t [ 5. 23.  1.]\n",
      "term: TATTCGCTTATAAAAGTAGAC : 0.0*UGA\n",
      "intron:  GTAAGTATTCATTGTTTAG\n",
      "NODE_21_length_245355_cov_1.707925\n",
      "\t [4. 6. 1.]\n",
      "term: TAGCAAAGACGTGGAAATCTT : 0.0*UGA\n",
      "NODE_22_length_243293_cov_1.665932\n",
      "\t [ 9. 17.  1.]\n",
      "term: AAATTCGGCATATTTAGAATG : 0.0*UGA\n",
      "NODE_23_length_242384_cov_1.872846\n",
      "\t [10.  6.  1.]\n",
      "term: TTCTTAGATGTTGCTTTAGTG : 0.0*UGA\n",
      "intron:  GTAATATTATAATACAG\n",
      "NODE_25_length_237540_cov_1.816270\n",
      "\t [10. 26.  1.]\n",
      "term: GAACTTATGCTCAAGATGCTT : 0.0*UGA\n",
      "intron:  GTAAACTTTTCTATCAATAG\n",
      "intron:  GTAAGATTTTAAAATTCAG\n",
      "NODE_27_length_235310_cov_1.764511\n",
      "\t [ 7. 19.  1.]\n",
      "term: ACAGTAAAAATATGGAGTTGT : 0.0*UGA\n",
      "NODE_29_length_230800_cov_1.889965\n",
      "\t [12. 21.  1.]\n",
      "term: GACACAGCTCTTTACACCTAG : 0.0*UGA\n",
      "intron:  GTAAGGCTATAATTCAG\n",
      "NODE_31_length_222000_cov_1.724356\n",
      "\t [ 9. 19.  1.]\n",
      "term: AGATACGAGTAGTTTTTCACT : 0.0*UGA\n",
      "intron:  GTAATTACTTTATACAG\n",
      "NODE_38_length_212895_cov_1.850617\n",
      "\t [16. 24.  1.]\n",
      "term: ATGTATTAGGAAGCCAAGATC : 0.0*UGA\n",
      "NODE_39_length_211856_cov_1.789892\n",
      "\t [8. 9. 1.]\n",
      "term: TTTGTAAGGCCTGCTGGATGT : 0.0*UGA\n",
      "NODE_42_length_207565_cov_1.992215\n",
      "\t [ 8. 13.  1.]\n",
      "term: ACAGTTAAAATTTGGGTGTGT : 0.0*UGA\n",
      "NODE_44_length_205421_cov_1.892588\n",
      "\t [11. 19.  1.]\n",
      "term: CCTTCTTCATGGGATATGAGA : 0.0*UGA\n",
      "intron:  GTAAATTTTATTTATGTAG\n",
      "NODE_54_length_192099_cov_1.873872\n",
      "\t [10. 21.  1.]\n",
      "term: CTTGCCATGCTAGTTGTTGCT : 0.0*UGA\n",
      "NODE_64_length_178741_cov_1.661034\n",
      "\t [ 4. 12.  1.]\n",
      "term: AGAGAATAGATGCTTTCCAAG : 0.0*UGA\n",
      "NODE_65_length_177181_cov_1.879014\n",
      "\t [9. 9. 1.]\n",
      "term: TCAGTTGTTCCGTATTTCCTT : 0.0*UGA\n",
      "intron:  GTAATTTCATTATGCAG\n",
      "NODE_69_length_175562_cov_1.893300\n",
      "\t [18. 17.  1.]\n",
      "term: CTGTATTCAAGAGCTAGATTA : 0.0*UGA\n",
      "NODE_70_length_174851_cov_1.725001\n",
      "\t [29. 57.  1.]\n",
      "term: TACTTCTACAAATTCATTCGT : 0.0*UGA\n",
      "NODE_71_length_174712_cov_2.292786\n",
      "\t [12. 17.  1.]\n",
      "term: TGGGCCTAGTATTGGAGACAA : 0.0*UGA\n",
      "NODE_75_length_172946_cov_1.727212\n",
      "\t [ 9. 15.  1.]\n",
      "term: GCTAAACAGGGCTGCTGCCAT : 0.0*UGA\n",
      "NODE_78_length_168807_cov_1.849484\n",
      "\t [ 4. 12.  1.]\n",
      "term: CGTACTCCCGATTATTTCCTT : 0.0*UGA\n",
      "intron:  GTAATTTTTTATGTAG\n",
      "\t [ 9. 23.  1.]\n",
      "term: ATGTATATCCCAAAAGTAATC : 0.0*UGA\n",
      "\t [ 9. 22.  1.]\n",
      "term: GTTTTTGTACCAGGAATTATC : 0.0*UGA\n",
      "NODE_80_length_167342_cov_1.702640\n",
      "\t [ 8. 16.  1.]\n",
      "term: AAAACTTTCGCTTCCTTTTAT : 0.0*UGA\n",
      "NODE_84_length_162193_cov_1.961824\n",
      "\t [ 3. 16.  1.]\n",
      "term: AAGAACTTCAGAGATCCCAAG : 0.0*UGA\n",
      "NODE_90_length_157735_cov_1.661280\n",
      "\t [2. 8. 1.]\n",
      "term: CTCTATGAGCAGCTTAAGTGT : 0.0*UGA\n",
      "NODE_91_length_156069_cov_1.747919\n",
      "\t [ 6. 22.  1.]\n",
      "term: TTTCTAGATGTTGCTCTAGGT : 0.0*UGA\n",
      "intron:  GTAATTTTTCATGCTTTAG\n",
      "NODE_103_length_148054_cov_1.718544\n",
      "\t [ 6. 18.  1.]\n",
      "term: ATTATGTAGTGGAAAAGAGCT : 0.0*UGA\n",
      "intron:  GTAATATTTATTAATTTTAG\n",
      "NODE_104_length_148052_cov_1.912118\n",
      "\t [ 5. 14.  1.]\n",
      "term: GTGAACTGTGGAGAAGTTCTT : 0.0*UGA\n",
      "NODE_114_length_139872_cov_1.858449\n",
      "\t [12. 20.  1.]\n",
      "term: CCTCCAAAGGAAAGAGTTTAT : 0.0*UGA\n",
      "NODE_127_length_132101_cov_1.744328\n",
      "\t [11. 18.  1.]\n",
      "term: ATGGTCAATAAGCCCAAACAA : 0.0*UGA\n",
      "intron:  GTAAATTTTAATTTTTAG\n",
      "\t [18. 28.  1.]\n",
      "term: TTTGGAAGAGGACTTATTGAT : 0.0*UGA\n",
      "NODE_132_length_130847_cov_1.874350\n",
      "\t [12. 14.  1.]\n",
      "term: GCGAAACCTTTTGCTTTTTTC : 0.0*UGA\n",
      "NODE_134_length_130250_cov_1.655841\n",
      "\t [12. 16.  1.]\n",
      "term: AATCCAGAATACCTTTTGCTC : 0.0*UGA\n",
      "NODE_135_length_129955_cov_1.831608\n",
      "\t [ 0. 10.  1.]\n",
      "term: GCAAATTTAATCCCTCCTATT : 0.0*UGA\n",
      "intron:  GTAAGCTTTTATATAG\n",
      "NODE_145_length_125098_cov_1.669611\n",
      "\t [16. 27.  1.]\n",
      "term: GTTAAACTGGCAGTATCAAAT : 0.0*UGA\n",
      "intron:  GTAATTTTTTAATATAG\n",
      "NODE_150_length_122897_cov_1.750631\n",
      "\t [ 7. 15.  1.]\n",
      "term: GGAGCTGAAAATTTCCTAAGT : 0.0*UGA\n",
      "NODE_153_length_122011_cov_1.703448\n",
      "\t [9. 5. 1.]\n",
      "term: CAAGTGCCAGATTATTTTCTG : 0.0*UGA\n",
      "intron:  GTGAGGAGTTATATAG\n",
      "\t [12.  8.  1.]\n",
      "term: TAGGATCTGTCTGAAATTTTG : 0.0*UGA\n",
      "NODE_155_length_121415_cov_1.724095\n",
      "\t [6. 8. 1.]\n",
      "term: GTTGCAGCTCTTTCTGATATT : 0.0*UGA\n",
      "NODE_159_length_120583_cov_1.600319\n",
      "\t [11. 19.  1.]\n",
      "term: TTAAAAAGCTCTTATTATATT : 0.0*UGA\n",
      "NODE_160_length_120340_cov_1.623793\n",
      "\t [6. 8. 1.]\n",
      "term: TTAAGAACAATAGAGAAAATA : 0.0*UGA\n",
      "NODE_164_length_119121_cov_1.871162\n",
      "\t [ 9. 14.  1.]\n",
      "term: AGTGCAGAGAAGACTGGATGC : 0.0*UGA\n",
      "NODE_176_length_115965_cov_1.774081\n",
      "\t [15. 29.  1.]\n",
      "term: AGATATTCTGAGTTGTTTTCA : 0.0*UGA\n",
      "intron:  GTAAAATTTTATGCAG\n",
      "intron:  GTAAGTATTAATATAG\n",
      "intron:  GTAAGCTTTAACTTAG\n",
      "NODE_177_length_115532_cov_1.858698\n",
      "\t [ 3. 12.  1.]\n",
      "term: AAGCAAATATTTGTGATGTCT : 0.0*UGA\n",
      "NODE_181_length_114511_cov_1.615750\n",
      "\t [29.  7.  1.]\n",
      "term: TTGTATTAGAGTGCTAAGATT : 0.0*UGA\n",
      "NODE_202_length_107880_cov_1.614990\n",
      "\t [ 4. 19.  1.]\n",
      "term: CCTCCAAAAGAAAGAGTTTAT : 0.0*UGA\n",
      "NODE_206_length_105986_cov_1.546198\n",
      "\t [12. 26.  1.]\n",
      "term: GCAGGCTTGAAAATGAAGGCT : 0.0*UGA\n",
      "NODE_207_length_105357_cov_1.894403\n",
      "\t [ 8. 14.  1.]\n",
      "term: GATCCAAAAAGACTAACTCCA : 0.0*UGA\n",
      "NODE_210_length_104547_cov_1.706570\n",
      "\t [ 2. 30.  1.]\n",
      "term: GACACAGCTCTTTACAACTAG : 0.0*UGA\n",
      "intron:  GTAAACCATAATACAG\n",
      "NODE_217_length_103446_cov_1.848866\n",
      "\t [15. 10.  1.]\n",
      "term: CTAAGGCAATTTACAAGTATT : 0.0*UGA\n",
      "NODE_222_length_101554_cov_1.892070\n",
      "\t [10.  4.  1.]\n",
      "term: CGTACGCCGGATTATTTCCTC : 0.0*UGA\n",
      "intron:  GTAATCATTTATATAG\n",
      "NODE_227_length_100883_cov_2.022033\n",
      "\t [14.  8.  1.]\n",
      "term: GCAAATTTAATCCCTCCCATT : 0.0*UGA\n",
      "intron:  GTAAGTACTAATATAG\n",
      "NODE_244_length_98187_cov_1.698399\n",
      "\t [ 7. 16.  1.]\n",
      "term: TTTTCACTTGTCAAAGTCGAC : 0.0*UGA\n",
      "intron:  GTAAGTTTTTTATATTTTAG\n",
      "NODE_246_length_97678_cov_1.774323\n",
      "\t [14. 23.  1.]\n",
      "term: AAACTGGCAGTTACAAGCAAT : 0.0*UGA\n",
      "intron:  GTAATTTTTATAATCCTAG\n",
      "NODE_250_length_96914_cov_1.797266\n",
      "\t [3. 9. 1.]\n",
      "term: CAAGATTTGTCTGAAATTCTT : 0.0*UGA\n",
      "NODE_259_length_95643_cov_1.674882\n",
      "\t [30. 46.  1.]\n",
      "term: ATTCCAGGAGTTGAGTTCCCT : 0.0*UGA\n",
      "intron:  GTAATTTTTGATATTATATAG\n",
      "NODE_262_length_95179_cov_1.968112\n",
      "\t [23. 12.  1.]\n",
      "term: AGTTGTTTCCATAAAGACAGC : 0.0*UGA\n",
      "intron:  GTAAATTCTTAATGTAG\n",
      "intron:  GTAAGTTAAATCTAGTTTAG\n",
      "intron:  GTAAGCTATAATATACAG\n",
      "NODE_264_length_94820_cov_1.815794\n",
      "\t [ 6. 11.  1.]\n",
      "term: TCTATCAAGAAATTATGGAAG : 0.0*UGA\n",
      "intron:  GTGAGATTTTATGTAG\n",
      "NODE_265_length_94667_cov_1.884356\n",
      "\t [1. 6. 1.]\n",
      "term: AAAAACGTCTTGTTCTATCTG : 0.0*UGA\n",
      "intron:  GTAAGGATTTTAATATATAG\n",
      "\t [3. 8. 1.]\n",
      "term: AAGCATCCCTACTTACTATCT : 0.0*UGA\n",
      "intron:  GTAAGATTTTTTATAAAAAG\n",
      "NODE_267_length_94345_cov_1.719173\n",
      "\t [10. 28.  1.]\n",
      "term: TTTCTAGATGTTGCTTTAGGG : 0.0*UGA\n",
      "intron:  GTAATCTTTACTATACAG\n",
      "NODE_271_length_94123_cov_1.840504\n",
      "\t [ 6. 19.  1.]\n",
      "term: AAGAGTCAAGATATCTGCTAT : 0.0*UGA\n",
      "intron:  GTAATTTAATAATATAG\n",
      "NODE_273_length_93569_cov_1.730303\n",
      "\t [27. 76.  1.]\n",
      "term: CTAGACTAGGAGTATGATCTA : 0.0*UGA\n",
      "NODE_278_length_92397_cov_1.658773\n",
      "\t [ 2. 10.  1.]\n",
      "term: GATGATCCTTTCATTGAATGG : 0.0*UGA\n",
      "\t [ 2. 22.  1.]\n",
      "term: TTCAGCAAGCCCATTCATTTT : 0.0*UGA\n",
      "NODE_279_length_92227_cov_1.518132\n",
      "\t [6. 8. 1.]\n",
      "term: GAAAACCCATTCAGAGATTGG : 0.0*UGA\n",
      "NODE_296_length_88345_cov_1.720182\n",
      "\t [6. 7. 1.]\n",
      "term: AGAACTCCAGACTATTTCCTT : 0.0*UGA\n",
      "intron:  GTAAGTAAATAATACAG\n",
      "NODE_297_length_88055_cov_1.968429\n",
      "\t [17. 25.  1.]\n",
      "term: ATCATGTAGTGGAAGCGTGTA : 0.0*UGA\n",
      "intron:  GTAAGCTTATCTTATTTTAG\n",
      "NODE_301_length_87574_cov_1.953023\n",
      "\t [11. 19.  1.]\n",
      "term: GTAAGACAAAGCAATTTTATT : 0.0*UGA\n",
      "NODE_322_length_82363_cov_1.603434\n",
      "\t [13. 49.  1.]\n",
      "term: ATTGAGAAAAATACAAGTTGG : 0.0*UGA\n",
      "NODE_324_length_81694_cov_1.607292\n",
      "\t [ 5. 10.  1.]\n",
      "term: TCCGTTGTTCCATACTTCTTA : 0.0*UGA\n",
      "intron:  GTAATTATTAATATAG\n",
      "NODE_336_length_80118_cov_1.672351\n",
      "\t [20. 42.  1.]\n",
      "term: AAACGCTGTTGCAGAAAATCC : 0.0*UGA\n",
      "NODE_339_length_79629_cov_1.996868\n",
      "\t [16. 25.  1.]\n",
      "term: CTATTTTACCAAAGAAAGCAA : 0.0*UGA\n",
      "NODE_344_length_79198_cov_1.698916\n",
      "\t [11. 35.  1.]\n",
      "term: TACTCAGGAGCATACGCAGTT : 0.0*UGA\n",
      "intron:  GTAATCTTTCATACAG\n",
      "intron:  GTAAGTATTTCATTTAG\n",
      "NODE_345_length_79091_cov_2.091434\n",
      "\t [ 9. 13.  1.]\n",
      "term: TTTTCTCTTATTAAAATAGAC : 0.0*UGA\n",
      "intron:  GTAAGTATTTTCATTTTAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NODE_358_length_77800_cov_1.675254\n",
      "\t [17. 56.  1.]\n",
      "term: ATAGAAAACTATACCTACTGG : 0.0*UGA\n",
      "NODE_364_length_77330_cov_3.110864\n",
      "\t [15. 27.  1.]\n",
      "term: CCTAACCTGGCTATGTATTTT : 0.0*UGA\n",
      "NODE_376_length_75967_cov_1.666139\n",
      "\t [19. 11.  1.]\n",
      "term: AGAATTTTCTTATACCGCCGT : 0.0*UGA\n",
      "NODE_379_length_75420_cov_1.946927\n",
      "\t [4. 9. 1.]\n",
      "term: CTCTATGCCTCACTAAAATGC : 0.0*UGA\n",
      "NODE_386_length_74730_cov_1.852566\n",
      "\t [ 8. 11.  1.]\n",
      "term: AGGAAAGGTGTTATAGGAAAG : 0.0*UGA\n",
      "intron:  GTAAGTATTTAATACAG\n",
      "NODE_389_length_74397_cov_1.930066\n",
      "\t [12. 18.  1.]\n",
      "term: ACAGTCAAAATTTGGGTATGC : 0.0*UGA\n",
      "NODE_410_length_72732_cov_1.702100\n",
      "\t [ 8. 27.  1.]\n",
      "term: CTATAGAAATTGTCTCTTACC : 0.0*UGA\n",
      "intron:  GTAAGTTATAATGTTTAG\n",
      "intron:  GTAATTTATAATTTAG\n",
      "intron:  GTAAACTTGTCTATTTTAG\n",
      "intron:  GTATTTTTTATTTATCTAG\n",
      "intron:  GTAAGCTTTTAACTTTTAG\n",
      "NODE_426_length_70982_cov_1.681053\n",
      "\t [14. 13.  1.]\n",
      "term: TCAGAGGCGATACAAAAATAT : 0.0*UGA\n",
      "intron:  GTAATTTATTATACAG\n",
      "NODE_427_length_70843_cov_2.049663\n",
      "\t [3. 9. 1.]\n",
      "term: AGAATCTTCTTATATCGCCGT : 0.0*UGA\n",
      "NODE_450_length_68074_cov_1.636967\n",
      "\t [16. 35.  1.]\n",
      "term: AGAATAGCCACTTCTCCTACT : 0.0*UGA\n",
      "NODE_455_length_67931_cov_1.705902\n",
      "\t [ 45. 126.   1.]\n",
      "term: CTACTTACATAGCTTGACTAT : 0.0*UGA\n",
      "intron:  GTAATATTATTATATAG\n",
      "intron:  GTAAGATTATTTATATAG\n",
      "NODE_469_length_66432_cov_1.646482\n",
      "\t [10. 11.  1.]\n",
      "term: TTGTTATCTCTTATTAAGACT : 0.0*UGA\n",
      "intron:  GTAAGTTTTTATGTAG\n",
      "NODE_476_length_65419_cov_1.711113\n",
      "\t [12. 11.  1.]\n",
      "term: ATTTTCAAATAGCAGTTATCT : 0.0*UGA\n",
      "NODE_479_length_65262_cov_1.758118\n",
      "\t [15. 24.  1.]\n",
      "term: GCTAAATGGGACGAAGACTTT : 0.0*UGA\n",
      "intron:  GTAAGGCTATAACACAG\n",
      "intron:  GTAAGTTAATAACTTTAG\n",
      "NODE_485_length_64803_cov_1.657740\n",
      "\t [14. 42.  1.]\n",
      "term: TTTAGGGAGCTCACAGAAAAG : 0.0*UGA\n",
      "intron:  GTAATTTTAATAATTGTAG\n",
      "NODE_487_length_64315_cov_1.673226\n",
      "\t [13. 16.  1.]\n",
      "term: GCTGGATTAAAAATGAAGGCT : 0.0*UGA\n",
      "NODE_502_length_62666_cov_1.675722\n",
      "\t [3. 6. 1.]\n",
      "term: GAAAATCCTATATGTTCTTTT : 0.0*UGA\n",
      "intron:  GTATTTTTATTTAAATTAG\n",
      "NODE_508_length_61973_cov_1.679430\n",
      "\t [15. 27.  1.]\n",
      "term: GACTGTGGAGAAATCCATCCT : 0.0*UGA\n",
      "\t [27. 50.  1.]\n",
      "term: GCAAGAGCGAATTTAAGAATT : 0.0*UGA\n",
      "intron:  GTAAGTATTAAATTTTAG\n",
      "NODE_518_length_61243_cov_1.724180\n",
      "\t [12. 25.  1.]\n",
      "term: TATAGAGAGCTCACTGAAAAA : 0.0*UGA\n",
      "intron:  GTAATTTTTTAATTTGTAG\n",
      "intron:  GTAAGTATTTATGTAG\n",
      "intron:  GTAAGTTAATTACACAG\n",
      "NODE_521_length_61204_cov_1.627896\n",
      "\t [10. 18.  1.]\n",
      "term: AAATTTGAGAATTCAGGATTT : 0.0*UGA\n",
      "intron:  GTAATTATTTATTTTAG\n",
      "NODE_547_length_58581_cov_1.583467\n",
      "\t [ 4. 16.  1.]\n",
      "term: GTTGATAGATTCAAACAACTA : 0.0*UGA\n",
      "NODE_555_length_57934_cov_1.702821\n",
      "\t [ 8. 13.  1.]\n",
      "term: AGAGAGCCCCTTAAGACTTTC : 0.0*UGA\n",
      "NODE_582_length_55810_cov_1.628343\n",
      "\t [ 9. 11.  1.]\n",
      "term: CCTTACAAAGTGATTCCACAA : 0.0*UGA\n",
      "intron:  GTAATTTATAATGCAG\n",
      "NODE_585_length_55535_cov_1.790265\n",
      "\t [ 7. 19.  1.]\n",
      "term: GCAAAAGTAGTTACTGTTGAT : 0.0*UGA\n",
      "NODE_586_length_55532_cov_1.796896\n",
      "\t [ 5. 14.  1.]\n",
      "term: TCTTCATACGTATGTCGTCTC : 0.0*UGA\n",
      "intron:  GTAAATTTTTACTGTAG\n",
      "intron:  GTAAAATTTTAATTAAG\n",
      "intron:  GTAATTTTCTATTTTAG\n",
      "NODE_615_length_53271_cov_1.620145\n",
      "\t [100. 104.   1.]\n",
      "term: GCCTTCTTAGCTTTGAGAAAC : 0.0*UGA\n",
      "intron:  GTAAGTTTAATTTACGCGCTATCATATCCACAATGTGGAATGATAGCTATTTCTTTCAACTGAATTGATATTAACTATACTATAG\n",
      "intron:  GTAAGTATATTTTCACTATGAATATCACATTTTTGAACAAATACTAACTAATATATAG\n",
      "NODE_624_length_53013_cov_1.613678\n",
      "\t [4. 7. 1.]\n",
      "term: AGGATTTTCTTATATCGCCGT : 0.0*UGA\n",
      "NODE_626_length_52975_cov_1.794202\n",
      "\t [2. 7. 1.]\n",
      "term: CTCTGGAGAAACACATTAGGT : 0.0*UGA\n",
      "NODE_632_length_52289_cov_1.832982\n",
      "\t [ 6. 15.  1.]\n",
      "term: GAAGAAATGCTTGGCCCAAAA : 0.0*UGA\n",
      "NODE_641_length_51465_cov_1.701001\n",
      "\t [22. 28.  1.]\n",
      "term: GTAATTGAAAAGCTAACATGG : 0.0*UGA\n",
      "intron:  GTAAGTATTTTTATATTAG\n",
      "NODE_662_length_50310_cov_1.693741\n",
      "\t [16. 20.  1.]\n",
      "term: CTCTTTTATAGATATAAAGGG : 0.0*UGA\n",
      "NODE_663_length_50272_cov_1.583887\n",
      "\t [14. 58.  1.]\n",
      "term: GCTAAGAGGAAAAACAAAGCT : 0.0*UGA\n",
      "intron:  GTAAGTTTTCATTATTCAG\n",
      "intron:  GTAAGAATATTAGTATATAG\n",
      "NODE_712_length_47909_cov_1.617806\n",
      "\t [7. 4. 1.]\n",
      "term: TAGGCAGTCATCAATAACTTG : 0.0*UGA\n",
      "intron:  GTAATTCTTAATTTAG\n",
      "NODE_721_length_47096_cov_1.882050\n",
      "\t [2. 6. 1.]\n",
      "term: ATGCTAAACACTGTCATCTTC : 0.0*UGA\n",
      "intron:  GTAATTTTTAATTGTAG\n",
      "NODE_725_length_46739_cov_1.861538\n",
      "\t [14. 21.  1.]\n",
      "term: TTTCTTGATGTTGCACTTGGT : 0.0*UGA\n",
      "intron:  GTAAGTTTTAACATTAG\n",
      "NODE_726_length_46702_cov_1.652088\n",
      "\t [19. 18.  1.]\n",
      "term: CCAAACTGGGACCAAGACTTT : 0.0*UGA\n",
      "intron:  GTAAGATCTCATTTCTTTACAG\n",
      "intron:  GTAAATTTTTTAACTTCAG\n",
      "NODE_748_length_45611_cov_1.600167\n",
      "\t [ 7. 19.  1.]\n",
      "term: TTGGCTAATATTCCCAAGCCA : 0.0*UGA\n",
      "intron:  GTAAGATTTTTATTAAG\n",
      "intron:  GTAACTATTAATATTTTAG\n",
      "NODE_762_length_45064_cov_1.971004\n",
      "\t [ 7. 19.  1.]\n",
      "term: ATAAAAACTTTAGAAGAAAAT : 0.0*UGA\n",
      "NODE_774_length_44628_cov_1.753938\n",
      "\t [ 6. 15.  1.]\n",
      "term: AATGAAATGAGGCCTTAGTAA : 0.0*UGA\n",
      "intron:  GTATATTTTAATTTCAG\n",
      "intron:  GTAATTTTTTTTAATTTCAG\n",
      "NODE_810_length_43155_cov_1.672516\n",
      "\t [20. 24.  1.]\n",
      "term: CAGCGTGTTTAGGATAATTAC : 0.0*UGA\n",
      "intron:  GTAAGTTTTGAACATTCAG\n",
      "intron:  GTAAGGTTATTATTTTAG\n",
      "intron:  GTAAGACTTTTATATAAAG\n",
      "NODE_847_length_41570_cov_1.563786\n",
      "\t [ 2. 10.  1.]\n",
      "term: CCTCCAAAGGAAAGAGTTTAT : 0.0*UGA\n",
      "NODE_889_length_39761_cov_1.904703\n",
      "\t [ 4. 26.  1.]\n",
      "term: GATCTCCTAACTAATGACGAA : 0.0*UGA\n",
      "intron:  GTAATTTTCAAGTTTAG\n",
      "intron:  GTAATTTTTCTTTATGTAAG\n",
      "intron:  GTAAGAATTCAAAATTTAG\n",
      "NODE_893_length_39578_cov_1.648298\n",
      "\t [12. 28.  1.]\n",
      "term: GCAAAGCAATCCTGCTGCCAC : 0.0*UGA\n",
      "NODE_896_length_39504_cov_1.683648\n",
      "\t [ 7. 30.  1.]\n",
      "term: ATGGTCAATAAACCAAAACAA : 0.0*UGA\n",
      "intron:  GTAATTTTTAATTATAG\n",
      "NODE_918_length_38891_cov_2.234522\n",
      "\t [ 52. 113.   1.]\n",
      "term: AAAGAAGGATTTTTGTTTGTC : 0.0*UGA\n",
      "intron:  GTAAGTTTACTATTTTAG\n",
      "intron:  GTAATTTTTCATTTTTAG\n",
      "intron:  GTAATTTTTTTTTAATTTCAG\n",
      "NODE_926_length_38670_cov_1.606881\n",
      "\t [ 6. 11.  1.]\n",
      "term: ATCATACACAAAAAATGCATT : 0.0*UGA\n",
      "NODE_998_length_36220_cov_1.933976\n",
      "\t [ 7. 19.  1.]\n",
      "term: AGAACTCCTGACTATTTCCTT : 0.0*UGA\n",
      "intron:  GTAAAATTTTCATTAGTTAG\n",
      "intron:  GTAAGTTATAAAATTAG\n",
      "NODE_1067_length_34315_cov_1.894934\n",
      "\t [11. 10.  1.]\n",
      "term: ATTGAAGAAAGATTTCTACTA : 0.0*UGA\n",
      "NODE_1083_length_33926_cov_1.664753\n",
      "\t [12. 14.  1.]\n",
      "term: CCAAACAACGAAAGAACTTTG : 0.0*UGA\n",
      "intron:  GTAATTATTAATACAG\n",
      "NODE_1087_length_33881_cov_1.716982\n",
      "\t [ 78. 169.   1.]\n",
      "term: GAAGGAGGTTTTGAGGAACAA : 0.0*UGA\n",
      "intron:  GTAAGGTTAAATTTAACCATTTATAG\n",
      "intron:  GTAAGGGCCATTTATATATTTAG\n",
      "NODE_1148_length_32028_cov_1.539450\n",
      "\t [ 7. 12.  1.]\n",
      "term: GCAGGAGTTATCAAAATGGGA : 0.0*UGA\n",
      "NODE_1178_length_31386_cov_1.635561\n",
      "\t [10. 16.  1.]\n",
      "term: GACGCTGCTCTATACAATTAG : 0.0*UGA\n",
      "intron:  GTAAGTTATAATTCAG\n",
      "NODE_1295_length_28975_cov_2.083403\n",
      "\t [ 6. 14.  1.]\n",
      "term: ATAGTACACAGAAAGTGTTTT : 0.0*UGA\n",
      "NODE_1401_length_27173_cov_1.713081\n",
      "\t [ 3. 16.  1.]\n",
      "term: GTTGCCGCTCTCTCTGACATC : 0.0*UGA\n",
      "NODE_1404_length_27111_cov_1.507449\n",
      "\t [ 8. 20.  1.]\n",
      "term: AGAAATGCTGGACCTAAATAC : 0.0*UGA\n",
      "NODE_1482_length_26050_cov_1.559272\n",
      "\t [13. 23.  1.]\n",
      "term: TTTATGGGTCTTAGAAAAATG : 0.0*UGA\n",
      "NODE_1721_length_22983_cov_1.829424\n",
      "\t [ 6. 14.  1.]\n",
      "term: AAAGAGGTGGAAAAAATGCTT : 0.0*UGA\n",
      "NODE_1864_length_21573_cov_1.733703\n",
      "\t [13. 47.  1.]\n",
      "term: CCTCCTAAGGAAAGAGTTTAT : 0.0*UGA\n",
      "NODE_2222_length_18295_cov_4.113496\n",
      "\t [12. 14.  1.]\n",
      "term: GCAGTTAAGTTAATTTTCGCT : 0.0*UGA\n",
      "NODE_3449_length_12448_cov_1.887347\n",
      "\t [ 9. 17.  1.]\n",
      "term: CAGTCAGAGTTGGACGATATG : 0.0*UGA\n",
      "NODE_4094_length_10741_cov_1.831543\n",
      "\t [19. 40.  1.]\n",
      "term: AATACTTTGCAAACCAACAGA : 0.0*UGA\n",
      "NODE_4127_length_10657_cov_2.887179\n",
      "\t [10. 13.  1.]\n",
      "term: AGAAGAGCTGGCCCTAAATAC : 0.0*UGA\n",
      "NODE_4354_length_10149_cov_3.036121\n",
      "\t [24. 23.  1.]\n",
      "term: ATTGAACGTAACACCTGTTGG : 0.0*UGA\n",
      "NODE_9798_length_4346_cov_32.044086\n",
      "\t [5. 8. 1.]\n",
      "term: ATCGTACACAGAAAATGTTTT : 0.0*UGA\n",
      "NODE_13726_length_2657_cov_6.379447\n",
      "\t [ 1. 32.  1.]\n",
      "term: ATCGAAAAAAATACAAGTTGG : 0.0*UGA\n"
     ]
    }
   ],
   "source": [
    "counts = np.zeros((num_of_genes,3))\n",
    "codon_usage_general = np.zeros(64)\n",
    "codon_usage_stop = np.zeros(64)\n",
    "codon_number = 0\n",
    "stop_codon_number = 0\n",
    "\n",
    "start_upstream = 3\n",
    "start_regions = []\n",
    "introns_by_length = {}\n",
    "ncs_base_counts = np.zeros(4)\n",
    "\n",
    "current_header = None\n",
    "current_sequence = \"\"\n",
    "gene_index = 0\n",
    "with open(sequence_filename) as scaffolds:\n",
    "    for row, line in enumerate(scaffolds):\n",
    "        if line.startswith(\">\"):\n",
    "            # Here query the analysis\n",
    "            if current_header is not None:\n",
    "                if current_header in annotation_dict:\n",
    "                    print(current_header)\n",
    "                    ann = annotation_dict[current_header]\n",
    "                    for gene in ann.genes:\n",
    "                        # Collect base-frequencies in flanking NCS\n",
    "                        \n",
    "                        start, end = gene.get_range()\n",
    "                        ncs_left = current_sequence[max(0, start -1 -hyperparameters[\"ncs_flank_size\"]): start]\n",
    "                        ncs_right = current_sequence[end: min(end+hyperparameters[\"ncs_flank_size\"], len(current_sequence))]\n",
    "                        ncs = ncs_left + ncs_right\n",
    "                        \n",
    "                        for i in range(4):\n",
    "                            ncs_base_counts[i] += ncs.count(base_order[i])\n",
    "                        \n",
    "                        # Collect number of inframe-stops, outframe-stops, and check for presence of stop as annotated\n",
    "                        counts[gene_index, :] = np.array(gene.count_UGA(current_sequence))\n",
    "                        print(\"\\t\", counts[gene_index,:])\n",
    "                        if counts[gene_index,2] == 0:\n",
    "                            print(\"\\t\\t\\t\\t!>!>!>missing STOP\")\n",
    "                            \n",
    "                        # Collect codon usage in CDS and stop regions\n",
    "                        cds_codon_counts, cds_num_of_codons, stop_region = gene.codon_usage(current_sequence)\n",
    "                        codon_number += cds_num_of_codons\n",
    "                        stop_codon_number += 7\n",
    "                        codon_usage_general += cds_codon_counts\n",
    "                        codon_usage_stop += stop_region\n",
    "                        \n",
    "                        gene_index += 1\n",
    "                        \n",
    "                        # Collect intron sequences:\n",
    "                        for f in [f for f in gene.features if f.feature == \"intron\"]:\n",
    "                            curr_intron = current_sequence[(f.start-1):f.end]\n",
    "                            curr_intron_length = f.end - f.start + 1\n",
    "                            if not f.strand:\n",
    "                                curr_intron = rev_comp(curr_intron)\n",
    "                            if curr_intron_length not in introns_by_length:\n",
    "                                introns_by_length[curr_intron_length] = [curr_intron]\n",
    "                            else:\n",
    "                                introns_by_length[curr_intron_length] += [curr_intron]\n",
    "                            \n",
    "                            print(\"intron: \", curr_intron)\n",
    "                            \n",
    "                        # Collect start-region:\n",
    "                        g_start, g_end = gene.get_range()\n",
    "                        if gene.strand:\n",
    "                            start_regions += [current_sequence[max(0, g_start-1-start_upstream):\n",
    "                                                               min(len(current_sequence), g_start+2)]]\n",
    "                        else:\n",
    "                            start_regions += [rev_comp(current_sequence[max(0, g_end-3):\n",
    "                                                                        min(len(current_sequence), g_end+start_upstream)])]\n",
    "                        if not start_regions[-1].endswith(\"ATG\"):\n",
    "                            print(\"! weird Start-region\", start_regions[-1], \"\\n\")\n",
    "            \n",
    "            current_header = line[1:-1]\n",
    "            current_sequence = \"\"\n",
    "        else:\n",
    "            current_sequence += line[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NCS-parameters\n",
    " - Estimate NCS-base-frequency from flanking NCS\n",
    " - How estimate NCS-length? Have user specify mean NCS-length estimate (allowed to be manual), thence geometric estimate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters[\"base_frequencies_NCS\"] = \"{%s}\" % \", \".join([str(f) for f in ncs_base_counts / np.sum(ncs_base_counts)])\n",
    "parameters[\"transition_probability_of_staying_in_NCS\"] = 1-1/hyperparameters[\"ncs_mean_length\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDS-parameters\n",
    " - Estimate probability to stay in CDS by assuming geometric **distr. of cds/exon length** (mean)\n",
    " - Estimate probability of going to stop by assuming geometric model of **number of introns** (mean)\n",
    " - Compute empirical **positional base frequency marginals**\n",
    " - For Stop-region: compute empirical **codon usage** outside stop-regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters[\"transition_probability_of_staying_in_CDS\"] = 1 - 3 / np.mean(exon_lengths)\n",
    "parameters[\"transition_probability_of_CDS_to_stop_given_that_CDS_is_being_left\"] = 1/(1+np.mean(intron_counts))\n",
    "\n",
    "cds_positional_marginals = positional_base_usage(codon_usage_general / codon_number)\n",
    "\n",
    "parameters[\"base_frequency_marginals_CDS\"] = format_matrix(cds_positional_marginals)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start region parameters\n",
    "These are difficult to infer automatically (especially the length of the pattern); can infer the base-probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sanity check: the last 3 rows should show deterministic ATG\n",
      "[[0.16233766 0.         0.77922078 0.05844156]\n",
      " [0.14935065 0.02597403 0.77922078 0.04545455]\n",
      " [0.14935065 0.01948052 0.74025974 0.09090909]\n",
      " [0.         0.         1.         0.        ]\n",
      " [1.         0.         0.         0.        ]\n",
      " [0.         0.         0.         1.        ]]\n"
     ]
    }
   ],
   "source": [
    "start_region_bases = np.zeros((start_upstream + 3, 4))\n",
    "# print(\"\\n\".join(start_regions))\n",
    "for r in [r for r in start_regions if len(r) == start_upstream + 3]:\n",
    "    # print(r)\n",
    "    for i in range(start_upstream + 3):\n",
    "        start_region_bases[i, base_order.index(r[i])] += 1\n",
    "        \n",
    "actual_start_region = start_region_bases / num_of_genes\n",
    "print(\"Sanity check: the last 3 rows should show deterministic ATG\")\n",
    "print(actual_start_region)\n",
    "actual_start_region = np.mean(actual_start_region[:3,:], axis=0)\n",
    "parameters[\"base_frequencies_start_region_upstream\"] = \"{%s}\" % \", \".join([str(p) for p in actual_start_region])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stop region parameters\n",
    "Size is fixed; like for start-region, this has to be done manually (_could_ try some sort of automation, but likely not very good, and not worth the effort)\n",
    " - Compute empirical **positional base frequency marginals**\n",
    " - Compute empirical **codon usage** in and outside stop region\n",
    " - Force UGA to 0\n",
    " - Force UAA to empirical (low) frequency\n",
    " - Find the codon which maximises the difference between codon-wise $stopregion - cds$ and base-wise $stopregion - cds$ (cf `StopCodons.ipynb`) and correct its probability\n",
    " - Find a compensatory codon, and correct its probability (4 codons explicit)\n",
    " \n",
    "The latter part can be extended: Could also pick the top 3 most devious codons, and then 2 compensatory ones etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support. ' +\n",
       "              'Please try Chrome, Safari or Firefox ≥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            if (mpl.ratio != 1) {\n",
       "                fig.send_message(\"set_dpi_ratio\", {'dpi_ratio': mpl.ratio});\n",
       "            }\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        fig.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var backingStore = this.context.backingStorePixelRatio ||\n",
       "\tthis.context.webkitBackingStorePixelRatio ||\n",
       "\tthis.context.mozBackingStorePixelRatio ||\n",
       "\tthis.context.msBackingStorePixelRatio ||\n",
       "\tthis.context.oBackingStorePixelRatio ||\n",
       "\tthis.context.backingStorePixelRatio || 1;\n",
       "\n",
       "    mpl.ratio = (window.devicePixelRatio || 1) / backingStore;\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width * mpl.ratio);\n",
       "        canvas.attr('height', height * mpl.ratio);\n",
       "        canvas.attr('style', 'width: ' + width + 'px; height: ' + height + 'px;');\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option);\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'] / mpl.ratio;\n",
       "    var y0 = (fig.canvas.height - msg['y0']) / mpl.ratio;\n",
       "    var x1 = msg['x1'] / mpl.ratio;\n",
       "    var y1 = (fig.canvas.height - msg['y1']) / mpl.ratio;\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width / mpl.ratio, fig.canvas.height / mpl.ratio);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x * mpl.ratio;\n",
       "    var y = canvas_pos.y * mpl.ratio;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overridden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    var width = fig.canvas.width/mpl.ratio\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\" width=\"' + width + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var width = this.canvas.width/mpl.ratio\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\" width=\"' + width + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        // select the cell after this one\n",
       "        var index = IPython.notebook.find_cell_index(this.cell_info[0]);\n",
       "        IPython.notebook.select(index + 1);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA+cAAAH9CAYAAAB4LA1oAAAgAElEQVR4nOzdeXwV1eH//xuyAWFxQ0AbliqQUJWSKmItSzeFutUdt1o+D/0AfWiAVinV9ottDRI+j1Kt8sFKqbZWKxXjgra2akU+skT8uBusCgZTQf2IJrTKovD+/cEv95HJXXLP3DszZ2Zez8fj/JHJzD0zc2bmzPvMXRICAAAAAACBSgS9AgAAAAAAxB3hHAAAAACAgBHOAQAAAAAIGOEcAAAAAICAEc4BAAAAAAgY4RwAAAAAgIARzgEAAAAACBjhHAAAAACAgBHOAQAAAAAIGOEcAAAAAICAEc4BAAAAAAgY4RwAAAAAgIARzgEAAAAACBjhHAAAAACAgBHOLbd37161tLSotbVVbW1tFAqFQqFQKBQKhUIJoLS2tqqlpUV79+71JPsRzi3X0tKiRCJBoVAoFAqFQqFQKBQLSktLiyfZj3BuudbW1uQBEPRIEYVCoVAoFAqFQqHEtbQ/OG1tbfUk+xHOLdfW1qZEIqG2tragVwUAAAAAYsvrbEY4txzhHAAAAACCRziPOcI5AAAAAASPcB5zhHMAAIDgDf7hw44CIH4I5zFHOAcAAAge4RwA4TzmCOcAAADBI5wDIJzHHOEcAAAgeIRzAITzmCOcAwAABI9wDoBwHnNRDOd0bgAAIGy4fwFAOI85wjkAAEDwuH8BQDiPOcI5AABA8Lh/AUA4jznCOQAAQPC4fwFAOI85wjkAAEDwuH8BQDiPOcI5AABA8Lh/AUA4jznCOQAAQPC4fwFAOI85wjlgB45bAPCPjddb+gEAhPOYI5zDhJt9S1vkhuMWAPxj4/WWfgAA4TzmCOcwQTj3DsctAPjHxust/QC6wjESfYTzmAtDOM8nDHLhKizCuXc4boHgcP7Fj43tzXGIrnCMRB/hPOYI5zBBOPcOxy0QHM6/cItK38RxiK5wjEQf4TzmCOcwEZUbIBtx3MIGcT0O47rdURGVvonjEF3hGIk+wnnMEc5hIio3QDby47il/dCVuF4/U7Z7Xh9ngdWicm2L6/mH3HGMRB/hPOZiEc65ySqYqNwA2YhwDhvE9caPfiPconJti+v5h9xxjEQf4TzmCOcwwQ2QdwjnubOx/aIirvuWfiPcuLYhLjhGoo9wHnOEc5jgBsg7+e7bXJah/dCVuO5b+o1w49pmVx3wDu0XfYTzmCOcw0TeN0AetYXXodYPgYTzHNrDtv0k2dl+URHXfUu/YTmD6xThPPg64B3aL/oI5zFHOIcJwrl3ChLOTW9gCef2sOQ6Fcl9mwP6DcsRzkNVB7xD+0Uf4TzmCOcFEpObOMK5dwjnubOx/fJmSRiM5L7NQeD9RsT7jrwRzu2pg+M2UHG9RscJ4TzmCOcFEpOOinDunciGcw/a3HRf2djeKSy54Q3FvvJA4P1GxPuOvBHO7amD4zZQcb1GxwnhPOYI5wXiR0dlQWdIOPcO4Tx3cQ3ntt64W79vcxB4vxHmOvwQhXDuwTlekH7DdP6oHFMhFYr+DHkhnMdc6MK5m86NcF4whPP/nwWBM+0yIQ3nvm+3jfwI5x7VYf2+zUHg/UaY6/BDFI5bwjkKIBT9GfJCOPfR4sWLNWTIEJWXl6umpkarV6/OOv+qVatUU1Oj8vJyDR06VEuWLEmZZ8WKFaqurlZZWZmqq6vV0NBgtE6E8wIhnOe0jC/h3JKQYyoU4dyj8y+S4dz0GLHkuLX1HPda4P1GmOvwA+G8IPMXpI6oHFMhZdr/IXwI5z655557VFpaqqVLl6qpqUkzZ85URUWFtmzZknb+zZs3q2fPnpo5c6aampq0dOlSlZaWasWKFcl51q5dq+LiYs2fP18bN27U/PnzVVJSovXr1+e8XoTzAolKOA/pjXskw3m+x7oHdfgSzv3YbjdM2zsq4dyD9subJedf3qJShx9C2jc5+BHOLanDD3kPSliyHaYiGc5D2hZeIZz7ZMyYMZo+fbpjWlVVlebOnZt2/jlz5qiqqsoxbdq0aRo7dmzy7/POO0+TJk1yzHPyySdrypQpOa8X4bxAPA6cvl2AQ3oDRDgnnHt53Lo6/ywJ5za2X94sOf/yFsU6vGLBuZE3S4KzjeE87+uOR3XYiHAefYRzH+zevVvFxcUpbzmvra3V+PHj0y4zbtw41dbWOqY1NDSopKREe/bskSRVVlZq0aJFjnkWLVqkQYMGZVyXXbt2qa2tLVlaWloI54XgceAknBe2Dj9CbT7bYNN22BjuCOcRC+de79sc6gi83wjwOpI3+qbceHCOE87tCudujsO86/Dj/POaH+dfiBDOffDOO+8okUhozZo1jul1dXUaPnx42mWGDRumuro6x7Q1a9YokUho69atkqTS0lLdddddjnnuuusulZWVZVyXefPmKZFIpBSbw7kfvO7cUljYuYV5O2wUyQ4UGeUdDtyIyrlR4PndLmPKj+uU7aHIlnCXwsJ95UpIB1d87/9idP75UYdp+9k6uOJ1H+AVwrkP2sP52rVrHdOvv/56jRgxIu0yw4YN0/z58x3Tnn76aSUSCW3btk3S/nB+9913O+b5wx/+oPLy8ozrEsYn537wPdR6gHBuF8J5vBDOc+d3cCac5/b6hPPsInuOE84zsvH8C6QOwrmvCOc+sOlt7Z2F4TPnfiCcE84LjXAeLzZeE2xFOLenDsJ57gjnuSOcm78+dQRfhy0I5z4ZM2aMZsyY4ZhWXV2d9QvhqqurHdOmT5+e8oVwkydPdswzadKkyH0hnB+icMITzu1CWIsX2jt3hHN76iCc545wnjvCufnrU0fwddiCcO6T9p9SW7ZsmZqamjRr1ixVVFSoublZkjR37lxdcsklyfnbf0pt9uzZampq0rJly1J+Sm3NmjUqLi7WggULtHHjRi1YsCCSP6XmhyjcVEf2BiikF2DCWrzQ3rkjnLuswwORvakmnLuro0AI5+avTx3B12ELwrmPFi9erMGDB6usrEw1NTV66qmnkv+79NJLNWHCBMf8q1at0ujRo1VWVqYhQ4ZoyZIlKa957733asSIESotLVVVVZXuu+8+o3UinO8XhZvqqFy4IrsdiDTaO3eEc5d1eCCyN9WEc3d1FAjh3Pz1I1WHB6Jyb5gLwnnMEc73i8JNdVTCQVQuwFFpD+SG9rYL4dz89cMcDlJYEkAI54TzXF8/UnV4ICr3hrkgnMcc4Xw/bqrtEacLMKKDcG4Xwrn564c5HKQIaQAJBOE8sDqicv4RzguLcB5zhPP9ItnhhlScLsCIjsjeuIcU4dz89cMcDlKENIAEgnAeWB1ROf8I54VFOI85wvl+kexwQypOF2BER2Rv3EOKcG7++mEOBylCGkCiwvdw7oeQnhuE8/AhnMcc4Xw/Olx7xOkCjOjgxt0uhHPz1w9zOEgR0gASFYTz3ETl/AsknEcY4TzmCOf7xeWEDwPCOcIoTjcOYeB7OPdIFI4pwnn8EM5zQzi3qw5bEM5jjnC+X1xO+DAgnCOM4nTjEAaEc3sQzuOHcJ4bwrldddiCcB5zhPP94nLChwHhHGEUpxuHMCCc24Nwjkj04VEJziE9N+J0/hHOY45wvl9cTvgwIJwjjOJ04xAGhHN7BHJuhDSARBZ9eFqEc7vqsAXhPOYI5/vF5YQPgzhdgBEdHLd2IZzbg3AOwnl6UQnnfojT+Uc4jznC+X5xOeHDIE4XYEQHx61dCOf2iEo4Rx5oi7QI57mLUx9LOI85wvl+cTnhwyBOF2BEB8etXaISzqOAcA7aIj3Cee7i1McSzmOOcL5fXE74MIjTBRjRwXEbP7R3bgjnQHqE89zFqY8lnMcc4Ry2idMFGNHBcRs/tLfFQhpAEC+E89zFqY8lnMcc4Ry2idMFGNHBcRs/tLfFQhpAEC+E89zFqY8lnMcc4Ry2idMFGNHBcQtYJKQBBPFCOM9dnPpYwnnMEc5hmzhdgBEdHLeARUIaQBAvfn9xJeE8HAjnMUc4h23idAEGAHggpAEE8UI4z12c7g0J5zFHOIdt4nQBBgB4IKQBBPFCOM9dnO4NCecxRziHbeJ0AQYAeCCkAQTxQjjPXZzuDQnnMUc4h23idAEGAHggpAEE8UI4z12c7g0J5zFHOIdt4nQBBgAA8ArhPHwI5zFHOIdt4nQBBgAA8ArhPHwI5zFHOIdt4nQBBgAA8ArhPHwI5zFHOIdt4nQBBgAA8EpUwnmcEM5jjnAO2xDOAQAA8kc4Dx/CecwRzmEbwjkAAED+uKcKH8J5zBHOYRs6EgAAgPxxTxU+hPOYI5zDNnQkAAAA+eOeKnwI5zFHOIdt6EgAAADyxz1V+BDOY45wDtvQkQAAAOSPe6rwIZzHHOEctqEjAQAAyB/3VOFDOI85wjlsQ0cCAACQP+6pwodwHnOEc9iGjgQAACB/3FOFD+E85gjnsA0dCQAAQP64pwofwrkPPvzwQ1188cXq06eP+vTpo4svvlgfffRR1mX27dunefPmaeDAgerevbsmTJigV155Jfn/7du364orrtDw4cPVo0cPVVZW6sorr1Rra6vRuhHOYRs6EgAAgPxxTxU+hHMfTJo0SUcddZTWrl2rtWvX6qijjtKpp56adZkFCxaod+/euu+++/Tyyy/r/PPP18CBA7Vjxw5J0ssvv6yzzjpLDz30kN5880098cQTGjZsmM4++2yjdSOcwzZ0JAAAAPnjnip8COcea2pqUiKR0Pr165PT1q1bp0Qioddeey3tMvv27dOAAQO0YMGC5LRdu3apb9++uvXWWzPW9ac//UllZWX69NNPc14/wjlsQ0cCAACQP+6pwodw7rFly5apb9++KdP79u2r3/72t2mX2bRpkxKJhJ577jnH9NNPP13f+c53Mta1dOlSHXLIIUbrRziHbehIAAAA8sc9VfgQzj1WV1enYcOGpUwfNmyY5s+fn3aZNWvWKJFI6J133nFMv/zyy3XSSSelXeaDDz7QoEGDdO2112Zdn127dqmtrS1ZWlpaCOewCh0JAAAA4ohw7tK8efOUSCSylg0bNqiurk7Dhw9PWf7II4/UDTfckPa128P51q1bHdMvu+wynXzyySnzt7W16fjjj9ekSZO0Z88eV+tNOIctCOcAAACII8K5S//3f/+njRs3Zi07d+70/G3tO3bs0AknnKCvf/3r2rlzZ5frzZNz2I5wDgAAgDginHus/QvhGhsbk9PWr1+f0xfC1dfXJ6ft3r075Qvh2traNHbsWE2YMEEff/yxq/XjM+ewDeEcAAAAcUQ498GkSZN0zDHHaN26dVq3bp2OPvrolJ9SGzFihBoaGpJ/L1iwQH379lVDQ4NefvllXXDBBY6fUtuxY4eOP/54HX300XrzzTe1bdu2ZPnss89yXjfCOWxDOAcAAEAcEc59sH37dl100UXq3bu3evfurYsuukgfffSRY55EIqHbb789+fe+ffs0b948DRgwQOXl5Ro/frxefvnl5P+ffPLJjJ91f+utt3JeN8I5bEM4BwAAQBwRzmOOcA7bEM4BAAAQR4TzmCOcwzaEcwAAAMQR4TzmCOewDeEcAAAAcUQ4jznCOWxDOAcAAEAcEc5jjnAO2xDOAQAAEEeE85gjnMM2hHMAAADEEeE85gjnsA3hHAAAAHFEOI85wjlsQzgHAABAHBHOY45wDtsQzgEAABBHhPOYI5zDNoRzAAAAxBHhPOYI57AN4RwAAABxRDiPOcI5AAAAAASPcB5zhHMAAAAACB7hPOYI5wAAAAAQPMJ5zBHOAQAAACB4hPOYa21tVSKRUEtLi9ra2igUCoVCoVAoFAqFEkBpaWlRIpFQa2urJ9mPcG659gOAQqFQKBQKhUKhUCjBl5aWFk+yH+Hccnv37lVLS4taW1sDHynKdSTJ5Cm/6TLU4V0dNq4TdYR7naJSh43rRB3hXifqCPc6UUe414k6wr1OQZfW1la1tLRo7969nmQ/wjkKpq3N/DMYpstQh3d12LhO1BHudYpKHTauE3WEe52oI9zrRB3hXifqCPc6RR3hHAVj6wlMHeFdJ+oI9zpFpQ4b14k6wr1O1BHudaKOcK8TdYR7naKOcI6CsfUEpo7wrhN1hHudolKHjetEHeFeJ+oI9zpRR7jXiTrCvU5RRzhHwezatUvz5s3Trl27PFuGOryrw8Z1oo5wr1NU6rBxnagj3OtEHeFeJ+oI9zpRR7jXKeoI5wAAAAAABIxwDgAAAABAwAjnAAAAAAAEjHAOAAAAAEDACOcw9rvf/c7Vlza0tbVp7969KdM/++wzV9/Q+P777xsv09GWLVu0b9++vF7DBj/96U/18ccfB7oOL730kmbOnJnXa0RlOwAAAAA3COcw1q1bN7333ntGyzQ0NGjYsGFpw9fHH3+s4cOH66GHHurydfbt26dHHnlEZ555psrKyhz/++STT7Ry5crk33PnztXs2bOT5aqrrtLOnTvz2o4PP/xQv/rVr9IOJrS2tmb8Xzp79+7VQw89pDPOOCPrfDt27FBbW1uy/Otf/3L833Q7Xn/9dU2ZMiXjNlxwwQXatGlTl6/T1tamW2+9Vccdd5yKioo0atSolP+3e+SRR/Tggw8my8MPP5zyeqbb8cknn+jBBx/Ujh070q7bgw8+mNMgUlfbYeqrX/2qPvroo7xeI18ffPCBfvnLX+b1GlEZvLKhPQph6tSpaY/1MHEzsFvoQV0pv4FdW9rh+eefz2v5oUOH6oMPPjBertDt8emnnxov05Et16l82+Opp57Ke18UQlTaI1/0GwgS4RzGioqKjEPtN7/5TS1dujTj/5ctW6aTTjop4/83bdqka6+9Vp/73Od0wAEH6KKLLlJDQ4NjnltvvVWnnnpq8u9evXrp+OOP18SJEzVx4kQNGDBAixYtyms7fvazn+mcc87J+P9zzz1X119/fdbXeP311zV37lwNHDhQ3bt3Twnnzz//vL71rW85tqNbt27JUlxcrGeeecb1dlx++eW6+uqrM/5/zpw5mj59esb/r1q1Spdccol69uypbt266Yc//KHeeOMNxzwrV67UF7/4Rcc2FBUVJUu3bt107733OpYx3Y4bb7xRX/va1zL+/+tf/7puueWWvLbjzTff1NSpU5N/V1ZW6sADD0yWQw45RK+99lpe2yEVZtBn3759evTRR3XuueeqrKxMhxxyiNE6dGY6WFLIgSspt8GrrgauJPP2eOKJJ1RdXZ1xO0aOHKnVq1fn/Hrt+6XjoI/pQKIUbHsENZBYqEFdKfPArh+DuoU6plpbW7V48WKNHj1a3bp1c/zPdDvcXKcK2R6vvvqqZs+erUMPPdQx/aOPPtJ///d/J/++8MILdeaZZybLOeec4whNpu1RyPM7U3v4cX4XelsytcfevXv10ksvJf9esmSJbrrppmS55ZZbHIM1YXjoQb/Rdb229eNxQziHsaKiIuMnDwMHDkwJPh298cYbGjhwoGPazp07deedd2rChAkqLy/XqaeequLiYr388stpX2PcuHGOwN6rVy/HE+A777xTY8eOdWyHaScyatQoPf744xn///jjjztCabtPPvlEd9xxh8aNG6fS0lJ169ZNN910U9qL0H/8x39o/vz5ju246667tGrVKj355JO65JJLdPHFFzu2w6Q9RowY4Qj3nT377LMaPny4Y9rWrVtVV1enI444QgMGDNDs2bO1YcMGlZSU6NVXX015jdNOO02/+c1vHNvQsS3q6+s1efJkxzKm23HcccdlvRFcuXKljjvuuLy2Y+bMmfrRj37k2I6FCxfqjjvu0B133KHJkydr2rRpKdvh56DPW2+9pZ/85CeqrKxUt27ddMkll+ixxx7TZ5995pjPdKDBdDsKMXAlZR+8Mh24crMdp512mmMQr7ObbrpJ3/72t7t8nccee0xTpkxR9+7d9bnPfU61tbXJ/5kOJLrZjigMJOY7qCt1PbDrx6BuvsfUE088oYsuukg9evRQVVWVrr32Wj333HOOefzYjnzb41//+peWLl2qsWPHqri4WCeeeGLKflm4cKEuuugix3acffbZ+u53v6vvfve7GjFihObNm+d6OwpxfnfVHn6c34XYllza46677tL48eMd2/K5z31OQ4YM0ZAhQ9SrVy9HX2/jQw/6jfD343FDOIexoqIifetb33KMZqcrHXXv3l0bN27M+JpNTU3q3r178u8ZM2bowAMP1NixY3XLLbck336XKURJUv/+/fXKK68k/z7kkEP01ltvJf/+xz/+oT59+ji2o66uzjEKnK501KtXL23ZsiXjdmzZskW9e/dO/t3Y2KjLL79cffr00bHHHqsbb7xR7777btbtGDFihGOEtXOwXb9+vQYNGuTYjqOPPlqjR4/OWtp1795dzc3NGbehublZPXr0cEwrLy/XxRdfrEcffdQxSp5pOwYPHqwNGzZk3IaXXnpJ/fr1cyxjuh0HHHBAl21xwAEH5LUdX/jCF/T3v/8943asWrVKRx55ZMp2vPnmm46R4HSlI9NBn127dunuu+/W1772NXXv3l1nnnmm7r333qzHlelAg2mn7nbgSsp98Mp04Kp9O0zaY9CgQWpqasq4HRs3blRlZWXa/23ZskXXXXedBg8erIMPPljdunXTihUrUuYzHUhs3w6TwasoDCS6GdSVzAZ23Qzqmg5OuzmmWlpa9POf/1xDhw7VoYceqiuuuCLr+e1mO5588km9+OKLWUtHbtvjf/7nf3TppZeqV69eOvroo1VcXKynn3467WuMGTNGjzzySMbtaGhocBy3ptcpt+e3SXv4cX7nsy0m7fGNb3xDd999d8ZtWbJkiSZOnOjYDtseetBveN9vSN62R9wQzmGsqKhI559/fnIkO1PpqKqqSnfeeWfG1/z973+vESNGJP8uLi7WNddck/JZmWw3J927d095i3FHGzduVHl5uWM7KisrkyPA6crQoUMdr9G3b1+tW7cuYx3r1q1T3759Hdsxa9aslPXKth09e/Z0hOdFixY5OoEtW7akbMdVV12l6667Lmtp179/fz3xxBMZt+Hxxx9X//79HdOGDx+uIUOG6JprrnEMsmTajvLycm3evDn594YNG7Rnz57k35s3b075zgDT7ejVq5eeffbZjNvx7LPPqlevXnltR69evRwDPLNmzXJ8TrO5udkxqNS+HR1HgTuX9v93rsdk0Ofggw/WuHHj9Otf/1offvhhl9shmQ80mA5emW6DZD54ZTpw1b4dJu1RXl7eZQDp3ObLly/XN7/5TfXs2VPnnHOOHnjgAe3evTvjdpgOJLZvxwEHHOB4t0O60nHf2DiQaDKwazqoK5kP7LoZ1DVpB8n8mJo8ebJ69+6tCy64QA8//HDyXTCFHpxuP/47l0zXKdP2qK+v14gRI3T44Yfrqquu0gsvvNDldhx88MH6xz/+kfz7S1/6klpaWpJ/b9q0SRUVFY7tMLlOuTm/TdvD7fk9bdo0x9uU05WOTLfFTXscfvjhyfmk1HO8qanJcbzb+NCDfsO7fkPypz3ihnAOY25GRq+55hoNGjRI7777bsr/tm3bpkGDBumaa65JTrvrrrv0jW98QxUVFTrvvPO0cuVKffrpp1k7kSOPPDLtaGO75cuX64gjjshrOyZOnKgf/vCHGf8/Z84cxyjyN7/5TfXu3VsXXnih/vKXvyS/KCXbdhx44IEZR7El6emnn07pDE2249xzz8369qrTTz897duann76aU2dOlW9evVSTU2NFi1apJKSkrSjxQMHDtRjjz2WsY6//vWvGjBggGOa6XYcf/zxWrBgQcb/33DDDTr++OPz2o4+ffqosbExYx2NjY0pHVVRUZEaGhq0atWqrKUj00GfAw44QOPHj9dtt93mGLjJdlyZDjSYDl6ZboNkPnhlOnDVvh0m7fH5z38+5fssOrrvvvtSBu2Ki4v1ox/9KOfBRNOBxPbtuOmmm5LvdMhU2tk6kGgysGs6qNu+HSYDu24GdU3aQTI/poqLizV79my9/vrrOW2D2+3YsGGDmpubs5aO3A6yd/6ITbbt6NGjR8aPrkn733XV8Z1dptcpt+e3SXu4Pb+//OUvJ9+inK589atfdSzj5rgybY/y8nK9+eabyb/ff/99x7vO3njjDcdAu40PPeg3vOs32rfD6/aIG8I5jLkJtTt27NAXvvAF9e7dWzNmzNCNN96om266SdOnT1fv3r01cuTItN8o+dZbb+n//b//p0GDBumQQw5J+0Vi7WprazVy5MiUL8SQ9r/dZuTIkY7P77j54pIVK1aopKREN998s6OD++yzz/SrX/1KpaWlKev39ttv66c//amGDBmi/v37q7a2NmMYlKSvfe1ruuqqqzKuw/e//33HF6GZbsdzzz2n8vJynX322WpsbFRra6taW1u1fv16nXXWWSovL9f//u//Zlz+X//6l2677TaNHTtWRUVFmjhxom677TbHW6fOP/98nXbaaRlf45RTTtF5553nmGa6Hb/+9a9VUVHh+IKUdg899JAqKir061//Oq/tOOGEE1RXV5fxNX72s5/phBNOcEzzY9Bn586d+sMf/qCvfvWr6tGjh8466yw1NDSotLQ0402W6UCD6XaYboNkPnhlOnDlZjuuuOIKHXXUURmvI0cddZSuvPJKx/TLL79cffv21Ze//GUtWbIk+W6GTNthOpDoZjuiMJBoOqgrmQ/s+jGoa3pMrV27Vpdddpn69OmjMWPG6Oabb9b7778f+OC0aXvU1dVp2LBhqqys1Jw5c5Khu6t3+Pzud7/LuA6//e1vNXLkSNfb4eb8Nm0PP85vN9vipj0GDRrk+JhBZw899FDKu2Nse+hBvxH+fjxuCOcw5ubiK+3/pscZM2booIMOSr597qCDDtKMGTO6/MmKffv26S9/+YvOPfdclZeX6/DDD0+50L377rsaMGCABg0apIULF+qBBx7Qgw8+qPr6elVWVmrgwIGOmwq323HNNdeoqKhIffr00Re/+EWNHj1afVQ4sVgAACAASURBVPr0SX7jdzZ/+9vfkl/2MWzYMP3oRz9KCcLtAwCdvwU10wBAtu3I9E2ZK1euVL9+/VLeqtWvXz89+OCDOe+LpqYm/eAHP9Chhx6qkpKS5PT2AYBzzjlHzzzzTHIAoLGxMeMAgJv2uOiii1RUVKTq6mp9+9vf1plnnqmqqip169ZNU6ZMyXs7brvtNvXs2TPtT7899NBD6tmzp2677ba8t8PNoE+7N998M/mFV0VFRbrwwgv1t7/9LeXpiOlAg+lgidttMBm8Mh24knJrj44/H/Tuu+/qsMMOU2Vlperr65PXkQULFqiyslKHHXZY2nDS/nm78ePHq7y8XKeffnrGzzmbDiRK/rSH1wOJpueG20FdKfeBXT8Gdd0eUx9//LGWLVumE088MfkZzhtvvDHtNptuRy5t0fmzqm7bY9WqVfrOd76jiooKHXPMMVk/4/zjH/9YlZWV2rZtW8r/tm7dqsrKSl177bXJaabt4bYtpNzbw8357abfcLstJu0xdepUffnLX077v3379umEE05wfNGojQ896DfC34/HDeEcxoqKihyfActFxwvEvn379P777+u9997L+HuY2S4o27dv1y9/+Usdc8wxKf/bvHmzTj75ZMdn6bp166aTTz455be7r7vuurQ/CZOLxsZG1dbW6lvf+pYmT56smTNnZn0q2Vn7z1V88YtfTPlcn7R/dDLTAEDni1pzc3PK78529S3L0v6LeUNDgxYuXKj6+nrdf//9rvfHp59+qvvuu88x7YEHHkjeFHcsBx98sO6///6U12hubnYcD6+88orji4k6ftaqo+XLl+uMM87QyJEjVV1drTPOOEPLly8v2HZMmTIlZQCgurpa3bp107nnnpvyGkOGDNE///lP47pNBn3S/V703r179ec//1lnn322ysrKdPDBBzv+bzrQ4PbJmtuBK6nrwSvTgStpf3tk+j3nTD8f1NzcrMmTJ6dcRyZPnuz4aEAm7effYYcdpj59+uiCCy5wHFemA4mS/+3h90BiJvkM6kpdD+z6Naib7zH12muv6eqrr9aAAQPUvXv3lHcmmW7HxIkT0+6/TD851y6f9tixY4eWLFmiMWPGqLi4WCeccIJ+8YtfpMxTXV2t3r1763vf+15yAGDGjBnq3bu3qqqqHGHYTXvk2xZS9vbw6/zOd1tyaY8333wz+Y6BP/3pT3rhhRf04osvavny5TruuOPUp08fx2etbXzoQb8R/n48bgjnMObmZDddxu0Fvt327dvV2NioxsZGbd++Pe08r7/+uqZMmZLx9xwvuOCClEDvhUxvIV+3bp1qa2s1efJkTZ48WbW1tVk/C2TyLcuZ5Pp7xrku8/HHH6uhoUH19fWqr69XQ0OD/v3vf6d9ndWrV+vYY49N/t3+8xodO7lsn2N3I9cBgD/+8Y8644wzVF1drerqap1++un64x//mHZeN08O2sN2roM+XdXx/vvvp9xkSWYDDW4Hr/IduJKyD16ZDFylk8vPB3Vcj2eeeUaNjY2OL97r+MQkm47nRuegYzKQ2K7jteqRRx7Rgw8+mCzpBl0kuwYSTQd2TQd1Oy/TWaaBXZO2+PTTT/Xiiy/qk08+SXn9jz/+WC+++GLKYGlHmY6pXH322We6//77035saNOmTcbHVMdls/3kXGeZ2iPXb4Z+6aWXNHPmzJRf7ZD276Np06bpwAMPTG7HgQceqGnTpqX05/kMsufbFlLm9jA9v90M6g4dOjQZIHPZlo7zd5atPRobG1VdXe34krT2fmT9+vWOeW196EG/sZ9f/YbkbXtEHeEcxmwN56ah6D//8z919dVXZ/z/nDlzNH36dMc000DvxwCAm29Z7iyXJ+2FWEbKHObPP//8lG//fuqpp9Tc3Ky33npLs2fP1llnnZXxdXMJ2n4MAPhx7OYzeJXrQIPNg1emA1eS2c8HZZLpiUkusgXGdAOJnW/kVq5c6fgJm169eqV8w7bXTxvyHUj0ox/I59zIZVD39ttv15e+9KWUj41I+4Pal770pZQvTfvss8+MAn22+T/55BO9+OKLaes32Q7J7CfnsunqaXs22cL8vn379N5772UdANi6davjc+4nnnii46c3jz32WFfvZCqUXNvCzaCuF+dTx19V6ez555/X8uXLtXz5csfvundEv5Eq7v2GVLj2iAvCOYwVFRXp97//vWP0LV3JZxm3dZh0VMOHD9czzzyT8f/PPvushg8f7ph2+eWXGwV60/kl887NzbcsS+6etOfzdL6rMH/EEUc4Lsydf17jueeec/yGrpugPWXKFM8HAIqKzH+v1s1Nlmkd6d4Kn43p4JWbmzI/buTc/HxQRyZPTNJ59dVX9f3vfz/nG7NMN3KnnnqqfvOb3yT/7nx+1NfXa/LkyY5lTJ+YZJs/25dC5crGcG4aik488cSM75qR9n/MZty4cY5ppoHejwEA05+cS8f0aXu79jB/1lln5RzmMw0A/OQnP9H3vve95N+9evVSbW1t8qc3jz/+eP3gBz9I/t9NmO9qmeOOOy7vAQBbH3pke9qeTlQeetBv7Jfrk3av+464IZzDWMdRt0yl81tYTJdxW4dJx9O9e/eUn4rpqLm52fGTLdL+32c0CfSm80vmgd70mzLdPGl3+3TeJMx3bo/77rvP8fa45uZmx02Zm6DtxwBAUZHZbzm3L2MStt3UYRpCTAev3AxEmS7j5qbMzc8HSfk9MTG9Mctl/sGDB2vDhg3Jvzsfuy+99JLjLammT0zcPmExuZHza5DWtA6T86Jfv35ZPz+6efNmHXLIIY5pX/nKV4wCven8knmgN/3JuXb5PG13E+a7WmbUqFH629/+lvy783nx6KOPOr7d/cc//rFRmHezjJsBgKKiIj355JOOwd90JZ9l3NYR9oce9Bve9Rtul3EzABAnhHMYs3WE1/SmrH///nriiScyvt7jjz+u/v37O6aZBno/BgAks2/KdPOk3XQZN2G+X79+evLJJzNu95NPPum46TUN2pI/AwBFRWa/5dy+jEnYdluHl4NXbo5b02XcDACY/nxQPk9MTG/MTOYvLy/X5s2bk39v2LDB8RbUzZs3O47d0047zeiJien8kvlNmV+DtKZ1mJwXPXv2TAkwHb344ovq2bOnY5ppoPdjAMD0J+ckd0/b3YR5k2X69u3rOE7PPPNMx5divfXWW47rlGmYd7OMmwGA9uMy0/Ga6Vg3WcZtHWF/6EG/4V2/4WYZW95qbzPCOYy5+WyU6TJuP39lclN27rnn6tvf/nbG1zv99NN1zjnnOKaZBno/BgA66+qbMt38nrHpMm4GAE499VTHT7J0dumll+qUU05J/m0atCV/BgDcDiyZhG23dZg8nTc9dt0ct34MALTL9eeD3DwxMb0xc3MjN3DgwKzfh/DXv/5VAwYMSP5t+sTEdH7J/KbMr0Fa0zpMBnVHjRqlJUuWZHy9xYsXa9SoUY5ppoHejwGAdrn+5Jxk/rTdTZg3XaaioiLj55+l/dfoioqK5N+mYd7NMm4GAIqKirRhwwY1NzdnLfks47YOk6ftNj70oN/wrt9ws4ybAYC4IZzDmM1Pzk2Waf8t7rPPPluNjY3J3+Jev359xt/iNg30fgwATJ06Ne1vy2b7pkzT3zM2XcbNAMDf//735Dd1dmzH9957T9///vdVXFzs2C+mQVvyZwDAjy/2cVuHydN502PXzU2ZHwMAnXX180GmT0wk8xszNzdy559/ftpv6G53yimn6Lzzzkv+bfrExHR+yfymLJfj9vnnn3c9v5tlTAd16+vrdfDBB6cNzy+88IIOPvhg1dfXO6abBno/BgA66+on5yTzp+1u3jpvukxNTY1uueWWjNt10003afTo0cm/TcO8m2XcDADYfF9l8rTdxoceheg32tratGTJEh133HHq1q0b/UYHpsu4GQCIG8I5jH33u99NGwYLuYybOtwElpUrV6pfv34pv8Xdr1+/lM8ySuaB3o8BgFy2O9O3LEu5/Z6xm2XcDAAsXrxYZWVl6tatmw444AAdeOCB6tatm8rKynTzzTc75jUN2pI/AwCmT6gld9/W7uZL50yezpseu25uyvwYAMjm5Zdf1qxZs9LeCOT6xEQyvzFzcyPX3h7nnHOOnnnmmWR7NDY2pm0P0ycmpvNL5jdlmcJBa2urFi9erNGjR+f0lvNM8xeyjkz27NmjiRMnqqSkRJMmTdKsWbM0e/ZsTZo0SSUlJRo/fnzKN16bBno/BgCy2b59u2688Ub17ds37f9zfdru5q3zpsssXLhQBx10UMZ9ddBBB2nhwoXJaaZh3s0ybgYAbA7nJk/bbXzoQb/hXb/hZhk3AwBxQziHMZvf1m6yTPsT508++UQNDQ1auHCh6uvrdf/992f9nU7TQO/1AIDb7e4s25N2N8t0ZDIA8Pbbb2vRokWaMWOGZsyYoUWLFuntt99Omc80aLfzegDgiSee0BFHHJHxy2dGjhyp1atXO6abhm23dXg5eOXmpsyPAYAnnnhC1dXVGfdVdXV11huLrp60d2RyY+Zm/gceeCAZhjqWgw8+WPfff79jXtMnJqbzS+Y3ZZ0HXZ944glddNFF6tGjh6qqqnTttdc6Qo3p/G6WcfN0fs+ePaqvr9eoUaPUs2dP9ejRQ6NGjVJ9fb327NmTdn6TQO/HAEAm2QY+Osvlabtk9tZ502X27Nmj8ePHq6SkRJMnT07uq8mTJ6ukpERf+cpXHPvKNMy7WcbNAMCYMWMcX9g3d+5czZ49O1muuuoq7dy5M69l3NRRqH6jqKgosIcebvqNTz75RCtXrsy6r7I9QOrqSXtHQfQbRUVFBek33CzjZgAgbgjnMGbrCK/p03Y3AwCmgd6PAQDTYOfmSbvpMoUcAMjGJGh35OUAwOmnn571G1ZvuummlBsF07Dtpg43x7sko2PXdCDKdBk3AwCnnXaa8b7KJNsTk45MAr3p/B9//LEaGhpUX1+v+vp6NTQ06N///nfKfKZPTEznl9zdyLW0tOjnP/+5hg4dqkMPPVRXXHFF1ic/pvObLuPm6Xw6H330kRYvXqyampq087sJ9F4OAHSWy8BHNl09bZdyD/Omy+zevVs33HCDRo0apR49eqhHjx465phjdMMNN2j37t2OfWUa5t0s42YA4NZbb9Wpp56a/LtXr146/vjjNXHiRE2cOFEDBgxIuY6ZLuOmDjf3YpJZvyGZB3qT+d30G13tq4EDB+b802i59hsmgd50fq/6DTfLuOk34oZwDmNu31Jr+lNRfrw12MZ3AJgG+qKiomQwzVY6zm/DZ6LdDACkk2vQlvwZAKisrMz61v2NGzeqsrLSMc00bLupw83T+WxPnNM9nZfMb8pMlzEdABg0aJDRvsr3SXtnud6YdTW/m/YwedLuZn7Tm7LJkyerd+/euuCCC/Twww8nPzuZKTibzu9mGTdP5zvKJ9R2FehN5zcN9G4GPtIxHciQcgvz+SyTbV+ZhHk3y7gZABg3bpzjJ+I6fw73zjvv1NixY/Naxk0dpk/bc3ni3PnpfMdlTfqOfPuNbOHfdF/l+6S9s0L1G27aw+RJu5tl3AwAxA3hHMZMw6CbZdzW4cfndm0bACgqKtJNN92kO+64I2vpOL8fgyt+fSzBhB8DAOXl5XrjjTcyvtYbb7yh7t27O6aZhm03dZg+nTd94uwmPPoxAGC6r9w8aTcN9G4GANy+AyDXJyZu5zcJ9MXFxZo9e7Zef/11x/RsX4BkMr/bZfx4mt+RDQMAbgY+CrVebsJ8rsu4XSfTgZKuljEdAOjfv79eeeWV5N+HHHKI49v3//GPf6hPnz55LeOmDj+ezpsGSLcDACb9hum+cvOk3TTQuxkAcNMeknk/YLqMmwGAOCGcw5hpGHSzjNs6TAOkmwEAr98B4PUAgF+DK14PAPj5vQQmPv/5zztG3Du77777NHToUMc00wDppg7Tp/OmT5zdhEc/BgBM95XpdrvZDjf7qtDvAOi8r9wOlEi535StXbtWl112mfr06aMxY8bo5ptv1vvvv58xEJrO72YZP57mS/YNALgZxMh3vdwE51yWsXGd0skW5rt3757y06Mdbdy4UeXl5Xkt46YOP57Oez0A4CbMm+4rL7a7c6B3MwDgxTsAOu8rt4MlbgYA4oJwDmO2fubcTYB0MwDgxzsAvAypfg2u+DEA4PVvfbsZALjiiit01FFHZRy5P+qoo1I+M2kaIN3U4fXTeTeh1o8BANN95eZdCabb4WZfef0OAC/eMZAp0H/88cdatmyZTjzxRJWWlqpbt2668cYbMw6Emc5vsowfT/NtHABwM/DhZr28/s4AG9eps1zC/JFHHqkVK1ZkfI3ly5friCOOyGsZN3X48XTe6wEAN0+PTfeVjdvtZr38+O6DfD76EBeEcxiz+dvaTQOkH3X4MQBg22/I+zUA4ObpvNcDAO+++64OO+wwVVZWqr6+Xg888IAefPBBLViwQJWVlTrssMMcv3krmQdIN3V4/XTeTaj1YwDAdF/58c4HN/vK63cA+PGOgXRee+01XX311RowYIC6d++e9YuC3Mzf1TJ+PM23cQCgncnAh+l6+fGdATauk2Qe5mtrazVy5MiMfcDIkSNVW1ub1zJu6vDj6bzXAwBuQq3pvrJxu92slx/ffeD2rfZxQjiHMZufnNv2c21+DACY8mNwxa9jxM3Tea8HACSpublZkydPTn6OqqioSN26ddPkyZMdnWk7N2HbtA6vn867CbV+DABIZvvKj3c+uNlXXr8DwI93DGTz2Wef6f77788pbLuZv6tlvHyab+MAQDpdDXyYrpcf3xlg4zq5CfPvvvuuBgwYoEGDBmnhwoXJPqC+vl6VlZUaOHBgSh9guoybOvx4Ou/1AICbUGu6r2zcbjfr5cd3H7gZAIgbwjkiIyrf1u72p0u8fH0/Bj78enu+1wMAHX344Yd65pln1NjYqA8//DBrXaZh27QOr5/Ouwm1fgwAdJTLvvLjnQ9u9pXX7wDw4x0DYeHV03ybBgCy6WrgI9f18uM7A2xcJ7cDJZs3b9bJJ5+c0gecfPLJjvCSzzKm8/vxdN7rAQA3oVYy21c2breb9fLjuw/cDADEDeEckeF1qHVThx8DADay9d0VfgwA5Msk0Jvw+um8m9f3YwDADa/f+eBmX5mulx8DBvkOltjOy6f5tgwA5CuX9fLyOwNsXKd8B0q2b9+uxsZGNTY2avv27V3O72aZXOf34+m81wMAbkJtR7nsKxu32816+fHdB24HS+KEcA5Yxo9Bhiiw9WMJNvP66byb1/d6ACAfXr7zwW1b5LpefgwY+DVYEmU2DQDkI9f1KvR3Bti8Tn4PlHjF66fzXg8AuAm1fuwnPwY+TNfLj+8+yHewJA4I5wBCyY/gHIV3MaTj1dP5fF7fywEAP5luu5dt4fWAgd+DJXDHzefzvVbo7wywfZ38HCjxildP5yVvBwDchlq3bNluN+vlx3cf+DVYEmaEcwDIIGpPzqPE6wGGKPFywMD2wRLAJjYOlNjEqwGAfEKtH7wc+DDl9Xcf+D1YEkaEcwAAkBcGSwDYzstQGzVefveB7YMlQSOcAwAAAAB8w2BJeoRzAAAAAAACRjgHAAAAACBghHMAAAAAAAJGOAcAAAAAIGCEcwAAAAAAAkY4BwAAAAAgYIRzAAAAAAACRjgHAAAAACBghHMAAAAAAAJGOAcAAAAAIGCEcwAAAAAAAkY4BwAAAAAgYIRzAAAAAAACRjgHAAAAACBghHMAAAAAAAJGOAcAAAAAIGCEc8vt3btXLS0tam1tVVtbG4VCoVAoFAqFQqFQAiitra1qaWnR3r17Pcl+hHPLtbS0KJFIUCgUCoVCoVAoFArFgtLS0uJJ9iOcW661tTV5AAQ9UkShUCgUCoVCoVAocS3tD05bW1s9yX6Ec8u1tbUpkUiora0t6FUBAAAAgNjyOpsRzi1HOAcAAACA4BHOY45wDgAAAADBI5zHHOEcAACE2eAfPuwoABBWhPOYI5wDAIAwI5wDiArCecwRzgEAQJgRzgFEBeE85gjnAAAgzAjnAKKCcB5zhHMAABBmhHMAUUE4jznCOQAACDPCOYCoIJzHHOEcAACEGeEcQFQQzmOOcA4AAMKMcA4gKgjnMUc4BwAAYUY4BxAVhPOYI5wDAIAwI5wDiArCecwRzgEAQJgRzgFEBeE85gjnAAAgzAjnAKKCcB5zhHMAABBmhHMAUUE4L5DFixdryJAhKi8vV01NjVavXp11/lWrVqmmpkbl5eUaOnSolixZ4vj/7bffrkQikVJ27tyZV72dEc4BAECYEc4BRAXhvADuuecelZaWaunSpWpqatLMmTNVUVGhLVu2pJ1/8+bN6tmzp2bOnKmmpiYtXbpUpaWlWrFiRXKe22+/XX369NG2bdscJZ960yGcAwCAMCOcA4gKwnkBjBkzRtOnT3dMq6qq0ty5c9POP2fOHFVVVTmmTZs2TWPHjk3+ffvtt6tv374FrTcdwjkAAAgzwjmAqCCc52n37t0qLi5WQ0ODY3ptba3Gjx+fdplx48aptrbWMa2hoUElJSXas2ePpP3hvLi4WIMGDdLhhx+uU045Rc8991xe9UrSrl271NbWliwtLS2EcwAAEFqEcwBRQTjP0zvvvKNEIqE1a9Y4ptfV1Wn48OFplxk2bJjq6uoc09asWaNEIqGtW7dKktatW6c777xTL7zwglavXq2zzz5bPXr00Ouvv+66XkmaN29e2s+yE84BAEAYEc4BRAXhPE/tIXnt2rWO6ddff71GjBiRdplhw4Zp/vz5jmlPP/20EolEyufK2+3du1ejRo3SlVde6bpeiSfnAAAgWgjnAKKCcJ4nr97Wns5ll12mSZMmua43HT5zDgAAwoxwDiAqCOcFMGbMGM2YMcMxrbq6OusXwlVXVzumTZ8+3fGFcJ3t27dPxx57rKZOneq63nQI5wAAIMwI5wCignBeAO0/abZs2TI1NTVp1qxZqqioUHNzsyRp7ty5uuSSS5Lzt/+U2uzZs9XU1KRly5al/JTaddddp0cffVSbNm3S888/r6lTp6qkpESNjY0515sLwjkAAAgzwjmAqCCcF8jixYs1ePBglZWVqaamRk899VTyf5deeqkmTJjgmH/VqlUaPXq0ysrKNGTIEC1ZssTx/1mzZmnQoEEqKytTv379dNJJJ6V8vryrenNBOAcAAGFGOAcQFYTzmCOcAwCAMCOcA4gKwnnMEc4BAECYEc4BRAXhPOYI5wAAIMwI5wCignAec4RzAAAQZoRzAFFBOI85wjkAAAgzwjmAqCCcxxzhHAAAhFlUwnlUtgOAe4TzmCOcAwCAMItKqI3KdgBwj3Aec4RzAAAQZlEJtVHZDgDuEc5jjnAOAADCLCqhNirbAcA9wnnMEc4BAECYRSXURmU7ALhHOI85wjkAAAizqITaqGwHAPcI5zFHOAcAAGEWlVAble0A4B7hPOYI5wAAIMyiEmqjsh0A3COcxxzhHAAA+GZeH2cpgKiE2qhsBwD3COcxF4ZwTkcFAEBEEM4zisp2AHCPcB5zhHMAAOAbwnlGUdkOAO4RzmOOcA4AiLwCh0HkgXCeUVS2A4B7hPOYI5wDACKPcG4PP8K5B3X4gXCOQuOYCh/CecwRzgEAkRfCoBZZhPOMCFIoNI6p8CGcxxzhHAAQeSEMapFFOM+IIIVC45gKH8J5zBHOAQCR53VQC2kYDAThPCOCFAqNYyp8COcxRziHCS7yAEKJcG4PwnlG9LEoNI6p8CGcxxzhHCa4yAMIJcK5PQjnGdHHotA4psKHcB5zhHOY4CIPIJQI5/awJJzb2JfRx6LQOKbCh3Aec4RzmOAiDyAMsoY1LxDOc0c4z4g+FoXGMRU+hPOYI5zDBBd5AGFAOLcY4Twj+lgUGsdU+BDOYy6K4ZwLkXei8rk+ANFGOLcY4Twj7l9QaBxT4UM4jznCOUwQzgGEAeHcYoTzjPy4f4nKPVIUtsEPUWnvOCGcx1zowrlhh8uFqLAI5wDCgHBuMcJ5RoTz3EVhG1L4cW7AeoTzmCOcwwTh3C4c60B6hHOLEc4zIpznLgrbkIJwDhHOY49wDhOBhHNuejPiWAfSI5xbjHCeUSDhPKTHrY3tlzfCOUQ4L5jFixdryJAhKi8vV01NjVavXp11/lWrVqmmpkbl5eUaOnSolixZ4vj/bbfdpq985Ss64IADdMABB+jrX/+6GhsbHfPMmzdPiUTCUfr372+03oRzmCCc24VjHUiPcG4PP/oNwnkedYT0uLWx/fJGOIcI5wVxzz33qLS0VEuXLlVTU5NmzpypiooKbdmyJe38mzdvVs+ePTVz5kw1NTVp6dKlKi0t1YoVK5LzXHjhhVq8eLGef/55bdy4UVOnTlXfvn31z3/+MznPvHnz9IUvfEHbtm1Llvfff99o3QnnaXCTlRHh3C50ukB6hHN7EM5zRzjPnY3tZ8rWcwPBIpwXwJgxYzR9+nTHtKqqKs2dOzft/HPmzFFVVZVj2rRp0zR27NiMdXz22Wfq3bu3fve73yWnzZs3T6NGjcpjzQnnaXHhyohwbhfCOZAe4dwetgYQG6+dhPPc2dh+pmw9NxAswnmedu/ereLiYjU0NDim19bWavz48WmXGTdunGprax3TGhoaVFJSoj179qRdZseOHerevbtWrlyZnDZv3jz17NlTAwcO1JAhQ3T++edr06ZNWdd3165damtrS5aWlhbCeWdcuDIinNuFcA6kRzi3h60BxMZrZ2TDuQd12Nh+pmw9NxAswnme3nnnHSUSCa1Zs8Yxva6uTsOHD0+7zLBhw1RXV+eYtmbNGiUSCW3dujXtMt/73vd0xBFHaOfOnclpf/7zn7VixQq99NJLeuyxxzRhwgT1799fH3zwQcb1Tfc5dcJ5J1y4MiKc24VwDqRHOLeHrQHExmsnOdr14AAAIABJREFU4Tx3NrafKVvPDQSLcJ6n9nC+du1ax/Trr79eI0aMSLvMsGHDNH/+fMe0p59+WolEQtu2bUuZv76+XgceeKBefPHFrOvy73//W/3799cvfvGLjPPw5DwHXLgyIpzbhXAOpEc4t4etAcTGayfhPHc2tp8pW88NBItwniev39b+X//1X+rbt682bNiQ0/p84xvfSPn8ezZ85jwNLlwZEc7tQjgH0rMxnMf1fLU1gNjYFoTz3NnYfqZsPTcQLMJ5AYwZM0YzZsxwTKuurs76hXDV1dWOadOnT0/5QriFCxeqT58+WrduXU7rsWvXLh1++OH66U9/mvO6E87d1RFXhHO7xPVmH+gK4dwetgYQG9uCcJ47G9vPlK3nBoJFOC+A9p9SW7ZsmZqamjRr1ixVVFSoublZkjR37lxdcsklyfnbf0pt9uzZampq0rJly1J+Sq2+vl5lZWVasWKF46fS/vWvfyXn+cEPfqBVq1Zp8+bNWr9+vU499VT17t07WW8uCOfu6ogrwrld4nqzD3SFcG4PWwOIjW1BOM+dje1nytZzA8EinBfI4sWLNXjwYJWVlammpkZPPfVU8n+XXnqpJkyY4Jh/1apVGj16tMrKyjRkyBAtWbLE8f/Bgwen/eK2efPmJec5//zzNXDgQJWWluqwww7TWWedpVdffdVovQnn7uqIK8K5Xfy+kfNKFG6yYBfCuT1sDSA2tgXhPHc2tp8pW88NBItwHnOEc3d1xBXh3C6EcyA9wrk9bA0gNrYF4Tx3NrafKVvPDQSLcB5zhHN3dcQV4dwuhHMgPcK5PWwNIDa2BeE8dza2nylbzw0Ei3Aec4Rzd3XEla3hPAqdtBuEcyC9SIbzkPZNtgYQG687hPPc2dh+pmw9NxAswnnMEc7d1RFXhHO7EM4txnUkUIRzexSk3zDdtyHtNwjnubOx/UwRznPnx7lhC8J5zBHO3dURV4RzuxDOLcZ1JFCEc3sQznNHOM+dje1ninCeO8J54RDOLUc4d1dHXAVyAxTSmyxT+R63Xt1kRSGcB9KpWzJwFVd5nxum+zYq4dzWcOB1OLfkXCKc5y6S/b4H1xHCefgQzmOOcO6ujny2IcwXFcK5dwjnhPOuROUmyw+m50be+5ZwnhHhPHeE89xFst8nnGcUlfvoXBDOY45w7q6OfLYhzBcVwrl3QhHOPerUCee5icpNlh9CEc49qCNvth63NobzkN4rEM7tQTjPXVTuo3NBOI85wrm7OvLZhjBfVAjn3iGcE867EpWbLD/4Ec7z7pv8CJCmLDlubWy/FCG9VyCc21OHreHcxntWG9fJK4TzmCOcu6sjn20I80014dw7hHPCeVeich3xg43hjnBOOO+KH+e43/2GL32TRwjn9txX2bhOXiGcxxzh3F0d+WxDmG+qbb0BisIFm3BOOO9KVK4jfrAx3BHOCeddIZy7qyMyfRPhPCMb18krhPOYi0U49+Mmy2Ab4nyTRTjPjHBOOO8K4Tx3NoY7wjnhvCuEc3d1RKZvsvDcsOW+ysZ18grhPOYI5wWqw2AbAulwLbnJ8iWchzSwEM4J510hnOfOxnBHOLc3gBDOveuTI9M3WdLehHN71skrhPOYI5x7UEeB1smUrXWEIpxbEHAI54TzrhDOc2djuCOc29XHEs4J50HXEYZzw6v2M0U4LxzCueUI5x7UUaB1MmVrHTaG80A6HtN1ykEkb4A8EEinbsn5F1c2hrtArp9dCEO/YUv7pbBkXxW0jlz40cdacG74UUcYzg2v2s9UIP14QAjnMUc496COAq2TKVvr8OMGyPeOxw3TdcqBH9vh+w2QBwLp1C05/+LKxnAXyPWzC2HoN2xpvxSW7KuC1pELS8Kd732TJe3t97nhVfuZilP/RziPOcK5B3UUaJ1M2VqHHzdAvnc8bpiuUw782A7fb4A84Gbf5s2S8y+ubAx3gVw/uxCGfsOW9kthyb4qaB25sCTc+d43WdLefp8bXrWfqTj1f4TzmCOce1BHgdbJlK11+HED5HvH44bpOuUg7+PQsA5fboA84Ga782bJ+RdXNoa7QK6fXQhDv2FL+6WwZF8VtI5cWBLufO+bLGlvv88Nr9rPVJz6P8J5zBHOPaijQOtkytY6/LgB8r3jccN0nXKQ93FoWEdkboD8YMn5l3cdIWVjuAvk+tmFMPQbtrRfCkv2VUHryIUl4S6SfZOF54ZX7WfKj3PDFoTzmCOce1BHgdbJlK11+HED5HvH44bpOuUg7+PQsI7I3AD5wXQ7PLq2mQpkX3nAxnBn4/UzDP2GLe2XwpJ9VdA6cuFHH+tH++WAc4NwHgTCecwRzj2oo0DrZMrWOvy4AfK9/dzwoI5Q3MAaboctx23eTOvw6NqWdx0hZeO5YeP1syDnhgXhwMZ960bg/XguPGi/fOuIU9/k97nhVfuZCqQfDwjhPOYI5x7UUaB1MmVrHX7cAPnefm54UEdUbmBtPG7zZlqHR9c23+uwhI3nho3nX0Ha2+PttqX9sm53iM4/P4KX33V4dZ2ysW/y+9zwqv1MRaVvygXhPOYI5x7UUaB1SmF7px7gDZDv7eeGB3VE5QY27/brQiCduo037n7U4YbHN9W2nBs2nn+29k1Rab+M+8nWOgq0TqZ15LsdXl2nbL+nCnP7mQqkbwoI4TzmCOce1FGgdUoR0jq8vrl0U0fe7eeGB3XYeAObdbu9ar8uBNKp+31TbUsdbnh8U23LuWHj+ReGfiPM7ZdxPxWojjDs21zqyHdf2dI3Za2ji9cPc/v50W8E0jcFhHAec4RzD+oo0DqlCGkdNnagoWi/HNh4A5t1u71qvy4E0qmb1uHRtc33Otzw4PVtPDdsPP/C0G+Euf0y7qcC1RGGfZtLHfnuK1v6pqx1dPH6YW4/G8+NMCOcxxzh3IM6CrROKUJah40daFTaLyo3sHnX0YVAOnVLbk78aL+8efD6cT03UnSxDNc2wjnt59P558E6hbn9Mu6nAOuwBeE85gjnHtRRoHVKEdI6vL65dFNHVNovrjdAKXV0oSDtZ8qSmxM/2i9vHrx+XM+NFF0sw7XNrvYzrSMM+zZO7ZciAvvWqzoy7qcA67AF4TzmCOce1FGgdUoR0jq8vrl0U0dU2i+uN0A2tl/W+QO8OfGj/fLmwevH9dxI0cUyXNvsaj/TOsKwb+PUfikisG996WP96Jv86MsKhHAec4RzD+rwYJ18uThKVuzbrOsUovbLuh0FqiOuN0BRab+Mr2/rvvXqBsjjawLnRuZlbD03aD/v2s/vfRun9ksRgX0b5vZzcNN+ASGcxxzh3IM6PFgnr+pIYcG+zbpOIWq/rNtRoDqi0oHauN0pbN/uENWRwuNrAudG5mXCcG7QfoVtP7/3bZzaL0UE9m2Y28/BTfsFhHAec4RzD+rwYJ28qiOFBfs26zqFqP2ybkeB6ohKB2rjdqewfbtDVEcKj68JnBuZlwnDuUH7Fbb9/N63cWq/FBHYt2FuPwc37RcQwnmBLF68WEOGDFF5eblqamq0evXqrPOvWrVKNTU1Ki8v19ChQ7VkyZKUeVasWKHq6mqVlZWpurpaDQ0NedfbGeHcgzo8WCev6khhwb7Nuk4har+s21GgOqLSgdq43Sls3+4Q1ZHC42sC50bmZcJwbtB+hW0/v/ct7RfufRvm9nNwc20LCOG8AO655x6VlpZq6dKlampq0syZM1VRUaEtW7aknX/z5s3q2bOnZs6cqaamJi1dulSlpaVasWJFcp61a9equLhY8+fP18aNGzV//nyVlJRo/fr1rutNh3DuQR0erJNXdaSwfd+GqP1ShPQYiWv7sW+9qyNFPvPTfnnVwbUtfu3H9ZO+ybY6bGk/WxDOC2DMmDGaPn26Y1pVVZXmzp2bdv45c+aoqqrKMW3atGkaO3Zs8u/zzjtPkyZNcsxz8skna8qUKa7rTYdw7kEdHqyTV3WksH3fhqj9UoT0GIlr+7FvfbwBymd+2i+vOri2xa/9uH7SN9lWhy3tZwvCeZ52796t4uLilLec19bWavz48WmXGTdunGprax3TGhoaVFJSoj179kiSKisrtWjRIsc8ixYt0qBBg1zXK0m7du1SW1tbsrz99ttKJBJqaWlxTLepVM76U7K0ze3tLF3M72aZvOvwYJ2iUkec2i+l0H6haj/2rXd1mJ4bWeen/fKqIwznBu1H+0Wi/UK6b+PUfraUlpYWJRIJtba25hZGDUU+nL/zzjtKJBJas2aNY3pdXZ2GDx+edplhw4aprq7OMW3NmjVKJBLaunWrJKm0tFR33XWXY5677rpLZWVlruuVpHnz5imRSFAoFAqFQqFQKBQKxcLS0tKSWxg1FJtwvnbtWsf066+/XiNGjEi7zLBhwzR//nzHtKefflqJRELbtm2TtD+c33333Y55/vCHP6i8vNx1vVLqk/OPPvpImzZtUmtra+AjRbmOJJk85Tddhjq8q8PGdaKOcK9TVOqwcZ2oI9zrRB3hXifqCPc6UUe41yno0traqpaWFu3duze3MGoo8uE8bG9rD7O2NvPPYJguQx3e1WHjOlFHuNcpKnXYuE7UEe51oo5wrxN1hHudqCPc6xR1kQ/n0v4vZpsxY4ZjWnV1ddYvhKuurnZMmz59esoXwk2ePNkxz6RJk1K+EM6k3rCz9QSmjvCuE3WEe52iUoeN60Qd4V4n6gj3OlFHuNeJOsK9TlEXi3De/pNmy5YtU1NTk2bNmqWKigo1NzdLkubOnatLLrkkOX/7T6nNnj1bTU1NWrZsWcpPqa1Zs0bFxcVasGCBNm7cqAULFmT8KbVM9UaNrScwdYR3nagj3OsUlTpsXCfqCPc6UUe414k6wr1O1BHudYq6WIRzSVq8eLEGDx6ssrIy1dTU6Kmnnkr+79JLL9WECRMc869atUqjR49WWVmZhgwZoiVLlqS85r333qsRI0aotLRUVVVVuu+++4zqjZpdu3Zp3rx52rVrl2fLUId3ddi4TtQR7nWKSh02rhN1hHudqCPc60Qd4V4n6gj3OkVdbMI5AAAAAAC2IpwDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMI5AAAAAAABI5wDAAAAABAwwjkAAAAAAAEjnAMAAAAAEDDCOQAAAAAAASOcAwAAAAAQMMK55fbu3auWlha1traqra2NQqFQKBQKhUKhUCgBlNbWVrW0tGjv3r2eZD/CueVaWlqUSCQoFAqFQqFQKBQKhWJBaWlp8ST7Ec4t19ramjwAgh4polAoFAqFQqFQKJS4lvYHp62trZ5kP8K55dra2pRIJNTW1hb0qgAAAABAbHmdzQjnliOcAwAAAEDwCOcxRzgHAAAAgOARzmOOcA4AAMJs8A8fdhTN6+MsABAShPOYI5wDAIAwI5wDiArCecwRzgEAQJgRzgFEBeE85gjnAAAgzAjnAKKCcB5zhHMAABBmhHMAUUE4jznCOQAACDPCOYCoIJzHHOEcAACEGeEcQFQQzmOOcA4AAMKMcA4gKgjnMUc4BwAAYUY4BxAVhPOYI5wDAIAwI5wDiArCecwRzgEAQJgRzgFEBeE85gjnAAAgzAjnAKKCcF4gixcv1pAhQ1ReXq6amhqtXr066/yrVq1STU2NysvLNXToUC1ZssTx/9tvv12JRCKl7Ny5M696OyOcAwCAMCOcA4gKwnkB3HPPPSotLdXSpUvV1NSkmTNnqqKiQlu2bEk7/+bNm9WzZ0/NnDlTTU1NWrp0qUpLS7VixYrkPLfffrv69Omjbdu2OUo+9aZDOAcAAGFGOAcQFYTzAhgzZoymT5/umFZVVaW5c+emnX/OnDmqqqpyTJs2bZrGjh2b/Pv2229X3759C1pvOoRzAAAQZoRzAFFBOM/T7t27VVxcrIaGBsf02tpajR8/Pu0y48aNU21trWNaQ0ODSkpKtGfPHkn7w3lxcbEGDRqkww8/XKeccoqee+65vOqVpF27dqmtrS1ZWlpaCOcAACC0COcAooJwnqd33nlHiURCa9ascUyvq6vT8OHD0y4zbNgw1dXVOaatWbNGiURCW7dulSStW7dOd955p1544QWtXr1aZ599tnr06KHXX3/ddb2SNG/evLSfZSecAwCAMCKcA4gKwnme2kPy2rVrHdOvv/56jRgxIu0yw4YN0/z58x3Tnn76aSUSiZTPlbfbu3evRo0apSuvvNJ1vRJPzgEAQLQQzgFEBeE8T169rT2dyy67TJMmTXJdbzp85hwAAIQZ4RxAVBDOC2DMmDGaMWOGY1p1dXXWL4Srrq52TJs+fbrjC+E627dvn4499lhNnTrVdb3pEM4BAECYEc4BRAXhvADaf9Js2bJlampq0qxZs1RRUaHm5mZJ0ty5c3XJJZck52//KbXZs2erqalJy5YtS/kpteuuu06PPvqoNm3apOeff15Tp05VSUmJGhsbc643F4RzAAAQZoRzAFFBOC+QxYsXa/DgwSorK1NNTY2eeuqp5P8uvfRSTZgwwTH/qlWrNHr0aJWVlWnIkCFasmSJ4/+zZs3SoEGDVFZWpn79+umkk05K+Xx5V/XmgnAOAADCjHAOICoI5zFHOAcAAGFGOAcQFYTzmCOcAwCAMCOcA4gKwnnMEc4BAECYEc4BRAXhPOYI5wAAIMwI5wCignAec4RzAAAQZoRzAFFBOI85wjkAAAizqITzlO0AEDuE85gjnAMAgDAjnAOICsJ5zBHOAQBAmBHOAUQF4fz/a+/co6uo7j2ekzd5AQJC0IRQBBK0IigPRTBavRgFLQiCUq5yV72AyxuIIiK2F20FCV2lUM2NQmmxFlsqhJe2VqVELiABrwrVoAIhSAvIEgSUp8D3/sE6Z2VyzpxzfnvOzNlz5vtZa/44c2bPb+/9m/34/vY8PA7FOSGEEELcDMU5ISRRoDj3OBTnhBBCCHEzFOeEkESB4tzjUJwTQgghxM1QnBNCEgWKc49DcU4IIYQQN0NxTghJFCjOPQ7FOSGEEELcDMU5ISRRoDj3OBTnhBBCCHEMG4QzxTkhJFGgOPc4bhDnHKgIIYSQBIHi3BSKc0IIxbnHoTgnhBBCiGNQnJtCcU4IoTj3OBTnhBBCEh4XCrWEheLcFIpzQgjFucehOCeEEJLwuFCoJSwU56ZQnJNYw2vKfVCcexyKc0IIIQmPC4VawkJxbgqFFIk1vKbcB8W5x6E4J4QQkvC4UKglLBTnplBIkVjDa8p9UJx7HIpzQgghCY8LhVrCQnFuCoUUiTW8ptwHxbnHoTgnhBCS8Ngt1FwqBuMCxbkpFFIk1vCach8U5x6H4pxIYCdPCHElFOf6QHFuCsdYEmt4TbkPinOPQ3FOJLCTJ4S4EopzfdBEnOs4lnGMJbGG15T7oDj3OBTnRAI7eUKIK6E41weKc1M4xpJYw2vKfVCcexyKcyKBnTwhxA2EFWt2QHEePRTnpnCMJbGG15T7oDj3OIkoztkR2UeiPNdHCElsKM41huLcFM5fSKzhNeU+KM49DsU5kUBxTghxAxTnGkNxbgrnLyTW8JpyHxTnHofinEigOCeEuAGKc42hODfFiflLosyREnIu4kTbINpDce5xXCfOhQMuO6LYQnGuF7zWCQkNxbnGUJybQnEePQk5F6E4J6A49zwU50QCxble8FonJDQU5xpDcW4KxXn0JORchOKcgOI8ZlRVVaGoqAgZGRno3bs31q9fH/b42tpa9O7dGxkZGejcuTOqq6sN/y9YsAA33ngjWrVqhVatWuEHP/gB6urqDMfMmDEDSUlJhq19+/aifFOcEwkU53rBa52Q0FCc64MT4wbFuQUbLr1uE3IuQnFOQHEeE/70pz8hLS0NCxcuRH19PSZNmoTs7Gzs3bs35PENDQ3IysrCpEmTUF9fj4ULFyItLQ3Lli0LHHP//fejqqoKH374IXbs2IFx48ahZcuW+Oc//xk4ZsaMGbjyyitx4MCBwHbo0CFR3inOQ5AonbwNUJzrBQddQkJDca4PFOfRQ3EePYkwF9G1bZD4QnEeA/r27YsJEyYY9hUXF2PatGkhj586dSqKi4sN+8aPH4/+/fub2jh37hxyc3Px8ssvB/bNmDEDPXv2tJBzivOQsOMyhZ28XlCcExIainN90FWA6Nh3UpxHTyLMRXRtGyS+UJxb5MyZM0hJSUFNTY1hf3l5OQYNGhQyzcCBA1FeXm7YV1NTg9TUVJw9ezZkmuPHjyMzMxNr1qwJ7JsxYwaysrKQn5+PoqIijBo1Crt37xbln+I8BOy4TGEnrxcU54SEhuJcH3QVIDr2nQkrzm2wkQhzEV3bBokvFOcW+de//oWkpCRs3LjRsH/mzJno1q1byDRdu3bFzJkzDfs2btyIpKQk7N+/P2Sahx9+GF26dMGpU6cC+/7yl79g2bJl2L59O95++23cdNNNaN++Pb766ivT/J4+fRrHjh0LbPv27aM4bw47LlPYyesFxTkhoaE41wddBYiOfSfFefQkwlxE17ZB4gvFuUX84nzTpk2G/c8++yy6d+8eMk3Xrl0xa9Ysw74NGzYgKSkJBw4cCDq+srISrVu3xrZt28Lm5dtvv0X79u3xy1/+0vSYUC+RozhvBjsuU9jJ6wXFOSGhoTjXB10FiI59J8V59CTCXETXtkHiC8W5Rey+rf0Xv/gFWrZsia1bt0aVn1tvvTXo+femcOU8CthxmcJOXi8ozgkJjY7i3KvtVVcBoqMvKM6jJxHmIrq2DRJfKM5jQN++fTFx4kTDvpKSkrAvhCspKTHsmzBhQtAL4ebMmYO8vDy89957UeXj9OnTuOyyy/DMM89EnXc+cx4CdlymsJPXC69O9gmJBMW5PugqQHT0BcV59CTCXETXtkHiC8V5DPB/Sm3RokWor6/H5MmTkZ2djcbGRgDAtGnTMHbs2MDx/k+pVVRUoL6+HosWLQr6lFplZSXS09OxbNkyw6fSvvnmm8Axjz32GGpra9HQ0IDNmzdjyJAhyM3NDdiNBopzNRtehZ28Xnh1sk9IJCjO9UFXAaKjLyjOoycR5iK6tg0SXyjOY0RVVRU6deqE9PR09O7dG++++27gvwceeAA33XST4fja2lr06tUL6enpKCoqQnV1teH/Tp06hXw2fMaMGYFjRo0ahfz8fKSlpaFjx44YPnw4PvnkE1G+Kc7VbHgVdvJ64fREzi50nCQTd0Nxrg+6ChAdfUFxHj2JMBfRtW2Q+EJx7nEoztVseBV28npBcU5IaCjO9UFXAaKjLyjOoycR5iK6tg0SXyjOPQ7FuZoNr8JOXi8ozgkJDcW5PugqQHT0BcV59CTCXETXtkHiC8W5x6E4V7PhVXTt5HWcZDkBxTkhoUlIca5JfytFVwGiY79DcR49Os5FpOjaNkh8oTj3OBTnaja8iq6dvI6TLCegONcYTdqGV6E41wddBYiO/Q7FefToOBeRomvbIPGF4tzjUJyr2fAqunbyOk6ynIDiXGM0aRteheJcH2IybkjrluI8ehsU53GD4jx6nGgbukBx7nEoztVseBVdJ0CJ0GFbvW7tmmQlgv/iMqizH4krltuG1H+JIs51FQcU5/bZoDiPG1av22iuEYpz90Fx7nEoztVseBVdJ0CJ0GFTnFOca2nDpVCcK+JVca5JW6I4jx4d/SeF4jx6KM5jB8W55lCcq9mwUgY3dyoU5/bhCnFu06BOcR4diTLJcgJp27Bctypjkw02LKPrdaujOHfpXIHiXB8ozqMnUebR0UBx7nEoztVsWCmDmzsVinP7oDinOI9EokyynMAJcW55bHJCQErR5LrV0X9BuHSuQHGuaMMGdBXnOs5ZdcyTXVCcexyKczUbVsrg5kk1xbl9UJxTnEciUfoRJ9BR3FGcU5xHwok2TnGuaMMGKM6jR8c82QXFucehOFezYaUMbp5U6zq5TIQOm+Kc4jwSidKPOIGO4k7X/tOAJtetjv4LQpO6iqmNGMGxSX5+ivPw6Jgnu6A49zgU587YiMukWpOJQyIMoE5AcU5xHgmK8+jRUdzp2n8a0OS61dF/QWhSVzG1ESM4NsnPr9P8U0chrGOe7ILi3ON4Qpw7MckSlMHLkyxHBlCXChaKc4rzSFCcR4+O4k7X/tOAJtetjv4LQpO6iqmNGJGQY5Mm/na6begihHXMk11QnHscinMbbMQoT1J0teGKAVQDgUNxTnEeCYrz6NFR3OnafxrQ5LrV0X9BaFJXMbURDTb4z2o5tJ1bCM6va9uwy39SKM5jB8W55lCc22AjRnmSoqsNHQfQRBl4EnICZANxGdQ1aX9eRUdxp2P/6YZxQxf/BaFJXcXURjQ4McZq0DacsOGGtmGX/6TEZRyPExTnHofi3AYbMcqTFF1t6DiAOiFqpagMPE6Uw3H/2UBcBnVN2p9X0VHc6dh/umHc0MV/QWhSVzG1EQ2aiDsd5xZWzq9r27DLf1LiMo7HCYpzj0NxboONGOV8LdXDAAAgAElEQVRJiq42dBxAnRC1UlQGHifK4bj/bCAug7om7c+r6CjudOw/3TBu6OK/IDSpq5jaiAZNxJ2Ocwsr59e1bdjlPyleGv8ozj0OxbkNNmKUJym62tBxAHVC1EpREZCWr0OhDUf8ZwMq5baMJu3Pq+go7nTsP90wbujivyA0qauY2ogGTcSdjnMLK+fXtW3Y5T8pXhr/KM49DsW5DTZilCcputrQcQC1PPDYgIqAtHwdCm044j8bUCm3ZTRpf15FR3GnY//phnFDF/8FoUldxdRGNGgi7nScW1g5v65twy7/SfHS+Edx7nEozm2wEaM8SdHVho4DqOWBxwZUBKTl61BoQxf/WTl/tOW2jLQcNvVtUuJSVzago7jTUUC6YdzQxX9BaFJXMbURDU6MsS6dW1g5v65twy7/SXGibegCxbnHoTi3wUaM8iRFVxs6DqCW/WcDKjZcMYEVlkOX69YyUhs29W2WbbgUHduGjgIyJm1DA3GgY92qEPdxPBps8J9VG14am5xuG3b5T0pcxvE4QXHucSjObbARozxJ0dWGjgOoZf/ZgIqNRJnA6njdWkZqw6a+zXEbmqBj29Cx/cXE3xqIAx3rVuX4uI/j0WCD/6za0GVuYeX8urYNu/wnJVHGpmigOPc4FOc22IhRnoLQfVB30QBq2X82oGIjUSawlv0XgbgM6jpO3DURB2HzFSN0bBs6tj9dx6ZE8Z9pPelqI0Z5ktqwWg67+ind51Ru9p+UuIxNcYLi3ONQnNtgI0Z5CsKlNnQcQC37zwZUbOg4gQ3CifYXgbgM6ok4cXeirmKEjm1Dx/bnhnHDzf4zracY2XBD3UZjw2pd6TI2SUkU/zkxbsRlbIoTFOceh+LcBhsxylMQLrXhRGfqCv8JytDpCYpzUTkiEJdBXWrDpr7NcRsq2Nwn6NI2dGx/bhg33Ow/03qKkQ031G00NqzWlS5jU1gbEc7vZv/FZY7kxNgUJyjOPQ7FuQ02YpSnIFxqw4nO1BX+E5RBp3K4ov1FIC6DutSGrnXrdF3FCK+2jSAipHHDuOFm/5nWU4xsuKFu3ew/tg2H5kiatD9doDj3OBTnNtiIUZ6CcKkNJyb6rvCfoAw6lcMV7S8CTvgvCE0mJ074zzI29wleahtBREjjhnHDS/6T2nBD3XrJf0EkQN3aZcO0nuJoQxcozj0OxbkNNmKUpyBcasNynqLAFf4TlEGncrD9KaLJ5MQJ/1nG5j7BS20jiAhp3DBueMl/UhtuqFsv+S+IBKhbR8ZYJ8YmJ8ayGEFx7nEozm2wYUOeHOkcAS3qVgVX1K2gDDqVw6vtLwgNJg5xr1u7JkA29wlsG+ZpdG0b9J99/nO6br3kvyASoG7d7D8DKv6LExTnHofi3AYbNuTJLhtBaFC3KriibgVl0KkcXm1/QehebhfZCMLK8Wwblmy4oW3Qf7H1n9N16yX/BZEAdetm/xlQ8V+coDj3OBTnNtiwIU922QhCg7pVwRV1KyiDTuXwavsLQvdyu8hGEFaOZ9uwZMMNbYP+i63/nK5bL/kviASoWzf7z4CK/+IExXmMqKqqQlFRETIyMtC7d2+sX78+7PG1tbXo3bs3MjIy0LlzZ1RXVwcds2zZMpSUlCA9PR0lJSWoqamxbLc5FOc22LAhT3bZCEKDulXBFXUrKINO5fBq+wtC93K7yEYQVo5n27Bkww1tg/6Lrf+crlv6z91162b/GVDp2+IExXkM+NOf/oS0tDQsXLgQ9fX1mDRpErKzs7F3796Qxzc0NCArKwuTJk1CfX09Fi5ciLS0NCxbtixwzKZNm5CSkoJZs2Zhx44dmDVrFlJTU7F582Zlu6GgOLfBhg15sstGEBrUrQquqFtBGXQqh1fbXxC6l9tFNoKwcjzbhiUbbmgb9F9s/ed03dJ/7q5bN/vPgErfFicozmNA3759MWHCBMO+4uJiTJs2LeTxU6dORXFxsWHf+PHj0b9//8Dve++9F7fffrvhmMGDB2P06NHKdkNBcW6DDRvyZJeNIHSvWxf5T4qu14hX/ce6tc9GEFaOp/8s2XDDuEH/xdZ/7D85NulmQxf/6QLFuUXOnDmDlJSUoFvOy8vLMWjQoJBpBg4ciPLycsO+mpoapKam4uzZswCAgoICzJ0713DM3LlzUVhYqGw3FBTnNtiwIU+JYsNL/pNC/+lbbtatzRMgK8fTf5ZsuKFt0H/0X6Q0rvCfDXlKFBu6+E8XKM4t8q9//QtJSUnYuHGjYf/MmTPRrVu3kGm6du2KmTNnGvZt3LgRSUlJ2L9/PwAgLS0NS5YsMRyzZMkSpKenK9sFgNOnT+PYsWOB7YsvvkBSUhL27dtn2K/TVjD5z4Ht2LRc4xbheJU0lm3YkKdEseEl/1m5znUqh1f9x7q1z0bQZuV4+s+SDTe0DfqP/ksI/7m0br3kP122ffv2ISkpCUePHo1OjArxjDjftGmTYf+zzz6L7t27h0zTtWtXzJo1y7Bvw4YNSEpKwoEDBwBcFOevvvqq4Zg//OEPyMjIULYLADNmzEBSUhI3bty4cePGjRs3bty4cdNw27dvX3RiVEjCi3O33dbefOX866+/xu7du3H06NG4R4qijSRJVvmlaWjDPhs65ok23J2nRLGhY55ow915og1354k23J0n2nB3nuK9HT16FPv27cP58+ejE6NCEl6cAxdfzDZx4kTDvpKSkrAvhCspKTHsmzBhQtAL4crKygzH3H777UEvhJPYdTvHjsmfwZCmoQ37bOiYJ9pwd54SxYaOeaINd+eJNtydJ9pwd55ow915SnQ8Ic79nzRbtGgR6uvrMXnyZGRnZ6OxsREAMG3aNIwdOzZwvP9TahUVFaivr8eiRYuCPqW2ceNGpKSkYPbs2dixYwdmz55t+ik1M7uJhq4NmDbcmyfacHeeEsWGjnmiDXfniTbcnSfacHeeaMPdeUp0PCHOAaCqqgqdOnVCeno6evfujXfffTfw3wMPPICbbrrJcHxtbS169eqF9PR0FBUVobq6Ouicr732Grp37460tDQUFxdj+fLlIruJhq4NmDbcmyfacHeeEsWGjnmiDXfniTbcnSfacHeeaMPdeUp0PCPOif2cPn0aM2bMwOnTp21LQxv22dAxT7Th7jwlig0d80Qb7s4Tbbg7T7Th7jzRhrvzlOhQnBNCCCGEEEIIIXGG4pwQQgghhBBCCIkzFOeEEEIIIYQQQkicoTgnnmXv3r24cOFCvLNBCCGEEEIIIRTnRM7LL7+s9NKGY8eO4fz580H7z507p/SGxkOHDonTNCU5ORlffvmlpXNY4fz581i9ejXuvvtuS+d55plncOLEiRjlSo3t27dj0qRJls4hLcfJkyexatUqHD9+POi/Y8eOYdWqVeLrNBblSBQYvCKE6A77Kb2gPwixDsU5EaMiamtqatC1a9eQ4uvEiRPo1q0bVq9eHfE8Fy5cwBtvvIFhw4YhPT3d8N/JkyexZs2awO9p06ahoqIisE2ZMgWnTp0K/O/z+cTlOHLkCH7961+HDCYcPXrU9L+mfP7555g2bRry8/ORmZkZUZwfP34cx44dC2zffPON4X+pPz7//HOMHj3atAz33Xcfdu/eHfE8x44dw4svvog+ffrA5/OhZ8+eQf/7eeONN7Bq1arA9vrrrwedT1qOefPm4ZZbbjH9/wc/+AFeeOEFS+XYtWsXxo0bF/hdUFCA1q1bB7a2bdvi008/NZzv5ptvxtdffx11Oezgq6++wq9+9StL50iU4FW8/eHvM5q3Dynjxo0LGYhyglj5QiWwG+ugLmAtsBtPPzTlww8/tJS+c+fO+Oqrr8TpYu2P7777TpymKfHup/xY9ce7775ruS5igdv9wXHDSLz7q1j5w2tQnBMxKqL2tttuw8KFC03/X7RoEf7t3/7N9P/du3fjqaeewuWXX45WrVphzJgxqKmpMRzz4osvYsiQIYHfOTk56NevH0pLS1FaWooOHTpg7ty5lsrxs5/9DCNGjDD9f+TIkXj22WeD9p88eRKLFy/GwIEDkZaWhuTkZMyfPz9IaAMXB/k77rjDUI7k5OTAlpKSgi1btiiX46GHHsLjjz9u+v/UqVMxYcIE0/9ra2sxduxYZGVlITk5GU888QR27txpOGbNmjW45pprDGXw+XyBLTk5Ga+99pohjbQcffr0CRvQWbNmDfr06WOpHJMmTcKTTz5pKMecOXOwePFiLF68GGVlZRg/frylcgCxCfpcuHABb775JkaOHIn09HS0bdvW8L800CAtRyzKAMiCV5ECVyrlWLt2LUpKSkzL0aNHD6xfvz7ied5++22MHj0amZmZuPzyy1FeXh74TxpIBOST3kQIJMYqqAuYB3alvlARH7G6po4ePYqqqir06tULycnJhv+cCE7H0h+ffPIJKioqcOmllxr2f/311/if//mfwO/7778fw4YNC2wjRowwiKZ4tW//8aH84UT7jnVZzPxx/vx5bN++PfC7uroa8+fPD2wvvPCCIVjjhkUPjhvm6DqOew2KcyLG5/OJVx7y8/ODhE9Tdu7cifz8fMO+U6dO4ZVXXsFNN92EjIwMDBkyBCkpKfjHP/4R8hwDBw40CPacnBzDCvArr7yC/v37G8oxc+ZMw0ATamtKz5498c4775iW45133jGI0rq6Ojz00EPIy8vDddddh3nz5uHgwYNITU3FJ598EvIc//Ef/4FZs2YZyrFkyRLU1tZi3bp1GDt2LH70ox8ZyiHxR/fu3Q3ivjnvv/8+unXrZti3f/9+zJw5E126dEGHDh1QUVGBrVu3mpZj6NCh+M1vfmMoQ1NfVFZWoqyszJBGWo5WrVph7969pv/v3bsXrVq1slSOK6+8En//+99Ny1FbW4srrrgiqBxOBX0AYM+ePfjpT3+KgoICJCcnY+zYsXj77bdx7tw5w3HSQIO0HFbKEG3wShq4UinH0KFDDUG85syfPx8//OEPQ/63d+9ePP300+jUqRPatGmD5ORkLFu2LOg4aSBRpRyJEEi0GtQFIgd2nQjqWrmmgIsT/zFjxqBFixYoLi7GU089hQ8++MBwjBPlsOqPb775BgsXLkT//v2RkpKCAQMGBNXLnDlzMGbMGEM57rnnHjz44IN48MEH0b17d8yYMUO5HFZ9AUT2hxPtOxZlicYfS5YswaBBgwxlufzyy1FUVISioiLk5OQYxnodFz04brh/HPcaFOdEjM/nwx133GGIZofampKZmYkdO3aYnrO+vh6ZmZmB3xMnTkTr1q3Rv39/vPDCC4Hb78KJ2vbt2+Pjjz8O/G7bti327NkT+P3ZZ58hLy/PUI6CgoLAIBNq69y5s8FGTk5OREGYm5sb+J2SkoLJkycH3focrhzdu3c3RFibC8LNmzejsLDQUI7vf//76NWrV9jNT2ZmJhobG03L0NjYiBYtWhj2ZWRk4Ec/+hHefPNNQ5TcrBydOnXC1q1bTcuwfft2tGvXzpBGWo6cnBy8//77puV4//33kZOTY6kcOTk5hmto8uTJhltBGxsbDdetvxy7du0yRIJDbU2RBn1Onz6NV199FbfccgsyMzMxbNgwvPbaa2GvK2mgQRq8kpYBkAevpIErfzkk/igsLER9fb1pOXbs2IGCggLDvqVLl+K2225DVlYWRowYgZUrV+LMmTOm5ZAGEv3lkASvEiGQqBLUBWSBXZWgrjQ4rXJN7du3Dz//+c/RuXNnXHrppXjkkUfC+kKlHOvWrcO2bdvCbk1R9cf//u//4oEHHkBOTg6+//3vIyUlBRs2bAh5jr59++KNN94wLUdNTY3hupX2Uyq+AGT+cKJ9WymLxB+33norXn31VdOyVFdXo7S01FAO3RY9OG7YN24AzvjDa1CcEzE+nw+jRo0KRLLNtqYUFxfjlVdeMT3n73//e3Tv3j3wOyUlBdOnTw96Vibc5CQzMzNIBDdlx44dyMjIMJRDGuFt2bIl3nvvPdP/33vvPbRs2TLw+7bbbkNubi7uv/9+/PWvfw28KCVcObKysgziee7cuYZBYO/evUHlmDJlCp5++umwm5/27dtj7dq1pmV455130L59e8O+bt26oaioCNOnTzcEWczKkZGRgYaGhsDvrVu34uzZs4HfDQ0NQe8MkJajX79+mD17tmk5nnvuOfTr189SOfLy8lBXV2dqo66uzhCM8ZejaRS4+eb/vynSoE+bNm0wcOBAvPTSSzhy5EjEcvhtSAIN0uCVtAyAPHglDVz5yyHxR0ZGRkQB0jwgk5KSgieffDLq/koaSPSXo1WrVoZHEUJtTetGx0CiJLArDeoC8sCuSlBX4gdAfk2VlZUhNzcX9913H15//fXAXTCxDk77r//mm1k/JfVHZWUlunfvjssuuwxTpkzBRx99FLEcbdq0wWeffRb4fe2112Lfvn2B37t370Z2drahHJJ+SqV9S/2h2r7Hjx9vuE051NYUaVlU/HHZZZcFjgOC23h9fb3hetdx0YPjhn3jhr8cdvvDa1CcEzEqonb69OkoLCzEwYMHg/47cOAACgsLMX369MC+JUuW4NZbb0V2djbuvfderFmzBt99913YQeSKK64IeSuQn6VLl6JLly6B3yrPeJWWluKJJ54w/X/q1KmGKDIAfPHFF3jmmWdQVFSE9u3bo7y8HKmpqaZR1tatW5tGsQFgw4YNQYOhpBwjR44Me6vbXXfdFfK2pg0bNmDcuHHIyclB7969MXfuXNNy5Ofn4+233za18be//Q0dOnQw7JOW46WXXkJ2drbhGSw/q1evRnZ2Nl566SVL5bj++usxc+ZM0zz87Gc/w/XXXx9UjpqaGtTW1obdmiIN+rRq1QqDBg3CggULDIGbcO1DGmiQ+kNaBkAevJIGrvzlkPjje9/7XtD7LJqyfPnyoMnlQw89hJYtW+KGG25AdXV1IGBiVg5pINFfjvnz5wceQzDb/OgaSJQEdqVBXUAe2FUJ6kr8AMivqZSUFFRUVODzzz+Pqgyq5di6dSsaGxvDbk1RDbI3f8QmXDlatGhh+ugacPGuq6Z3dkn7KZX2LfWHavu+4YYbArcoh9puvvlmS2VR8UdGRgZ27doV+H3o0CHDXWc7d+40BNp1XPTguGHfuAE44w+vQXFOxKh0vsePH8eVV16J3NxcTJw4EfPmzcP8+fMxYcIE5ObmokePHiHfKLlnzx7893//NwoLC9G2bduQLxLzU15ejh49egS9EAO4+CxMjx49DC/XUCnHsmXLkJqaiueff94wwJ07dw6//vWvkZaWZpo/AHjrrbcCL/vo2rUrnnzySfzf//2f4ZhbbrkFU6ZMMT3Ho48+anhLuTTI8MEHHyAjIwP33HMP6urqcPToURw9ehSbN2/G8OHDkZGREZSnpnzzzTdYsGAB+vfvD5/Ph9LSUixYsMBw69SoUaMwdOhQ03PceeeduPfeew37VIIlY8aMgc/nQ0lJCX74wx9i2LBhKC4uRnJyMkaPHh02bTTlWLBgAbKyskK+XX716tXIysrCggULDPtVritp0OfUqVP4wx/+gJtvvhktWrTA8OHDUVNTg7S0NNNJljTQIPWHSuAKkAWvpIErQO6PRx55BFdddZVpP3LVVVfhv/7rv0L+t3jxYgwaNAgZGRm46667TG+llgYSVcqRCIFEaVAXkAd2pb5Qad/Sa2rTpk348Y9/jLy8PPTt2xfPP/88Dh06FNPgtBNB9pkzZ6Jr164oKCjA1KlTA20h0uM3L7/8smkefvvb36JHjx6B39J+SqV9S/3hRPtWKYuKPwoLCw2PGTRn9erVhlVOHRc9OG64fxz3GhTnRIzKIAJcfNPjxIkTcckllwRun7vkkkswceLEiJ+suHDhAv76179i5MiRyMjIwGWXXRbU0R08eBAdOnRAYWEh5syZg5UrV2LVqlWorKxEQUEB8vPzDZOKp59+Wun74NOnT4fP50NeXh6uueYa9OrVC3l5eYE3fkeD/42Y11xzTdCtg/4AQPO3oJoFAML5w+wzFmvWrEG7du2CbtVq164dVq1aFW1VoL6+Ho899hguvfRSpKamBvb7AwAjRozAli1bAgGAuro60wCA6nW1dOlS3H333ejRowdKSkpw9913Y+nSpaJzmJUDAEaPHh0UACgpKUFycjJGjhwZdC6ngz67du0KvPDK5/Ph/vvvx1tvvRW0OiINNEjLYTVwBUQOXkkDV9GWo+nngw4ePIiOHTuioKAAlZWVgX5k9uzZKCgoQMeOHUOKk6b431TbsWNH5OXl4b777sPy5csD/0sDiYB80qtjIFF6TakGdYHoA7tSX6iID9Vr6sSJE1i0aBEGDBgQeMHSvHnzQpbZjuB082dVVf1RW1uLf//3f0d2djauvvrqsM84/+QnP0FBQQEOHDgQ9N/+/ftRUFCAp556SlSOplhp39H6Q6V9q4wbqmWR+GPcuHG44YYbQv534cIFXH/99YavgOi46MFxw/3juNegOCdifD6f4RmwaGjaQVy4cAGHDh3Cl19+Gbj9JdzxzTl8+DB+9atf4eqrrw76r6GhAYMHDzY8S5ecnIzBgwcHfbvbyve+6+rqUF5ejjvuuANlZWWYNGlS2FuGwxFqlXrq1KmmAYDmnVpjY2PQd2ej+YzFyZMnUVNTgzlz5qCyshIrVqxQClYAFweppoMIAKxcuTIwKW66tWnTBitWrAg6R2Njo+F6+Pjjjw0vJmr6rJVdhCoHAPzxj3/E3XffjZKSEpSUlOCuu+7CH//4x5DnKCoqwj//+U+xbUnQJ9T3os+fP4+//OUvuOeee5Ceno42bdoE2ZAEGlSCV7EIXAHmwStp4Aq46A+z7zmbfT6osbERZWVlQf1IWVmZ4Rm/SDQNjjW99VMaSATUVzrdEkg0w0pQF4gc2JX6QjWIaPWa+vTTT/H444+jQ4cOyMzMDLozSVqO0tLSkPVn9sk5P1b8cfz4cVRXV6Nv375ISUnB9ddfj1/+8pdBx5SUlCA3NxcPP/xwIAAwceJE5Obmori42CCGVfqpWLTvcP5wqn1bLUs0/ti1a1fgjoE///nP+Oijj7Bt2zYsXboUffr0QV5enuFZax0XPThuOD9uALH1h9egOCdiVBq7NI3qQOXn8OHDqKurQ11dHQ4fPhzymP/8z/+09L3vaLASAHjvvfdQXl6OsrIylJWVoby8POyzQJJPIJlhttKumubEiROoqalBZWUlKisrUVNTg2+//TbkedavX4/rrrsu8Nv/eY2mg1y459hVxLwdAQCVlTW/2I426BPJxqFDh4ImWX6iDTSoXruxDFwBwcErSeAqFNF8PsjPkSNHsGXLFtTV1RlevNd0xSRamvtLEkj009QXb7zxBlatWhXYQt0RAegVSJQGdqVB3eZpmmMW2JX44rvvvsO2bdtw8uTJoPOfOHEC27ZtCwqWNsXsmoqWc+fOYcWKFSEfG9q9e7f4mmqaNtwn55pj5o9o3wy9fft2TJo0KeirHcDFOho/fjxat24dKEfr1q0xfvz4oPF8//79hlvpBwwYYPi6x3XXXWcaLLXqC8DcH9L2rRLU7dy5c0BARlOWpsc3J5w/6urqUFJSYnhJmj/Iu3nzZsOxui56cNy4SDzGDSD2/kh0KM6JGF3FuVQUdevWTfy9b+nA89BDD9keAFD5BFJzollpj0UawFzMjxo1yvCJlZycHLz77rtobGzEnj17UFFRgeHDhwf+VxHzTgQAnLh2VWyEWm0Ph87BK2ngCpB9PsgMsxWTSGkeffRR0zRmgcTmE7k1a9YYPmGTk5MT9IZtq6sN4SZx4Z47jdYfTowDVgK70QR1f/e73+Haa68NemwEuCjUrr322qCXpp07d04k6MMdf/LkSWzbti2kfUk5ANkn58IRabU9HOHE/IULF/Dll1+GDQD89Kc/xcMPPxz4nZOTg/Ly8sDXPfr164fHHntMlKdYEq0vVIK6drSnpl9Vac6HH36IpUuXYunSpYbvujeF40YwiT5uAGpjh4o/vALFORHj8/nw+9//3tD4Qm1W0qjakAxUKt/7lort7t272x4AUPkEEqC20m5ldT6SmO/SpYuhY27+eY0PPvjA8A3d0aNHi8S8ShoVMe/zyb9XqzLJktqwO3ilMmFyInil8vmgpkhWTKyk8WM2kRsyZAh+85vfBH43bx+VlZUoKyszpJGsmDgxidNRnEvbxYABA0wfaQEuvgNj4MCBhn1SQe9EAED6yblQSFfb/fjF/PDhw6MW82YBgJ49e+Ktt94K/G7eLt58803DC+RUVtojpenTp4/So0xN0XXRI9xqeygSZdGD48ZFollpdyoA4CUozomYpo3ObGv+bKI0jaoNycCj8r1vqdh2IgAg/YyFykq76uq8RMw3r6vly5cbnl1rbGw0TMqkYl4ljUoAwOeTfcvZn0YitlVt2Bm8UpkwSduTSgBA5fNBgNqKieoqSzSTsk6dOmHr1q2B382v3e3btxtuSZVOmIYOHSqexAGyiZxTQVqpDUm7aNeuXdjnRxsaGtC2bVvDvhtvvFEk6KXHA3JBL/3knB8rq+0qYj5SmpYtWxqu02HDhhmeu92zZ4+hn/rJT34iXmmXplEJAPh8Pqxbt85wZ1aozUoaVRtuX/TguGHfuAGojR0qt9p7CYpzIkbXCK90UqbyvW/pwONEAACQfcZCZaVdmkZFzLdr1w7r1q0zLfe6desMk16pmFdJoxIA8Plk33L2p5GIbVUbkgCA9NpVuW6dCABIPx+ksmKiusoimZRlZGSgoaEh8Hvr1q2GW1AbGhoM1650wiSdxAHyiZxTQVqpDclYk5WVFSRgmrJt2zZkZWUZ9kkFvRMBAOkn5wC11XYVMS9Jk52dbXqLNXCxj87Ozg78lq60q6RRCQD4r0uz69XsWpekUbXh9kUPjhv2jRuAMwEAr0FxTsSoPBslTaP6/JVkUqbyvW/pwONEAKA5kT5jIV1pV9hfgrMAAAxpSURBVEmjEgAYMmSI4ZMszXnggQdw5513Bn5LxbxKGpUAgGpgSSK2VW1IAgDSa1flunUiAOAn2s8HqayYSNOoTMry8/PDvg/hb3/7Gzp06BD4LZ0wSSdxgHwi51SQVmpDEtTt2bMnqqurTc9XVVWFnj17GvZJBb0TAQA/0X5yDpCvtquIeWma3r1744UXXjAt9/z589GrV6/Ab+lKu0oalQCAz+fD1q1b0djYGHazkkbVhmS1XcdFD44b9o0bgDMBAK9BcU7E6LxyLk0j/d63dOBxIgAwbty4kN+WDfcJJMlKu0oalQDA3//+98CbOpv68csvv8Sjjz6KlJQUQ71IxbxKGpUAgBMv9lG1IQkASK9dlRUTJwIAzYn0+SDpiolKGpWJ3KhRo0K+odvPnXfeiXvvvTfwWzphkk7iAPlELprr9sMPP1Q+XiWNNKhbWVmJNm3ahBTPH330Edq0aYPKykrDfqmgdyIA0JxIn5wD5KvtKrfOS9PMmTMHl1xyiak/LrnkEsyZMyewT7rSrpJGJQCg87xKstqu46JHLMaNY8eOobq6Gn369EFycjLHjSY4EQDwGhTnRMyDDz4YUgzGMo2KDalg8Ytayfe+VQYeuwMA0ZQ71CeQ/ERaaVdNoxIAqKqqQnp6OpKTk9GqVSu0bt0aycnJSE9Px/PPP284VirmVdKoBACkt48Dam9rt/ulc4Ds2lVZMXEiABCOf/zjH5g8eXLIiUC0KyYqaVQmcv66GjFiBLZs2RKoq7q6upB1JZ0wSSdxgHwiZ3YNHj16FFVVVejVq1dUt5ybHR9LG2acPXsWpaWlSE1Nxe23347JkyejoqICt99+O1JTUzFo0KCgN15LBb0TAYBwHD58GPPmzUPLli1D/h/tarvKrfPSNGfPnsWgQYOQmpqKsrKygD/KysqQmpqKG2+80eAP6Uq7ShqVAIDO4ly62h5q3PD5fHFb9OC4Yd+4ATgTAPAaFOdEjM63tdttA5AJFicCANJyq6y0q6RpiiQA8MUXX2Du3LmYOHEiJk6ciLlz5+KLL74IeaxEzKukUQkArF27Fl26dDF9+UyPHj2wfv16w36p2FaxoXq9R3vtqgSuAPsDAGvXrkVJSYlpXZWUlISdjERaabeSRjqRW7lyZUAMNd3atGmDFStWGI6VTpikkzhAPpFrHnRdu3YtxowZgxYtWqC4uBhPPfWUQdRIj1dJo7I6f/bsWVRWVqJnz57IyspCixYt0LNnT1RWVuLs2bMhj5cIeicCAGaEC3w0J5rVdkB267xKmjNnzuC5555Dz5490aJFC7Ro0QJXX301nnvuOZw5c8bgD+lKu0oalQBA3759DS/smzZtGioqKgLblClTcOrUKUtpVGyoCHog+nEDiN2ih1kAQGXcOHnyJNasWRP4Haquwi0gRVppt5ImFuOGz+eLybgBOBMA8BoU50SMrhFe6Wq7ig2p2HYiACAVdior7dI0sQwAREIi5lXSSAMAd911V9jPn8yfPz9ooiAV2yo2VAIA4URtqACA9C4RP3YGAIYOHSquKzPCrZhYSSMJAJw4cQI1NTWorKxEZWUlampq8O233wYdpyK2JeIfUJvI7du3Dz//+c/RuXNnXHrppXjkkUfCrvxIj5emUVmdD8XXX3+Nqqoq9O7dO+TxKoLezgBAc6IJfIQj0mo7EL2Yt5oGMPeHdKVdJY1KAODFF1/EkCFDAr9zcnLQr18/lJaWorS0FB06dAjqx6RpVGxI50nRiNrmAQBAvtrut2XXuBGprvLz86P+zJld44YkAGDnuAHYHwDwGhTnRIzqLbXST0XpeGuwig27AwA+ny8gGsNtVvJkx10JKgGApqiIeScCAAUFBWFv3d+xYwcKCgoM+6RiW8WGNACgKmolKyZOBAAKCwtFdaWy0m51db4pZpMylbqSim0g+kkcIJ/IlZWVITc3F/fddx9ef/31wLOTZsJZerxKGpXV+aZYEbWRBL30eKmgVwl8hEIayACiE/MqaaLxh2SlXSWNSgBg4MCBhk/ENX8O95VXXkH//v0tpVGxIV1tVwkA+Il27FAJAEjFv7SuVFbara7ON8Vs3FCpK4nQbordAQAvQXFOxEjFoEoaVRtS4eyEDScCAPPnz8fixYvDblbypBJc0SEAYDWNipjPyMjAzp07Tf/fuXMnMjMzDfukYlvFhjQAEGtRG0o8OhEAkNaVSp6kaVTEvGpdRTthUvEfIAsApKSkoKKiAp9//rlhf7gXIEmOV03jxGp+U3QIAKgEPmKVLxUxHy5NLIIM0kBJpDTSAED79u3x8ccfB363bdvW8Pb9zz77DHl5eZbSqNiQim2VAIBUQKoGACTjhrSuVFbapWlUxLxqXUmCtE4GALwCxTkRIxWDKmlUbUgFpBM2nAgAOBGUkAZXvBoA+N73vmeYnDRn+fLl6Ny5s2GfVECq2LA7AKAiHp0IAEjrSponlTQ61pWV2/+jncht2rQJP/7xj5GXl4e+ffvi+eefx6FDh0yFlPR4lTROrOYD+gUAVIIYVvOlIuYjpbEaZLAjT2aEE/OZmZlBnx5tyo4dO5CRkWEpjYoNqdjWMQCgIh6ldeXEnQ8qAQAn6sqJAIDXoDgnYnR95typ27ulNuwOANi90q6SxssBgEceeQRXXXVVyOfqTp48iauuuiromUmpgFSxYXcAQEXUOhEAkNaVyl0J0jQ61pVKnlRX20+cOIFFixZhwIABSEtLQ3JyMubNm2d6l4r0eEkaJ1bzdQwAqAQ+VPJl9zsDnLhTQjWNn2jE/BVXXIFly5aZnmPp0qXo0qWLpTQqNqRiW8cAgIp4lNaVE3c+qAQAnKgrJwIAXoPinIjR+W3tEgHphA0dgwxOBVe8GgA4ePAgOnbsiIKCAlRWVmLlypVYtWoVZs+ejYKCAnTs2NHwzVtALiBVbNgdAFARtU4EAKR1pXJXgjSNjnWlkqdYvGzv008/xeOPP44OHTogMzMz7IuCVI6PlMaJ1XwdAwB+JIEPab6ceGeA3XdKqKaRivny8nL06NHDtL/t0aMHysvLLaVRsSEV2zoGAFRErbSunLjzQSUA4ERdOREA8BoU50SMzivnutlwIgAgxYngipcDAADQ2NiIsrKywHNUPp8PycnJKCsrMwxaflTEttSG3QEAFVHrRAAAkNWVyl0J0jQ61pVKnlSCJWacO3cOK1asiEpsqxwfKY2dq/k6BgBCESnwIc2XU+8MAOy7U0IljYqYP3jwIDp06IDCwkLMmTMn0N9WVlaioKAA+fn5QWOANI2KDanY1jEAoCJqpXXlxJ0PKgEAJ+rKiQCA16A4JwmD6mfL7LThhEiVomtQQse6tRooOXLkCLZs2YK6ujocOXIkrC2p2JbasDsAoCJqnQgASOtKpZ6kaXSsKycelXALdq3m6xQACEekwEe0+XLinQGhiOWdEippVAMMDQ0NGDx4cFB/O3jwYIN4sZJGerxUbOsYAFARtYCsrpy480ElAOBEXTkRAPAaFOckYbBb1Dplw4kgg914OQBgFYmgl2JnAEBF1ErzpCIgVVCpJ0kaHevKiUcl3Iadq/m6BACsEk2+7HxnQDhieaeEJI3VAMPhw4dRV1eHuro6HD58OKqyStNEe7yK2NYtAKAiaqV15cSdDyoBACfqyqlgiZegOCdEM+IhCN2IrgEAN2BXAEBV/EebJ1VRq4pKPUWbRse6svtRCRKMTgEAK0Sbr1i/MyAW2JUnpwMldqGyog/oEwBQEbUq2H3ng0o9SW3oGgDwGhTnhBBiAgMlaui4+q8rOtaVnY9KEOdReT7fbmL9zgDd8+RkoMQuVFb0o8XOAICqqFXFzjsfVOspWhu6BgC8BsU5IYQQ12GnqE00dAwAEOJFdAyU6IRdAQArolZHdAyURJsvp4MlboTinBBCCCGWYLCEEKI7doraREPXAIAXoDgnhBBCCCGEEOIYDJaEhuKcEEIIIYQQQgiJMxTnhBBCCCGEEEJInKE4J4QQQgghhBBC4gzFOSGEEEIIIYQQEmcozgkhhBBCCCGEkDhDcU4IIYQQQgghhMQZinNCCCGEEEIIISTOUJwTQgghhBBCCCFxhuKcEEIIIYQQQgiJMxTnhBBCCCGEEEJInKE4J4QQQgghhBBC4gzFOSGEEEIIIYQQEmcozgkhhBBCCCGEkDhDcU4IIYQQQgghhMQZinNCCCGEEEIIISTOUJwTQgghhBBCCCFxhuKcEEIIIYQQQgiJM/8PLVyLh03HFbgAAAAASUVORK5CYII=\" width=\"999\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{TGA: 0, TAA: 0.0018552875695732839, AAA: 0.06029684601113173, GCT: 0.029871664283204254}\n",
      "blue: empirical\n",
      "orange: base-wise/approximate\n"
     ]
    }
   ],
   "source": [
    "stop_positional_marginals = positional_base_usage(codon_usage_stop / stop_codon_number)\n",
    "parameters[\"base_frequency_marginals_stop_region\"] = format_matrix(stop_positional_marginals)\n",
    "\n",
    "def difference_of_approximation(empirical_close, empirical_far, approx_close, approx_far):\n",
    "    return (empirical_close - empirical_far) - (approx_close - approx_far)\n",
    "\n",
    "empirical_codon_frequencies_stop = codon_usage_stop / stop_codon_number\n",
    "empirical_codon_frequencies_general = codon_usage_general / codon_number\n",
    "\n",
    "explicit = {\"TGA\": 0, \"TAA\": empirical_codon_frequencies_stop[codon2index(\"TAA\")]}\n",
    "\n",
    "# Finding another two codons to be included:\n",
    "basewise_codon_frequencies_stop = fit_codon_usage(empirical_codon_frequencies_stop)\n",
    "basewise_codon_frequencies_general = fit_codon_usage(empirical_codon_frequencies_general)\n",
    "\n",
    "# Plot: which codons are the most informative w.r.t. stop vs general:\n",
    "w = 0.3\n",
    "plt.subplot(3,1,1)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop - empirical_codon_frequencies_general, width=w)\n",
    "plt.bar(np.arange(64)+w/2, basewise_codon_frequencies_stop - basewise_codon_frequencies_general, width=w)\n",
    "plt.xticks(range(64), codon_list, rotation=90)\n",
    "\n",
    "# Plot before:\n",
    "plt.subplot(3,1,2)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop, width=w)\n",
    "plt.bar(np.arange(64)+w/2, basewise_codon_frequencies_stop, width=w)\n",
    "plt.xticks(range(64), [\"\"] * 64)\n",
    "\n",
    "# Corrections:\n",
    "diff = difference_of_approximation(empirical_codon_frequencies_stop, \n",
    "                                   empirical_codon_frequencies_general, \n",
    "                                   basewise_codon_frequencies_stop, \n",
    "                                   basewise_codon_frequencies_general)\n",
    "diff[codon2index(\"TGA\")] = 0\n",
    "diff[codon2index(\"TAA\")] = 0\n",
    "abs_diff = np.absolute(diff)\n",
    "\n",
    "most_prominent_deviant = codon_list[np.argmax(abs_diff)]\n",
    "explicit[most_prominent_deviant] = empirical_codon_frequencies_stop[np.argmax(abs_diff)]\n",
    "diff[np.argmax(abs_diff)] = 0\n",
    "\n",
    "corrected_sum = 0\n",
    "basewise_sum = 0\n",
    "for codon in explicit:\n",
    "    corrected_sum += explicit[codon]\n",
    "    basewise_sum += basewise_codon_frequencies_stop[codon2index(codon)]\n",
    "    # print(\"+\", codon, corrected_sum, basewise_sum)\n",
    "desired_deviation = basewise_sum - corrected_sum\n",
    "# print(desired_deviation)\n",
    "\n",
    "# Find compensator:\n",
    "diff = empirical_codon_frequencies_stop - basewise_codon_frequencies_stop\n",
    "candidates = sorted([c for c in codon_list if c not in [\"TGA\", \"TAA\", most_prominent_deviant]], \n",
    "                    key=lambda codon: abs(diff[codon2index(codon)] - desired_deviation))\n",
    "\n",
    "explicit[candidates[0]] = empirical_codon_frequencies_stop[codon2index(candidates[0])]\n",
    "corrected_sum = 0\n",
    "basewise_sum = 0\n",
    "for codon in explicit:\n",
    "    corrected_sum += explicit[codon]\n",
    "    basewise_sum += basewise_codon_frequencies_stop[codon2index(codon)]\n",
    "    # print(\"+\", codon, corrected_sum, basewise_sum)\n",
    "\n",
    "explicit[candidates[0]] += basewise_sum - corrected_sum\n",
    "\n",
    "parameters[\"explicit_codon_probabilities_stop_region\"] = \"{%s}\" % \", \".join([\"%s: %s\" % (k, explicit[k]) \n",
    "                                                                             for k in explicit])\n",
    "print(parameters[\"explicit_codon_probabilities_stop_region\"])\n",
    "\n",
    "\n",
    "plt.subplot(3,1,3)\n",
    "plt.bar(np.arange(64)-w/2, empirical_codon_frequencies_stop, width=w)\n",
    "plt.bar(np.arange(64)+w/2, [explicit[codon] if codon in explicit else basewise_codon_frequencies_stop[codon2index(codon)] \n",
    "                            for codon in codon_list], width=w)\n",
    "plt.xticks(range(64), codon_list, rotation=90)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "print(\"blue: empirical\")\n",
    "print(\"orange: base-wise/approximate\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intron parameters\n",
    " - collect **intron lengths**\n",
    " - Compute intron length median/mean (for breakpoint)\n",
    " - find min and max intron length; find 5% and 95% quantiles; extend from quantiles (subsequent AUGUSTUS can still concern itself with long introns)\n",
    " - Maybe derive a maximum-likelihood approach for inferring the parameters (would be good)\n",
    " - manually set SDS- and SAS-size? Or try sliding a categorical model fit to the intron-middle onto the edges, and stop where they deviate significantly?\n",
    " - collect **empirical base-frequencies at SDS, SAS, and in between**\n",
    " \n",
    " - **For Poisson** only need the mean (+ min and max for truncation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intron lengths: 16 17.979166666666668 26\n",
      "0.6363636363636364\n"
     ]
    }
   ],
   "source": [
    "# TODO: remove the unreliable annotation with the big introns!\n",
    "intron_lengths = [l for l in intron_lengths if l < 40]\n",
    "print(\"intron lengths:\", np.min(intron_lengths), np.mean(intron_lengths), np.max(intron_lengths))\n",
    "print(np.mean(intron_counts))\n",
    "\n",
    "parameters[\"intron_minimum_length\"] = np.min(intron_lengths) - hyperparameters[\"intron_length_tolerance\"]\n",
    "parameters[\"intron_maximum_length\"] = np.max(intron_lengths) + hyperparameters[\"intron_length_tolerance\"]\n",
    "parameters[\"intron_mean_length\"] = np.mean(intron_lengths)\n",
    "\n",
    "# # v # SEQUENCE information # v # #\n",
    "\n",
    "sds_frequencies = np.zeros((hyperparameters[\"sds_size\"], 4))\n",
    "sas_frequencies = np.zeros((hyperparameters[\"sas_size\"], 4))\n",
    "intron_base_frequencies = np.zeros(4)\n",
    "\n",
    "# ADD PSEUDOCOUNTS!\n",
    "sds_frequencies[2:,:] = 1\n",
    "sas_frequencies[:-2,:] = 1\n",
    "\n",
    "for l in introns_by_length:\n",
    "    for intron in introns_by_length[l]:\n",
    "        sds = intron[:hyperparameters[\"sds_size\"]]\n",
    "        for i in range(len(sds)):\n",
    "            sds_frequencies[i, base_order.index(sds[i])] += 1\n",
    "        \n",
    "        middle = intron[hyperparameters[\"sds_size\"]:-hyperparameters[\"sas_size\"]]\n",
    "        for i in range(len(middle)):\n",
    "            intron_base_frequencies[base_order.index(middle[i])] += 1\n",
    "            \n",
    "        sas = intron[-hyperparameters[\"sas_size\"]:]\n",
    "        for i in range(len(sas)):\n",
    "            sas_frequencies[i, base_order.index(sas[i])] += 1\n",
    "        # print(\"%s %s %s\" % (sds, middle, sas))\n",
    "        \n",
    "parameters[\"intron_base_frequencies_splice_donor_site\"] = format_matrix((sds_frequencies.T / np.sum(sds_frequencies, \n",
    "                                                                                                    axis=1)).T)\n",
    "parameters[\"intron_base_frequencies_splice_acceptor_site\"] = format_matrix((sas_frequencies.T / np.sum(sas_frequencies, \n",
    "                                                                                                       axis=1)).T)\n",
    "parameters[\"intron_base_frequencies\"] = \"{%s}\" % \", \".join([str(f) \n",
    "                                                            for f \n",
    "                                                            in intron_base_frequencies / np.sum(intron_base_frequencies)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Writing the parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_filename, \"w\") as out:\n",
    "    result = \"# automatically generated parameter file\\n\"\n",
    "    result += \"# from annotation file: \" + annotation_filename + \"\\n\"\n",
    "    result += \"# and sequence file: \" + sequence_filename + \"\\n\"\n",
    "    result += \"### Hyperparameters: \\n\"\n",
    "    for key in hyperparameters:\n",
    "        result += \"# %s: %s\\n\" % (key, hyperparameters[key])\n",
    "    \n",
    "    for key in parameters:\n",
    "        result += \"%s: %s\" % (key, parameters[key])\n",
    "        result += \"\\n\"\n",
    "    \n",
    "    out.write(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
